{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "[Bam-файлы и отчеты](https://drive.google.com/drive/folders/1athn52a93obOwfuHynDCRPRhjAPB80aO?usp=sharing)"
      ],
      "metadata": {
        "id": "xCZd8SLxqBpJ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Установка FastQC"
      ],
      "metadata": {
        "id": "0UyrWfnhqErC"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WYDFSMmfkywa",
        "outputId": "574d945e-614a-4d79-d8fc-7b04dbea22bc"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-02-16 00:03:18--  https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip\n",
            "Resolving www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)... 149.155.133.4\n",
            "Connecting to www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)|149.155.133.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 10249221 (9.8M) [application/zip]\n",
            "Saving to: ‘fastqc_v0.11.9.zip’\n",
            "\n",
            "fastqc_v0.11.9.zip  100%[===================>]   9.77M  7.21MB/s    in 1.4s    \n",
            "\n",
            "2025-02-16 00:03:20 (7.21 MB/s) - ‘fastqc_v0.11.9.zip’ saved [10249221/10249221]\n",
            "\n",
            "Archive:  fastqc_v0.11.9.zip\n",
            "  inflating: FastQC/cisd-jhdf5.jar   \n",
            "   creating: FastQC/Configuration/\n",
            "  inflating: FastQC/Configuration/adapter_list.txt  \n",
            "  inflating: FastQC/Configuration/contaminant_list.txt  \n",
            "  inflating: FastQC/Configuration/limits.txt  \n",
            "  inflating: FastQC/fastqc           \n",
            "  inflating: FastQC/fastqc_icon.ico  \n",
            "   creating: FastQC/Help/\n",
            "   creating: FastQC/Help/1 Introduction/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/\n",
            "  inflating: FastQC/Help/1 Introduction/.svn/entries  \n",
            "   creating: FastQC/Help/1 Introduction/.svn/props/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/text-base/\n",
            "  inflating: FastQC/Help/1 Introduction/.svn/text-base/1.1 What is FastQC.html.svn-base  \n",
            "   creating: FastQC/Help/1 Introduction/.svn/tmp/\n",
            "   creating: FastQC/Help/1 Introduction/.svn/tmp/props/\n",
            "  inflating: FastQC/Help/1 Introduction/1.1 What is FastQC.html  \n",
            "   creating: FastQC/Help/2 Basic Operations/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/\n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/entries  \n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/props/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/text-base/\n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.1 Opening a sequence file.html.svn-base  \n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.2 Evaluating Results.html.svn-base  \n",
            "  inflating: FastQC/Help/2 Basic Operations/.svn/text-base/2.3 Saving a Report.html.svn-base  \n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/tmp/\n",
            "   creating: FastQC/Help/2 Basic Operations/.svn/tmp/props/\n",
            "  inflating: FastQC/Help/2 Basic Operations/2.1 Opening a sequence file.html  \n",
            "  inflating: FastQC/Help/2 Basic Operations/2.2 Evaluating Results.html  \n",
            "  inflating: FastQC/Help/2 Basic Operations/2.3 Saving a Report.html  \n",
            "   creating: FastQC/Help/3 Analysis Modules/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/entries  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/prop-base/\n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/duplication_levels.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/prop-base/kmer_profiles.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_gc_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_n_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_base_sequence_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_sequence_gc_content.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_sequence_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/per_tile_quality.png.svn-base  \n",
            " extracting: FastQC/Help/3 Analysis Modules/.svn/prop-base/sequence_length_distribution.png.svn-base  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/props/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/text-base/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/1 Basic Statistics.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/10 Adapter Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/11 Kmer Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/12 Per Tile Sequence Quality.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/2 Per Base Sequence Quality.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/3 Per Sequence Quality Scores.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/4 Per Base Sequence Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/5 Per Sequence GC Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/6 Per Base N Content.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/7 Sequence Length Distribution.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/8 Duplicate Sequences.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/9 Overrepresented Sequences.html.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/duplication_levels.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/kmer_profiles.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_gc_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_n_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_base_sequence_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_sequence_gc_content.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_sequence_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/per_tile_quality.png.svn-base  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/.svn/text-base/sequence_length_distribution.png.svn-base  \n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/tmp/\n",
            "   creating: FastQC/Help/3 Analysis Modules/.svn/tmp/props/\n",
            "  inflating: FastQC/Help/3 Analysis Modules/1 Basic Statistics.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/10 Adapter Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/11 Kmer Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/12 Per Tile Sequence Quality.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/2 Per Base Sequence Quality.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/3 Per Sequence Quality Scores.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/4 Per Base Sequence Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/5 Per Sequence GC Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/6 Per Base N Content.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/7 Sequence Length Distribution.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/8 Duplicate Sequences.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/9 Overrepresented Sequences.html  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/duplication_levels.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/kmer_profiles.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_gc_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_n_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_base_sequence_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_sequence_gc_content.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_sequence_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/per_tile_quality.png  \n",
            "  inflating: FastQC/Help/3 Analysis Modules/sequence_length_distribution.png  \n",
            "  inflating: FastQC/INSTALL.txt      \n",
            "  inflating: FastQC/jbzip2-0.9.jar   \n",
            "  inflating: FastQC/LICENSE          \n",
            "  inflating: FastQC/LICENSE.txt      \n",
            "  inflating: FastQC/LICENSE_JHDF5.txt  \n",
            "   creating: FastQC/net/\n",
            "   creating: FastQC/net/sourceforge/\n",
            "   creating: FastQC/net/sourceforge/iharder/\n",
            "   creating: FastQC/net/sourceforge/iharder/base64/\n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$1.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$InputStream.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64$OutputStream.class  \n",
            "  inflating: FastQC/net/sourceforge/iharder/base64/Base64.class  \n",
            "   creating: FastQC/org/\n",
            "   creating: FastQC/org/apache/\n",
            "   creating: FastQC/org/apache/commons/\n",
            "   creating: FastQC/org/apache/commons/math3/\n",
            "   creating: FastQC/org/apache/commons/math3/analysis/\n",
            "   creating: FastQC/org/apache/commons/math3/analysis/solvers/\n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/AbstractUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/AllowedSolution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BaseAbstractUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BaseUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BracketedUnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/BrentSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/UnivariateSolver.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/solvers/UnivariateSolverUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/analysis/UnivariateFunction.class  \n",
            "   creating: FastQC/org/apache/commons/math3/distribution/\n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractIntegerDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractRealDistribution$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/AbstractRealDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/BetaDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/BinomialDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/CauchyDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/ChiSquaredDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/FDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/GammaDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/HypergeometricDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/IntegerDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/NormalDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/PascalDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/PoissonDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/RealDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/SaddlePointExpansion.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/TDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/WeibullDistribution.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/distribution/ZipfDistribution.class  \n",
            "   creating: FastQC/org/apache/commons/math3/exception/\n",
            "  inflating: FastQC/org/apache/commons/math3/exception/ConvergenceException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/DimensionMismatchException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathArithmeticException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalArgumentException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalNumberException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathIllegalStateException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MathInternalError.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/MaxCountExceededException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NoBracketingException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotFiniteNumberException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotPositiveException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NotStrictlyPositiveException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NullArgumentException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NumberIsTooLargeException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/NumberIsTooSmallException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/OutOfRangeException.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/TooManyEvaluationsException.class  \n",
            "   creating: FastQC/org/apache/commons/math3/exception/util/\n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ArgUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ExceptionContext.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/ExceptionContextProvider.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/Localizable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/exception/util/LocalizedFormats.class  \n",
            "   creating: FastQC/org/apache/commons/math3/random/\n",
            "  inflating: FastQC/org/apache/commons/math3/random/AbstractWell.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/BitsStreamGenerator.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomData.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomDataImpl.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/RandomGenerator.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/random/Well19937c.class  \n",
            "   creating: FastQC/org/apache/commons/math3/special/\n",
            "  inflating: FastQC/org/apache/commons/math3/special/Beta$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Beta.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Erf.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Gamma$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/special/Gamma.class  \n",
            "   creating: FastQC/org/apache/commons/math3/util/\n",
            "  inflating: FastQC/org/apache/commons/math3/util/ArithmeticUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/ContinuedFraction.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/DoubleArray.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$ExpFracTable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$ExpIntTable.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath$lnMant.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMath.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMathCalc.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/FastMathLiteralArrays.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor$1.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor$MaxCountExceededCallback.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Incrementor.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/MathUtils.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/Precision.class  \n",
            "  inflating: FastQC/org/apache/commons/math3/util/ResizableDoubleArray.class  \n",
            "  inflating: FastQC/README.md        \n",
            "  inflating: FastQC/README.txt       \n",
            "  inflating: FastQC/RELEASE_NOTES.txt  \n",
            "  inflating: FastQC/run_fastqc.bat   \n",
            "  inflating: FastQC/sam-1.103.jar    \n",
            "   creating: FastQC/Templates/\n",
            "  inflating: FastQC/Templates/fastqc2fo.xsl  \n",
            "  inflating: FastQC/Templates/header_template.html  \n",
            "   creating: FastQC/Templates/Icons/\n",
            " extracting: FastQC/Templates/Icons/error.png  \n",
            " extracting: FastQC/Templates/Icons/fastqc_icon.png  \n",
            " extracting: FastQC/Templates/Icons/tick.png  \n",
            " extracting: FastQC/Templates/Icons/warning.png  \n",
            "   creating: FastQC/uk/\n",
            "   creating: FastQC/uk/ac/\n",
            "   creating: FastQC/uk/ac/babraham/\n",
            "   creating: FastQC/uk/ac/babraham/FastQC/\n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Analysis/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisListener.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisQueue.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisRunner.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Analysis/OfflineRunner.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Dialogs/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/AboutDialog$1.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/AboutDialog.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/FastQCTitlePanel$SmoothJLabel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/FastQCTitlePanel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Dialogs/WelcomePanel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCApplication$1.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCApplication.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCConfig.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FastQCMenuBar.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/FileFilters/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/BAMFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/CasavaFastQFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/FastQFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/GobyFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/MappedBAMFileFilter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/FileFilters/SequenceFileFilter.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Graphs/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/BaseGroup.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/LineGraph.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/QualityBoxPlot.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Graphs/TileGraph.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Help/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpDialog.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpIndexRoot$FileSorter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpIndexRoot.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPage.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPageDisplay$HelpEditor.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpPageDisplay.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Help/HelpSearchPanel.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Modules/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AbstractQCModule.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent$Adapter.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/AdapterContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/BasicStats$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/BasicStats.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/DuplicationLevel.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/GCModel.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/GCModel/GCModelValue.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent$Kmer.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/KmerContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/ModuleConfig.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/ModuleFactory.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/NContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs$OverrepresentedSeq.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs$ResultsTable.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerBaseQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerBaseSequenceContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerSequenceGCContent.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerSequenceQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/PerTileQualityScores.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/QCModule.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Modules/SequenceLengthDistribution.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Report/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Report/HTMLReportArchive.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Report/stylesheet.txt  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Resources/\n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/error.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon.png  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon.svg  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon_100.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/tick.png  \n",
            " extracting: FastQC/uk/ac/babraham/FastQC/Resources/warning.png  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Results/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Results/ResultsPanel$ModuleRenderer.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Results/ResultsPanel.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/BAMFile.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/Contaminant.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/ContaminantHit.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/ContaminentFinder.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Fast5File.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/FastQFile.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Sequence/QualityEncoding/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/QualityEncoding/PhredEncoding.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/Sequence.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFactory.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFile.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFileGroup.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFormatException.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Statistics/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Statistics/NormalDistribution.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Statistics/PearsonCorrelation.class  \n",
            "   creating: FastQC/uk/ac/babraham/FastQC/Utilities/\n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/CasavaBasename.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/HotColdColourGradient.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/ImageToBase64.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/MultiMemberGZIPInputStream.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/NameFormatException.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/NanoporeBasename.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/QualityCount.class  \n",
            "  inflating: FastQC/uk/ac/babraham/FastQC/Utilities/RGB.class  \n"
          ]
        }
      ],
      "source": [
        "!wget https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip\n",
        "!unzip fastqc_v0.11.9.zip\n",
        "!chmod a+x FastQC/fastqc"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Установка Bismark с зависимостями (+bowtie2, samtools)"
      ],
      "metadata": {
        "id": "Bp7J-ug7skGi"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "! pip install -q condacolab"
      ],
      "metadata": {
        "id": "mdx767DZEWnL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import condacolab\n",
        "condacolab.install()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Odze-zcwsk3d",
        "outputId": "175e24e4-5149-4743-f8e3-ff0cc3ba49a3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "✨🍰✨ Everything looks OK!\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import condacolab\n",
        "condacolab.check()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FFDw2DBWspJ8",
        "outputId": "537f4770-23d6-4899-c895-9a53170f6bfe"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "✨🍰✨ Everything looks OK!\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! conda install -c bioconda bismark"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "4yzrFcbasqGt",
        "outputId": "47931e1b-de35-47c4-8d98-5f01fefed294"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Channels:\n",
            " - bioconda\n",
            " - conda-forge\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Solving environment: \\ \b\b| \b\b/ \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - bismark\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    bismark-0.24.2             |       hdfd78af_0         7.7 MB  bioconda\n",
            "    bowtie2-2.5.4              |       he96a11b_5        14.3 MB  bioconda\n",
            "    ca-certificates-2025.1.31  |       hbcca054_0         154 KB  conda-forge\n",
            "    certifi-2025.1.31          |     pyhd8ed1ab_0         159 KB  conda-forge\n",
            "    conda-24.11.3              |  py311h38be061_0         1.1 MB  conda-forge\n",
            "    coreutils-8.25             |                1         9.1 MB  bioconda\n",
            "    hisat2-2.2.1               |       h503566f_8        15.9 MB  bioconda\n",
            "    htslib-1.21                |       h566b1c6_1         3.0 MB  bioconda\n",
            "    libdeflate-1.23            |       h4ddbbb0_0          71 KB  conda-forge\n",
            "    minimap2-2.1.1             |                0         255 KB  bioconda\n",
            "    openssl-3.4.1              |       h7b32b05_0         2.8 MB  conda-forge\n",
            "    perl-5.32.1                | 7_hd590300_perl5        12.7 MB  conda-forge\n",
            "    samtools-1.21              |       h96c455f_1         476 KB  bioconda\n",
            "    zlib-1.3.1                 |       hb9d3cd8_2          90 KB  conda-forge\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:        67.9 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  bismark            bioconda/noarch::bismark-0.24.2-hdfd78af_0 \n",
            "  bowtie2            bioconda/linux-64::bowtie2-2.5.4-he96a11b_5 \n",
            "  coreutils          bioconda/linux-64::coreutils-8.25-1 \n",
            "  hisat2             bioconda/linux-64::hisat2-2.2.1-h503566f_8 \n",
            "  htslib             bioconda/linux-64::htslib-1.21-h566b1c6_1 \n",
            "  libdeflate         conda-forge/linux-64::libdeflate-1.23-h4ddbbb0_0 \n",
            "  minimap2           bioconda/linux-64::minimap2-2.1.1-0 \n",
            "  perl               conda-forge/linux-64::perl-5.32.1-7_hd590300_perl5 \n",
            "  samtools           bioconda/linux-64::samtools-1.21-h96c455f_1 \n",
            "  zlib               conda-forge/linux-64::zlib-1.3.1-hb9d3cd8_2 \n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "  ca-certificates                     2024.12.14-hbcca054_0 --> 2025.1.31-hbcca054_0 \n",
            "  certifi                           2024.12.14-pyhd8ed1ab_0 --> 2025.1.31-pyhd8ed1ab_0 \n",
            "  conda                             24.11.2-py311h38be061_1 --> 24.11.3-py311h38be061_0 \n",
            "  openssl                                  3.4.0-h7b32b05_1 --> 3.4.1-h7b32b05_0 \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages:\n",
            "hisat2-2.2.1         | 15.9 MB   | :   0% 0/1 [00:00<?, ?it/s]\n",
            "bowtie2-2.5.4        | 14.3 MB   | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "coreutils-8.25       | 9.1 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "bismark-0.24.2       | 7.7 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "htslib-1.21          | 3.0 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "openssl-3.4.1        | 2.8 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "conda-24.11.3        | 1.1 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "samtools-1.21        | 476 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "minimap2-2.1.1       | 255 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "certifi-2025.1.31    | 159 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "ca-certificates-2025 | 154 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "zlib-1.3.1           | 90 KB     | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libdeflate-1.23      | 71 KB     | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "hisat2-2.2.1         | 15.9 MB   | :   0% 0.0009822275799240931/1 [00:00<01:49, 109.60s/it]\n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | :   0% 0.0024555502907835257/1 [00:00<00:44, 45.09s/it]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "coreutils-8.25       | 9.1 MB    | :   0% 0.0034200769412939327/1 [00:00<00:32, 32.39s/it]\u001b[A\u001b[A\u001b[A\n",
            "bowtie2-2.5.4        | 14.3 MB   | :   0% 0.0010928429591871448/1 [00:00<01:45, 105.23s/it]\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "hisat2-2.2.1         | 15.9 MB   | :  12% 0.11590285443104299/1 [00:00<00:01,  1.52s/it]   \n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | :  19% 0.19398847297189853/1 [00:00<00:00,  1.09it/s]  \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "coreutils-8.25       | 9.1 MB    | :  25% 0.254795732126398/1 [00:00<00:00,  1.42it/s]    \u001b[A\u001b[A\u001b[A\n",
            "bowtie2-2.5.4        | 14.3 MB   | :  14% 0.13769821285758024/1 [00:00<00:01,  1.32s/it]   \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "bismark-0.24.2       | 7.7 MB    | :  76% 0.7593393265558255/1 [00:00<00:00,  3.09it/s]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "coreutils-8.25       | 9.1 MB    | :  63% 0.6310041956687306/1 [00:00<00:00,  2.46it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "hisat2-2.2.1         | 15.9 MB   | :  35% 0.347708563293129/1 [00:00<00:00,  1.34it/s]  \n",
            "bowtie2-2.5.4        | 14.3 MB   | :  43% 0.43167296887892215/1 [00:00<00:00,  1.64it/s]\u001b[A\n",
            "\n",
            "hisat2-2.2.1         | 15.9 MB   | :  73% 0.727830636723753/1 [00:00<00:00,  2.27it/s]\n",
            "\n",
            "\n",
            "\n",
            "bismark-0.24.2       | 7.7 MB    | : 100% 1.0/1 [00:00<00:00,  3.09it/s]               \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "bowtie2-2.5.4        | 14.3 MB   | :  73% 0.7311119396961998/1 [00:00<00:00,  2.07it/s] \u001b[A\n",
            "\n",
            "\n",
            "coreutils-8.25       | 9.1 MB    | : 100% 1.0/1 [00:00<00:00,  2.57it/s]               \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "coreutils-8.25       | 9.1 MB    | : 100% 1.0/1 [00:00<00:00,  2.57it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | : 100% 1.0/1 [00:00<00:00,  2.96it/s]               \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "htslib-1.21          | 3.0 MB    | :   1% 0.005149561702649271/1 [00:00<01:33, 93.59s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "hisat2-2.2.1         | 15.9 MB   | :  99% 0.9861564902437896/1 [00:00<00:00,  2.36it/s]\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "conda-24.11.3        | 1.1 MB    | :   1% 0.01364371451460054/1 [00:00<00:37, 38.22s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "conda-24.11.3        | 1.1 MB    | : 100% 1.0/1 [00:00<00:00, 38.22s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "openssl-3.4.1        | 2.8 MB    | : 100% 1.0/1 [00:00<00:00, 87.76s/it]                  \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "htslib-1.21          | 3.0 MB    | : 100% 1.0/1 [00:00<00:00, 93.59s/it]                 \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "samtools-1.21        | 476 KB    | :   3% 0.03362972815422383/1 [00:00<00:17, 17.65s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "bowtie2-2.5.4        | 14.3 MB   | : 100% 1.0/1 [00:00<00:00,  1.89it/s]               \u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "minimap2-2.1.1       | 255 KB    | :   6% 0.0627410133378265/1 [00:00<00:08,  9.54s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "bowtie2-2.5.4        | 14.3 MB   | : 100% 1.0/1 [00:00<00:00,  1.89it/s]\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "minimap2-2.1.1       | 255 KB    | : 100% 1.0/1 [00:00<00:00,  9.54s/it]               \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "samtools-1.21        | 476 KB    | : 100% 1.0/1 [00:00<00:00, 17.65s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "certifi-2025.1.31    | 159 KB    | :  10% 0.1006876801396255/1 [00:00<00:05,  6.01s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "hisat2-2.2.1         | 15.9 MB   | : 100% 1.0/1 [00:00<00:00,  2.36it/s]               \n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "zlib-1.3.1           | 90 KB     | :  18% 0.17753505407104003/1 [00:00<00:02,  3.62s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "zlib-1.3.1           | 90 KB     | : 100% 1.0/1 [00:00<00:00,  3.62s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libdeflate-1.23      | 71 KB     | :  23% 0.22675247387724032/1 [00:00<00:02,  2.87s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libdeflate-1.23      | 71 KB     | : 100% 1.0/1 [00:00<00:00,  2.87s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "ca-certificates-2025 | 154 KB    | :  10% 0.10360178065560502/1 [00:00<00:05,  6.30s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "ca-certificates-2025 | 154 KB    | : 100% 1.0/1 [00:00<00:00,  6.30s/it]                \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "bismark-0.24.2       | 7.7 MB    | : 100% 1.0/1 [00:02<00:00,  3.09it/s]\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "coreutils-8.25       | 9.1 MB    | : 100% 1.0/1 [00:03<00:00,  2.57it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "conda-24.11.3        | 1.1 MB    | : 100% 1.0/1 [00:03<00:00,  3.60s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "conda-24.11.3        | 1.1 MB    | : 100% 1.0/1 [00:03<00:00,  3.60s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "openssl-3.4.1        | 2.8 MB    | : 100% 1.0/1 [00:03<00:00,  3.76s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "openssl-3.4.1        | 2.8 MB    | : 100% 1.0/1 [00:03<00:00,  3.76s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "perl-5.32.1          | 12.7 MB   | : 100% 1.0/1 [00:04<00:00,  2.96it/s]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "htslib-1.21          | 3.0 MB    | : 100% 1.0/1 [00:04<00:00,  4.34s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "htslib-1.21          | 3.0 MB    | : 100% 1.0/1 [00:04<00:00,  4.34s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "minimap2-2.1.1       | 255 KB    | : 100% 1.0/1 [00:04<00:00,  4.43s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "minimap2-2.1.1       | 255 KB    | : 100% 1.0/1 [00:04<00:00,  4.43s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "samtools-1.21        | 476 KB    | : 100% 1.0/1 [00:04<00:00,  4.50s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "samtools-1.21        | 476 KB    | : 100% 1.0/1 [00:04<00:00,  4.50s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "certifi-2025.1.31    | 159 KB    | : 100% 1.0/1 [00:04<00:00,  4.61s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "certifi-2025.1.31    | 159 KB    | : 100% 1.0/1 [00:04<00:00,  4.61s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "bowtie2-2.5.4        | 14.3 MB   | : 100% 1.0/1 [00:06<00:00,  1.89it/s]\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "zlib-1.3.1           | 90 KB     | : 100% 1.0/1 [00:06<00:00,  6.68s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "zlib-1.3.1           | 90 KB     | : 100% 1.0/1 [00:06<00:00,  6.68s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libdeflate-1.23      | 71 KB     | : 100% 1.0/1 [00:06<00:00,  6.80s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "libdeflate-1.23      | 71 KB     | : 100% 1.0/1 [00:06<00:00,  6.80s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "ca-certificates-2025 | 154 KB    | : 100% 1.0/1 [00:06<00:00,  6.55s/it]\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \n",
            "                                                                        \u001b[A\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\u001b[A\n",
            "Preparing transaction: - \b\bdone\n",
            "Verifying transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Executing transaction: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! samtools view"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nf_I3IZks0zZ",
        "outputId": "f058b676-8990-4e21-b361-fb89172995dc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Usage: samtools view [options] <in.bam>|<in.sam>|<in.cram> [region ...]\n",
            "\n",
            "Output options:\n",
            "  -b, --bam                  Output BAM\n",
            "  -C, --cram                 Output CRAM (requires -T)\n",
            "  -1, --fast                 Use fast BAM compression (and default to --bam)\n",
            "  -u, --uncompressed         Uncompressed BAM output (and default to --bam)\n",
            "  -h, --with-header          Include header in SAM output\n",
            "  -H, --header-only          Print SAM header only (no alignments)\n",
            "      --no-header            Print SAM alignment records only [default]\n",
            "  -c, --count                Print only the count of matching records\n",
            "  -o, --output FILE          Write output to FILE [standard output]\n",
            "  -U, --unoutput FILE, --output-unselected FILE\n",
            "                             Output reads not selected by filters to FILE\n",
            "  -p, --unmap                Set flag to UNMAP on reads not selected\n",
            "                             then write to output file.\n",
            "  -P, --fetch-pairs          Retrieve complete pairs even when outside of region\n",
            "Input options:\n",
            "  -t, --fai-reference FILE   FILE listing reference names and lengths\n",
            "  -M, --use-index            Use index and multi-region iterator for regions\n",
            "      --region[s]-file FILE  Use index to include only reads overlapping FILE\n",
            "  -X, --customized-index     Expect extra index file argument after <in.bam>\n",
            "\n",
            "Filtering options (Only include in output reads that...):\n",
            "  -L, --target[s]-file FILE  ...overlap (BED) regions in FILE\n",
            "  -N, --qname-file [^]FILE   ...whose read name is listed in FILE (\"^\" negates)\n",
            "  -r, --read-group STR       ...are in read group STR\n",
            "  -R, --read-group-file [^]FILE\n",
            "                             ...are in a read group listed in FILE\n",
            "  -d, --tag STR1[:STR2]      ...have a tag STR1 (with associated value STR2)\n",
            "  -D, --tag-file STR:FILE    ...have a tag STR whose value is listed in FILE\n",
            "  -q, --min-MQ INT           ...have mapping quality >= INT\n",
            "  -l, --library STR          ...are in library STR\n",
            "  -m, --min-qlen INT         ...cover >= INT query bases (as measured via CIGAR)\n",
            "  -e, --expr STR             ...match the filter expression STR\n",
            "  -f, --require-flags FLAG   ...have all of the FLAGs present\n",
            "  -F, --excl[ude]-flags FLAG ...have none of the FLAGs present\n",
            "      --rf, --incl-flags, --include-flags FLAG\n",
            "                             ...have some of the FLAGs present\n",
            "  -G FLAG                    EXCLUDE reads with all of the FLAGs present\n",
            "      --subsample FLOAT      Keep only FLOAT fraction of templates/read pairs\n",
            "      --subsample-seed INT   Influence WHICH reads are kept in subsampling [0]\n",
            "  -s INT.FRAC                Same as --subsample 0.FRAC --subsample-seed INT\n",
            "\n",
            "Processing options:\n",
            "      --add-flags FLAG       Add FLAGs to reads\n",
            "      --remove-flags FLAG    Remove FLAGs from reads\n",
            "  -x, --remove-tag STR\n",
            "               Comma-separated read tags to strip (repeatable) [null]\n",
            "      --keep-tag STR\n",
            "               Comma-separated read tags to preserve (repeatable) [null].\n",
            "               Equivalent to \"-x ^STR\"\n",
            "  -B, --remove-B             Collapse the backward CIGAR operation\n",
            "  -z, --sanitize FLAGS       Perform sanitity checking and fixing on records.\n",
            "                             FLAGS is comma separated (see manual). [off]\n",
            "\n",
            "General options:\n",
            "  -?, --help   Print long help, including note about region specification\n",
            "  -S           Ignored (input format is auto-detected)\n",
            "      --no-PG  Do not add a PG line\n",
            "      --input-fmt-option OPT[=VAL]\n",
            "               Specify a single input file format option in the form\n",
            "               of OPTION or OPTION=VALUE\n",
            "  -O, --output-fmt FORMAT[,OPT[=VAL]]...\n",
            "               Specify output format (SAM, BAM, CRAM)\n",
            "      --output-fmt-option OPT[=VAL]\n",
            "               Specify a single output file format option in the form\n",
            "               of OPTION or OPTION=VALUE\n",
            "  -T, --reference FILE\n",
            "               Reference sequence FASTA FILE [null]\n",
            "  -@, --threads INT\n",
            "               Number of additional threads to use [0]\n",
            "      --write-index\n",
            "               Automatically index the output files [off]\n",
            "      --verbosity INT\n",
            "               Set level of verbosity\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "1. Скачиваем чтения из SRA с помощью SRAtoolkit или любым другим способом\n",
        "\n",
        "2. Индексируем геном\n",
        "\n",
        "3. Строим выравнивание\n",
        "\n",
        "(Описание шагов 2-3 см https://www.bioinformatics.babraham.ac.uk/projects/bismark/Bismark_User_Guide.pdf)"
      ],
      "metadata": {
        "id": "J-923BR4tIcL"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Можно получать данные секвенировия с ENA. Например, https://www.ebi.ac.uk/ena/browser/view/SRR5836473"
      ],
      "metadata": {
        "id": "VgNip6acn9gR"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "! wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/003/SRR5836473/SRR5836473_1.fastq.gz\n",
        "# ! wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/003/SRR5836473/SRR5836473_2.fastq.gz"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7DBMAu61n2_G",
        "outputId": "df52cddf-1699-4a88-81b2-74e0e1ce1af5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-02-09 12:23:14--  ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR583/003/SRR5836473/SRR5836473_1.fastq.gz\n",
            "           => ‘SRR5836473_1.fastq.gz’\n",
            "Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.193.165\n",
            "Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.193.165|:21... connected.\n",
            "Logging in as anonymous ... Logged in!\n",
            "==> SYST ... done.    ==> PWD ... done.\n",
            "==> TYPE I ... done.  ==> CWD (1) /vol1/fastq/SRR583/003/SRR5836473 ... done.\n",
            "==> SIZE SRR5836473_1.fastq.gz ... 6856386252\n",
            "==> PASV ... done.    ==> RETR SRR5836473_1.fastq.gz ... done.\n",
            "Length: 6856386252 (6.4G) (unauthoritative)\n",
            "\n",
            "SRR5836473_1.fastq. 100%[===================>]   6.38G  35.7MB/s    in 2m 53s  \n",
            "\n",
            "2025-02-09 12:26:09 (37.9 MB/s) - ‘SRR5836473_1.fastq.gz’ saved [6856386252]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir /content/fastqc_out\n",
        "!gzip -dk SRR5836473_1.fastq.gz"
      ],
      "metadata": {
        "id": "2WtUFePpvC1C"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!./FastQC/fastqc -o fastqc_out SRR5836473_1.fastq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IaW8vWz5xP1W",
        "outputId": "725da7a9-396c-41fb-c55f-1ee34b349ce3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Started analysis of SRR5836473_1.fastq\n",
            "Approx 5% complete for SRR5836473_1.fastq\n",
            "Approx 10% complete for SRR5836473_1.fastq\n",
            "Approx 15% complete for SRR5836473_1.fastq\n",
            "Approx 20% complete for SRR5836473_1.fastq\n",
            "Approx 25% complete for SRR5836473_1.fastq\n",
            "Approx 30% complete for SRR5836473_1.fastq\n",
            "Approx 35% complete for SRR5836473_1.fastq\n",
            "Approx 40% complete for SRR5836473_1.fastq\n",
            "Approx 45% complete for SRR5836473_1.fastq\n",
            "Approx 50% complete for SRR5836473_1.fastq\n",
            "Approx 55% complete for SRR5836473_1.fastq\n",
            "Approx 60% complete for SRR5836473_1.fastq\n",
            "Approx 65% complete for SRR5836473_1.fastq\n",
            "Approx 70% complete for SRR5836473_1.fastq\n",
            "Approx 75% complete for SRR5836473_1.fastq\n",
            "Approx 80% complete for SRR5836473_1.fastq\n",
            "Approx 85% complete for SRR5836473_1.fastq\n",
            "Approx 90% complete for SRR5836473_1.fastq\n",
            "Approx 95% complete for SRR5836473_1.fastq\n",
            "Analysis complete for SRR5836473_1.fastq\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "!cp /content/drive/MyDrive/SRR5836473_1_bismark_bt2_pe.bam /content/SRR5836473_1_bismark_bt2_pe.bam\n",
        "!cp /content/drive/MyDrive/SRR3824222_1_bismark_bt2_pe.bam /content/SRR3824222_1_bismark_bt2_pe.bam\n",
        "!cp /content/drive/MyDrive/SRR5836475_1_bismark_bt2_pe.bam /content/SRR5836475_1_bismark_bt2_pe.bam\n",
        "!cp /content/drive/MyDrive/SRR5836473_1_bismark_bt2_PE_report.txt /content/SRR5836473_1_bismark_bt2_PE_report.txt\n",
        "!cp /content/drive/MyDrive/SRR3824222_1_bismark_bt2_PE_report.txt /content/SRR3824222_1_bismark_bt2_PE_report.txt\n",
        "!cp /content/drive/MyDrive/SRR5836475_1_bismark_bt2_PE_report.txt /content/SRR5836475_1_bismark_bt2_PE_report.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "PVUC9c3A3GBa",
        "outputId": "090a7820-e1c5-4950-8f99-26e4ae5ff963"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! ls -lh"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "mFZzhcNIQUxc",
        "outputId": "07f0a27b-0668-4592-d0a6-0c148618ecd8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "total 42G\n",
            "-rw-r--r-- 1 root root  21K Feb  9 12:32 condacolab_install.log\n",
            "drwx------ 6 root root 4.0K Feb  9 15:06 drive\n",
            "drwxr-xr-x 8 root root 4.0K Feb  9 12:06 FastQC\n",
            "drwxr-xr-x 2 root root 4.0K Feb  9 13:11 fastqc_out\n",
            "-rw-r--r-- 1 root root 9.8M Jan 15  2020 fastqc_v0.11.9.zip\n",
            "drwxr-xr-x 1 root root 4.0K Feb  6 14:19 sample_data\n",
            "-rw------- 1 root root 1.9G Feb  9 15:07 SRR3824222_1_bismark_bt2_pe.bam\n",
            "-rw------- 1 root root 1.8K Feb  9 15:07 SRR3824222_1_bismark_bt2_PE_report.txt\n",
            "-rw------- 1 root root 769M Feb  9 15:06 SRR5836473_1_bismark_bt2_pe.bam\n",
            "-rw------- 1 root root 1.8K Feb  9 15:07 SRR5836473_1_bismark_bt2_PE_report.txt\n",
            "-rw-r--r-- 1 root root  32G Feb  9 12:26 SRR5836473_1.fastq\n",
            "-rw-r--r-- 1 root root 6.4G Feb  9 12:26 SRR5836473_1.fastq.gz\n",
            "-rw------- 1 root root 1.2G Feb  9 15:07 SRR5836475_1_bismark_bt2_pe.bam\n",
            "-rw------- 1 root root 1.8K Feb  9 15:07 SRR5836475_1_bismark_bt2_PE_report.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! samtools view SRR5836473_1_bismark_bt2_pe.bam | head"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "T_57U3cCQW-E",
        "outputId": "d5492eb1-024a-414d-f739-08be47835327"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "SRR5836473.41_41_length=140\t99\tchr11\t40076664\t0\t140M\t=\t40076703\t178\tAATTTTTTTGGGGATTGTATTGAATTTGTAGATTGTTTTTGGTAAGAGAGTTATTTTTATTATATTGATTTTGTTAATTTATGAATATGGGAGATTTTTTTATTTTTTGAGATTTTTTTTAATTTTTTTTTTTAGAGATT\tKKKKKKKKKKKKKKKFKKKKKKKKKKKKKKFKKKKKKKKKFKKKKKKKKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKF7<77FF\tNM:i:28\tMD:Z:6G0A9C7C9C6C4T2C0C7C9C0C2C0C3C0C5C9C3C0C2C2C6C2C8C3C2C5C1\tXM:Z:.................h.......x.........h......h.......hh.......h.........hx..hh...hh.....h.........h...hh..h..x......h..h........h...h..x.....h.\tXR:Z:CT\tXG:Z:CT\n",
            "SRR5836473.41_41_length=140\t147\tchr11\t40076703\t0\t139M\t=\t40076664\t-178\tTGGTAAGAGAGTTATTTTTATTATATTGATTTTGTTAATTTATGAATATGGGAGATTTTTTTATTTTTTGAGATTTTTTTTAATTTTTTTTTTTAGAGATTTGAAGTTTTTATTATATAGATTTTTTATTTTTTTAGTG\tKKKKKAAFFFKKKKKKKKKKKKKFFKKKKKKKKKKFAFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFKKKKKKKKKKKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK\tNM:i:31\tMD:Z:3C4T2C0C7C9C0C2C0C3C0C5C9C3C0C2C2C6C2C8C3C2C5C8C4C3A4C3C1C2C6T0\tXM:Z:...h.......hh.......h.........hx..hh...hh.....h.........h...hh..h..x......h..h........h...h..x.....h........h....h........h...h.h..h.......\tXR:Z:GA\tXG:Z:CT\n",
            "SRR5836473.231_231_length=140\t83\tchr11\t48401999\t0\t140M\t=\t48401992\t-147\tACAAATTACTCCTTCTAATTCATTAAAAAATTAAATTAAAATTATAATAAAAATTACATTAAATCTATAAATTACTTTTAACAAAATAACCATTTTTACAATATTAATCCTACCAATCCATAAACACAAAAAAACTTTCC\tKKKKKKKKF<KKKKKKKKKFKKKKKKKKKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFKKKKKKKKKKKKKKKK\tNM:i:34\tMD:Z:0G2G3G13G2G2G4G1G2G0G6G2G0G0G0G3G4G5G2G3G5G0G3G3G13G2G5G9G1G2T1G0G1G1T6\tXM:Z:h..x...h.............z..h..h....h.h..hh......h..hhhh...h....h.....x..h...h.....hh...h...h.............h..h.....x.........h.h....hh.h........\tXR:Z:CT\tXG:Z:GA\n",
            "SRR5836473.231_231_length=140\t163\tchr11\t48401992\t0\t139M\t=\t48401999\t147\tTAAATTTACAAATTACTCCTTCTAATTCATTAAAAAATTAAATTAAAATTATAATAAAAATTACATTAAATCTATAAATTACTTTTAACAAAATAACCATTTTTACAATATTAATCCTACCAATCCATAAACACAAAAA\tFFKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKAKFFKKKKKKKKKKKKKKKKFKAF<FF<FKKKKKKKKKKFKKKK<FFKAFF<FKF<FK7AF7A,7<AAF<KF\tNM:i:35\tMD:Z:1G3A1G2G3G13G2G2G4G1G2G0G6G2G0G0G0G3G4G5G2G3G5G0G3G3G13G2G5G9G1G2T1G0G1G0\tXM:Z:.h.....h..x...h.............z..h..h....h.h..hh......h..hhhh...h....h.....x..h...h.....hh...h...h.............h..h.....x.........h.h....hh.h\tXR:Z:GA\tXG:Z:GA\n",
            "SRR5836473.1421_1421_length=140\t99\tchr11\t30560802\t0\t140M\t=\t30560952\t289\tTGTTGAGGGGTATTTGGGTTTTTTTTAGTTTTCGGTTATTATAAATAAGGTTGTTATGAATATAGTGTAGTATGTGTTTTTTTTATTGGTTGGAATATTTTTTGGATATATGTTTAGATGAGGTATTGTGGGATTTTTTG\tKKKKKKKKFKFKKKKKKKFKKKKKFKKKKKAK,FAKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFKKKKKKKKKKKKKKKKKKAKKKKKKKKKKKKKKKKKKKKKKKKKKKKKAKKKKAFFK<AFKKKKK\tNM:i:30\tMD:Z:10C2C6C3C0C2C2C0T2C9C4C2C6C6G2A6C3C3C0C8C2C2C10C0C0C3A9C5C0C1C2\tXM:Z:..........h..x......h...hx..h..x...h.........h....x..h......h................h...h...xz........h..h..x..........hhx.............z.....hh.x..\tXR:Z:CT\tXG:Z:CT\n",
            "SRR5836473.1421_1421_length=140\t147\tchr11\t30560952\t0\t139M\t=\t30560802\t-289\tGTTTAATTTTTTGAAGAATTGTTAGATTGATTTTTAGAGTGGTTGTAAAAGTTTGTAATTTTATTAATAATGGAGGAGTGTTTTTTTTTTTTTATATTTTTTTTAGTATTTGTTGTTATTTGAATTTTTGATTTTAGTT\t<<<<<<KF<KKKKKFFFKKAA<KAFKFKKKKKKKKKFFFFFFKKKFFAFAAKKKKFKKKKKKKKKKKKKKKKKKKKFKKKKKKKKKKKKKKKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKFKKKKKKKFKKKKKKFF\tNM:i:41\tMD:Z:2C0C6C3G3C2C0C3C6C0C12C3C3C3C0C0C1C0C2C14C0C1C3C1C0C1C1G0C0C1C0G0C0C2C2C2C3C1C0C12C5C0\tXM:Z:..hh......x.......x..hx...x......hx................h...h...hhh.hh..h..............hh.h...h.hh.x..hh.z.hx..h..x..x...h.hx............h.....h\tXR:Z:GA\tXG:Z:CT\n",
            "SRR5836473.2061_2061_length=140\t99\tchr11\t73270947\t0\t139M\t=\t73270962\t154\tATAAGAGATAGTTATATTAGGGTTTTGTTAGTAAAATTTTGTTGGTATTTGTAGTAGTGTTTGGGTTTGGTGGTTGTTTATGGGATGGATTTATGGGTGGGGTAGTTTTTGGATGGTTTTTTTTTTTATTTTAGTTTTG\tKKKKKKKKKKKFKKKKKKKKKKFKKKKKKKKKKKKKKFKKKKKKKKKKKKKKKKFKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFFKKKKKKKFKKKK7AFAFKKKKKF7FKKKKKKFKFAKKKKKKKKF7FFKF\tNM:i:31\tMD:Z:1C6C2C5C5C0C3C2C5C3C3C2A2C1A4A1C1A27C0C10C3C1C9C2C0C2C0C2C1C2C1C2\tXM:Z:.h......x..h.....x.....hx...x..h.....h...x...h.....h........h.............................hh..........x...h.x.........h..hh..hh..h.x..h.x..\tXR:Z:CT\tXG:Z:CT\n",
            "SRR5836473.2061_2061_length=140\t147\tchr11\t73270962\t0\t139M\t=\t73270947\t-154\tATTAGGGTTTTGTTAGTAAAATTTTGTTGGTATTTGTAGTAGTGTTTGGGTTTGGTGGTTGTTTATGGGATGGATTTATGGGTGGGGTAGTTTTTGGATGGTTTTTTTTTTTATTTTAGTTTTGAATTTTGTTTTTGTA\tA7<AAAAA77AKKKFKKKFKKKKAKF<KKKFKKFKKFKAAKKKKFKAKKKKKKKFKKKKKKKKKKKKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK\tNM:i:31\tMD:Z:2C5C0C3C2C5C3C3C2A2C1A4A1C1A27C0C10C3C1C9C2C0C2C0C2C1C2C1C4C5C1C4\tXM:Z:..x.....hx...x..h.....h...x...h.....h........h.............................hh..........x...h.x.........h..hh..hh..h.x..h.x....h.....h.x....\tXR:Z:GA\tXG:Z:CT\n",
            "SRR5836473.3331_3331_length=140\t99\tchr11\t43350802\t40\t140M\t=\t43350876\t214\tTTTTAATGTAAGTTGTATAAGGAGTATAATTAGGTTTAGGTGATGTTGAAAATAATATAATTTTATTTATAAATATTATTATTATTATGATAAAGGGAAAGGTGATTATTTTTAAAAGATTATGATGGTTTGTTGGTAAT\tKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFKFFKKKKKKKKKF\tNM:i:29\tMD:Z:0C5C1C6C0T0C7G0C3C3C1C23C2C1C0C6C2C1C0C1C23C0C3C0C0C7C7C0C6C3\tXM:Z:h.....z.h......h.h........h...x...h.x.......................h..h.hh......h..h.hh.h.......................hh...hhh.......h.......hx......h...\tXR:Z:CT\tXG:Z:CT\n",
            "SRR5836473.3331_3331_length=140\t147\tchr11\t43350876\t40\t140M\t=\t43350802\t-214\tATTATTATTATTATGATAAAGGGAAAGGTGATTATTTTTAAAAGATTATGATGGTTTGTTGGTAATGTTATTGTTATACGTAGTATGTTTTGGTGTTATTTGGATTTTGTATTTTTTATAGTTTTTTTATTTTGTTAATT\tKFKKKKK<KKKKKFKAKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKFKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKKK\tNM:i:34\tMD:Z:2C1C0C1C23C0C3C0C0C7C7C0C6C4C2C2C6C2C3C0C7C2C0C3C0C1C1C4C0C11C1C0C4C3C0\tXM:Z:..h.hh.h.......................hh...hhh.......h.......hx......h....h..x..h....Z.x..h...hh.......h..xz...hh.z.h....hh...........h.hh....h...h\tXR:Z:GA\tXG:Z:CT\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! samtools sort SRR5836473_1_bismark_bt2_pe.bam -o sorted_8cell.bam\n",
        "! samtools sort SRR3824222_1_bismark_bt2_pe.bam -o sorted_epiblast.bam\n",
        "! samtools sort SRR5836475_1_bismark_bt2_pe.bam -o sorted_icm.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "w3Gkcu_LQbhS",
        "outputId": "64cad5aa-e474-4700-b64d-9d5dfc9a8510"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[bam_sort_core] merging from 4 files and 1 in-memory blocks...\n",
            "[bam_sort_core] merging from 9 files and 1 in-memory blocks...\n",
            "[bam_sort_core] merging from 5 files and 1 in-memory blocks...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! samtools index  sorted_8cell.bam\n",
        "! samtools index  sorted_epiblast.bam\n",
        "! samtools index  sorted_icm.bam"
      ],
      "metadata": {
        "id": "bMl63S7AT_3e"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "! samtools view sorted_8cell.bam chr11:11347700-11367700 | cut -f1 | sort -u | wc -l\n",
        "! samtools view sorted_epiblast.bam chr11:11347700-11367700 | cut -f1 | sort -u | wc -l\n",
        "! samtools view sorted_icm.bam chr11:11347700-11367700 | cut -f1 | sort -u | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tmjmOhC7VG8N",
        "outputId": "145982ec-ca93-46b8-d4bd-006ae317f875"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "551\n",
            "1344\n",
            "797\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! samtools view sorted_8cell.bam chr11:40185800-40195800 | cut -f1 | sort -u | wc -l\n",
        "! samtools view sorted_epiblast.bam chr11:40185800-40195800 | cut -f1 | sort -u | wc -l\n",
        "! samtools view sorted_icm.bam chr11:40185800-40195800 | cut -f1 | sort -u | wc -l"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "t-ZXD4DtVMRu",
        "outputId": "5bab3f03-323d-446f-b81d-1a4446e9433c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "194\n",
            "565\n",
            "274\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "df = pd.DataFrame({\n",
        "        'Образцы': ['8cell', 'epiblast', 'ICM'],\n",
        "        'chr11:11347700-11367700': [551, 1344, 797],\n",
        "        'chr11:40185800-40195800': [194, 565, 274]\n",
        "     })\n",
        "df"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 142
        },
        "id": "v9ceV8qmi9lW",
        "outputId": "812ea5eb-01af-4a64-84c7-bf44b98d6215"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "    Образцы  chr11:11347700-11367700  chr11:40185800-40195800\n",
              "0     8cell                      551                      194\n",
              "1  epiblast                     1344                      565\n",
              "2       ICM                      797                      274"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-45b24770-7e6a-4a72-8450-1fb8a1fe0098\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Образцы</th>\n",
              "      <th>chr11:11347700-11367700</th>\n",
              "      <th>chr11:40185800-40195800</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>8cell</td>\n",
              "      <td>551</td>\n",
              "      <td>194</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>epiblast</td>\n",
              "      <td>1344</td>\n",
              "      <td>565</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ICM</td>\n",
              "      <td>797</td>\n",
              "      <td>274</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-45b24770-7e6a-4a72-8450-1fb8a1fe0098')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-45b24770-7e6a-4a72-8450-1fb8a1fe0098 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-45b24770-7e6a-4a72-8450-1fb8a1fe0098');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-45599c5e-4f88-479d-a2bf-b3896be70b2d\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-45599c5e-4f88-479d-a2bf-b3896be70b2d')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-45599c5e-4f88-479d-a2bf-b3896be70b2d button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df",
              "summary": "{\n  \"name\": \"df\",\n  \"rows\": 3,\n  \"fields\": [\n    {\n      \"column\": \"\\u041e\\u0431\\u0440\\u0430\\u0437\\u0446\\u044b\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 3,\n        \"samples\": [\n          \"8cell\",\n          \"epiblast\",\n          \"ICM\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"chr11:11347700-11367700\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 405,\n        \"min\": 551,\n        \"max\": 1344,\n        \"num_unique_values\": 3,\n        \"samples\": [\n          551,\n          1344,\n          797\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"chr11:40185800-40195800\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 195,\n        \"min\": 194,\n        \"max\": 565,\n        \"num_unique_values\": 3,\n        \"samples\": [\n          194,\n          565,\n          274\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 56
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Дедупликация"
      ],
      "metadata": {
        "id": "iYFEJamUi-3w"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "! deduplicate_bismark  --bam  --paired  -o s_8_cell  SRR5836473_1_bismark_bt2_pe.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2vxKhuSxVUbd",
        "outputId": "9096858b-c796-46ba-9986-d20afa5513de"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Output filename was given as: s_8_cell\n",
            "Processing paired-end Bismark output file(s) (SAM format):\n",
            "SRR5836473_1_bismark_bt2_pe.bam\n",
            "\n",
            "\n",
            "If there are several alignments to a single position in the genome the first alignment will be chosen. Since the input files are not in any way sorted this is a near-enough random selection of reads.\n",
            "\n",
            "Checking file >>SRR5836473_1_bismark_bt2_pe.bam<< for signs of file truncation...\n",
            "\n",
            "\n",
            "\n",
            "Now testing Bismark result file SRR5836473_1_bismark_bt2_pe.bam for positional sorting (which would be bad...)\t...passed!\n",
            "Output file is: s_8_cell.deduplicated.bam\n",
            "\n",
            "skipping header line:\t@HD\tVN:1.0\tSO:unsorted\n",
            "skipping header line:\t@SQ\tSN:chr11\tLN:121973369\n",
            "skipping header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome_folder /mnt/cache-1/for_minor/mm39_11/ -q --non_directional -parallel 10 -1 SRR3824222_1.fq,SRR5836473_1.fq,SRR5836473_1.fq -2 SRR3824222_2.fq,SRR5836473_2.fq,SRR5836473_2.fq\"\n",
            "skipping header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.21\tCL:/usr/local/bin/samtools view -h --threads 1 SRR5836473_1_bismark_bt2_pe.bam\n",
            "\n",
            "Total number of alignments analysed in SRR5836473_1_bismark_bt2_pe.bam:\t2858170\n",
            "Total number duplicated alignments removed:\t522512 (18.28%)\n",
            "Duplicated alignments were found at:\t230677 different position(s)\n",
            "\n",
            "Total count of deduplicated leftover sequences: 2335658 (81.72% of total)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! deduplicate_bismark  --bam  --paired  -o s_epiblast  SRR3824222_1_bismark_bt2_pe.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ol_-zAj_Xrd3",
        "outputId": "dac72068-f61c-4278-d7a3-693cacf6edfe"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Output filename was given as: s_epiblast\n",
            "Processing paired-end Bismark output file(s) (SAM format):\n",
            "SRR3824222_1_bismark_bt2_pe.bam\n",
            "\n",
            "\n",
            "If there are several alignments to a single position in the genome the first alignment will be chosen. Since the input files are not in any way sorted this is a near-enough random selection of reads.\n",
            "\n",
            "Checking file >>SRR3824222_1_bismark_bt2_pe.bam<< for signs of file truncation...\n",
            "\n",
            "\n",
            "\n",
            "Now testing Bismark result file SRR3824222_1_bismark_bt2_pe.bam for positional sorting (which would be bad...)\t...passed!\n",
            "Output file is: s_epiblast.deduplicated.bam\n",
            "\n",
            "skipping header line:\t@HD\tVN:1.0\tSO:unsorted\n",
            "skipping header line:\t@SQ\tSN:chr11\tLN:121973369\n",
            "skipping header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome_folder /mnt/cache-1/for_minor/mm39_11/ -q --non_directional -parallel 10 -1 SRR3824222_1.fq,SRR5836473_1.fq,SRR5836473_1.fq -2 SRR3824222_2.fq,SRR5836473_2.fq,SRR5836473_2.fq\"\n",
            "skipping header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.21\tCL:/usr/local/bin/samtools view -h --threads 1 SRR3824222_1_bismark_bt2_pe.bam\n",
            "\n",
            "Total number of alignments analysed in SRR3824222_1_bismark_bt2_pe.bam:\t7051731\n",
            "Total number duplicated alignments removed:\t205287 (2.91%)\n",
            "Duplicated alignments were found at:\t164998 different position(s)\n",
            "\n",
            "Total count of deduplicated leftover sequences: 6846444 (97.09% of total)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! deduplicate_bismark  --bam  --paired  -o s_icm  SRR5836475_1_bismark_bt2_pe.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aqaC6Rh-Z19I",
        "outputId": "ecc04fc5-8d63-4388-e68a-3ce1c6ee8dda"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Output filename was given as: s_icm\n",
            "Processing paired-end Bismark output file(s) (SAM format):\n",
            "SRR5836475_1_bismark_bt2_pe.bam\n",
            "\n",
            "\n",
            "If there are several alignments to a single position in the genome the first alignment will be chosen. Since the input files are not in any way sorted this is a near-enough random selection of reads.\n",
            "\n",
            "Checking file >>SRR5836475_1_bismark_bt2_pe.bam<< for signs of file truncation...\n",
            "\n",
            "\n",
            "\n",
            "Now testing Bismark result file SRR5836475_1_bismark_bt2_pe.bam for positional sorting (which would be bad...)\t...passed!\n",
            "Output file is: s_icm.deduplicated.bam\n",
            "\n",
            "skipping header line:\t@HD\tVN:1.0\tSO:unsorted\n",
            "skipping header line:\t@SQ\tSN:chr11\tLN:121973369\n",
            "skipping header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome_folder /mnt/cache-1/for_minor/mm39_11/ -q --non_directional -parallel 10 -1 SRR5836475_1.fq -2 SRR5836475_2.fq\"\n",
            "skipping header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.21\tCL:/usr/local/bin/samtools view -h --threads 1 SRR5836475_1_bismark_bt2_pe.bam\n",
            "\n",
            "Total number of alignments analysed in SRR5836475_1_bismark_bt2_pe.bam:\t4170098\n",
            "Total number duplicated alignments removed:\t378125 (9.07%)\n",
            "Duplicated alignments were found at:\t229940 different position(s)\n",
            "\n",
            "Total count of deduplicated leftover sequences: 3791973 (90.93% of total)\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! ls"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "cM9EW8IPckXF",
        "outputId": "374a7362-2aa6-42d3-d94f-3966357c41c5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "condacolab_install.log\t\t     sorted_8cell.bam.bai\n",
            "drive\t\t\t\t     sorted_epiblast.bam\n",
            "FastQC\t\t\t\t     sorted_epiblast.bam.bai\n",
            "fastqc_out\t\t\t     sorted_icm.bam\n",
            "fastqc_v0.11.9.zip\t\t     sorted_icm.bam.bai\n",
            "s_8_cell.deduplicated.bam\t     SRR3824222_1_bismark_bt2_pe.bam\n",
            "s_8_cell.deduplication_report.txt    SRR3824222_1_bismark_bt2_PE_report.txt\n",
            "sample_data\t\t\t     SRR5836473_1_bismark_bt2_pe.bam\n",
            "s_epiblast.deduplicated.bam\t     SRR5836473_1_bismark_bt2_PE_report.txt\n",
            "s_epiblast.deduplication_report.txt  SRR5836473_1.fastq\n",
            "s_icm.deduplicated.bam\t\t     SRR5836473_1.fastq.gz\n",
            "s_icm.deduplication_report.txt\t     SRR5836475_1_bismark_bt2_pe.bam\n",
            "sorted_8cell.bam\t\t     SRR5836475_1_bismark_bt2_PE_report.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "a = pd.DataFrame({\n",
        "        'Образцы WGBS': ['8cell', 'epiblast', 'ICM'],\n",
        "        'Сколько процентов прочтений дуплицированно в каждом из образцов': [18.28, 2.91, 9.07],\n",
        "     })\n",
        "a"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 159
        },
        "id": "lxouJOsVcj0s",
        "outputId": "2c9d9c02-c822-4476-fc67-d7dedfa8344f"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  Образцы WGBS  \\\n",
              "0        8cell   \n",
              "1     epiblast   \n",
              "2          ICM   \n",
              "\n",
              "   Сколько процентов прочтений дуплицированно в каждом из образцов  \n",
              "0                                              18.28                \n",
              "1                                               2.91                \n",
              "2                                               9.07                "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3c0a67d1-39b2-4ce0-b415-d6e134c260f5\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Образцы WGBS</th>\n",
              "      <th>Сколько процентов прочтений дуплицированно в каждом из образцов</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>8cell</td>\n",
              "      <td>18.28</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>epiblast</td>\n",
              "      <td>2.91</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ICM</td>\n",
              "      <td>9.07</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3c0a67d1-39b2-4ce0-b415-d6e134c260f5')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-3c0a67d1-39b2-4ce0-b415-d6e134c260f5 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-3c0a67d1-39b2-4ce0-b415-d6e134c260f5');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-d155f277-e01d-42ad-b762-8b8ed372706b\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-d155f277-e01d-42ad-b762-8b8ed372706b')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-d155f277-e01d-42ad-b762-8b8ed372706b button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "a",
              "summary": "{\n  \"name\": \"a\",\n  \"rows\": 3,\n  \"fields\": [\n    {\n      \"column\": \"\\u041e\\u0431\\u0440\\u0430\\u0437\\u0446\\u044b WGBS\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 3,\n        \"samples\": [\n          \"8cell\",\n          \"epiblast\",\n          \"ICM\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"\\u0421\\u043a\\u043e\\u043b\\u044c\\u043a\\u043e \\u043f\\u0440\\u043e\\u0446\\u0435\\u043d\\u0442\\u043e\\u0432 \\u043f\\u0440\\u043e\\u0447\\u0442\\u0435\\u043d\\u0438\\u0439 \\u0434\\u0443\\u043f\\u043b\\u0438\\u0446\\u0438\\u0440\\u043e\\u0432\\u0430\\u043d\\u043d\\u043e \\u0432 \\u043a\\u0430\\u0436\\u0434\\u043e\\u043c \\u0438\\u0437 \\u043e\\u0431\\u0440\\u0430\\u0437\\u0446\\u043e\\u0432\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 7.7352720271063085,\n        \"min\": 2.91,\n        \"max\": 18.28,\n        \"num_unique_values\": 3,\n        \"samples\": [\n          18.28,\n          2.91,\n          9.07\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 19
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!mkdir /content/mouse_chr"
      ],
      "metadata": {
        "id": "JzcbOKBjeEwh"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import requests\n",
        "from google.colab import files\n",
        "\n",
        "url = \"https://hgdownload.soe.ucsc.edu/goldenPath/mm39/chromosomes/chr11.fa.gz\"\n",
        "response = requests.get(url)\n",
        "\n",
        "with open(\"chr11.fa.gz\", \"wb\") as file:\n",
        "    file.write(response.content)\n",
        "\n",
        "files.download(\"chr11.fa.gz\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "n6SuB-oDjkSc",
        "outputId": "c3a9ea37-1e46-4137-d093-5e79c84993e3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_51dc71a1-f3b4-4df3-bce3-1d5e6d0ff3c9\", \"chr11.fa.gz\", 39009907)"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gzip -dk chr11.fa.gz"
      ],
      "metadata": {
        "id": "dTWdX8pNm066"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "! bismark_methylation_extractor --gzip --bedGraph --cytosine_report --genome_folder ./mouse_chr s_8_cell.deduplicated.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "H2AzMjcVm5aW",
        "outputId": "3e7356ed-fef1-48a8-8dc5-faa912a5c6b4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            " *** Bismark methylation extractor version v0.24.2 ***\n",
            "\n",
            "Trying to determine the type of mapping from the SAM header line of file s_8_cell.deduplicated.bam\n",
            "Treating file(s) as paired-end data (as extracted from @PG line)\n",
            "\n",
            "Setting option '--no_overlap' since this is (normally) the right thing to do for paired-end data\n",
            "\n",
            "\n",
            "Setting core usage to single-threaded (default). Consider using --multicore <int> to speed up the extraction process.\n",
            "\n",
            "Summarising Bismark methylation extractor parameters:\n",
            "===============================================================\n",
            "Bismark paired-end SAM format specified (default)\n",
            "Number of cores to be used: 1\n",
            "Output will be written to the current directory ('/content')\n",
            "\n",
            "\n",
            "Summarising bedGraph parameters:\n",
            "===============================================================\n",
            "Generating additional output in bedGraph and coverage format\n",
            "bedGraph format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage>\n",
            "coverage format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage> <count methylated> <count non-methylated>\n",
            "\n",
            "Using a cutoff of 1 read(s) to report cytosine positions\n",
            "Reporting and sorting cytosine methylation information in CpG context only (default)\n",
            "The bedGraph UNIX sort command will use the following memory setting:\t'2G'. Temporary directory used for sorting is the output directory\n",
            "\n",
            "\n",
            "Summarising genome-wide cytosine methylation report parameters:\n",
            "===============================================================\n",
            "Generating comprehensive genome-wide cytosine report\n",
            "(output format: <Chromosome> <Position> <Strand> <count methylated> <count non-methylated>  <C-context>  <trinucleotide context> )\n",
            "Reporting cytosine methylation in CpG context only (default)\n",
            "Using 1-based genomic coordinates (default)\n",
            "Genome folder was specified as ./mouse_chr/\n",
            "\n",
            "Checking file >>s_8_cell.deduplicated.bam<< for signs of file truncation...\n",
            "\n",
            "\n",
            "Now testing Bismark result file >s_8_cell.deduplicated.bam< for positional sorting (which would be bad...)\t...passed!\n",
            "Writing result file containing methylation information for C in CpG context from the original top strand to CpG_OT_s_8_cell.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the complementary to original top strand to CpG_CTOT_s_8_cell.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the complementary to original bottom strand to CpG_CTOB_s_8_cell.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the original bottom strand to CpG_OB_s_8_cell.deduplicated.txt.gz\n",
            "\n",
            "Writing result file containing methylation information for C in CHG context from the original top strand to CHG_OT_s_8_cell.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the complementary to original top strand to CHG_CTOT_s_8_cell.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the complementary to original bottom strand to CHG_CTOB_s_8_cell.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the original bottom strand to CHG_OB_s_8_cell.deduplicated.txt.gz\n",
            "\n",
            "Writing result file containing methylation information for C in CHH context from the original top strand to CHH_OT_s_8_cell.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the complementary to original top strand to CHH_CTOT_s_8_cell.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the complementary to original bottom strand to CHH_CTOB_s_8_cell.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the original bottom strand to CHH_OB_s_8_cell.deduplicated.txt.gz\n",
            "\n",
            "Now reading in Bismark result file s_8_cell.deduplicated.bam\n",
            "skipping SAM header line:\t@HD\tVN:1.0\tSO:unsorted\n",
            "skipping SAM header line:\t@SQ\tSN:chr11\tLN:121973369\n",
            "skipping SAM header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome_folder /mnt/cache-1/for_minor/mm39_11/ -q --non_directional -parallel 10 -1 SRR3824222_1.fq,SRR5836473_1.fq,SRR5836473_1.fq -2 SRR3824222_2.fq,SRR5836473_2.fq,SRR5836473_2.fq\"\n",
            "skipping SAM header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.21\tCL:/usr/local/bin/samtools view -h --threads 1 SRR5836473_1_bismark_bt2_pe.bam\n",
            "skipping SAM header line:\t@PG\tID:samtools.1\tPN:samtools\tPP:samtools\tVN:1.21\tCL:/usr/local/bin/samtools view -bS --threads 1 -\n",
            "skipping SAM header line:\t@PG\tID:samtools.2\tPN:samtools\tPP:samtools.1\tVN:1.21\tCL:/usr/local/bin/samtools view -h s_8_cell.deduplicated.bam\n",
            "Processed lines: 500000\n",
            "Processed lines: 1000000\n",
            "Processed lines: 1500000\n",
            "Processed lines: 2000000\n",
            "\n",
            "Processed 2335658 lines in total\n",
            "Total number of methylation call strings processed: 4671316\n",
            "\n",
            "Final Cytosine Methylation Report\n",
            "=================================\n",
            "Total number of C's analysed:\t104748620\n",
            "\n",
            "Total methylated C's in CpG context:\t2103916\n",
            "Total methylated C's in CHG context:\t270195\n",
            "Total methylated C's in CHH context:\t1236285\n",
            "\n",
            "Total C to T conversions in CpG context:\t2729982\n",
            "Total C to T conversions in CHG context:\t22951565\n",
            "Total C to T conversions in CHH context:\t75456677\n",
            "\n",
            "C methylated in CpG context:\t43.5%\n",
            "C methylated in CHG context:\t1.2%\n",
            "C methylated in CHH context:\t1.6%\n",
            "\n",
            "\n",
            "Determining maximum read lengths for M-Bias plots\n",
            "Maximum read length of Read 1: 140\n",
            "Maximum read length of Read 2: 140\n",
            "\n",
            "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
            "Determining maximum read lengths for M-Bias plots\n",
            "Maximum read length of Read 1: 140\n",
            "Maximum read length of Read 2: 140\n",
            "\n",
            "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
            "Deleting unused files ...\n",
            "\n",
            "CpG_OT_s_8_cell.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_CTOT_s_8_cell.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_CTOB_s_8_cell.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_OB_s_8_cell.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_OT_s_8_cell.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_CTOT_s_8_cell.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_CTOB_s_8_cell.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_OB_s_8_cell.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_OT_s_8_cell.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_CTOT_s_8_cell.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_CTOB_s_8_cell.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_OB_s_8_cell.deduplicated.txt.gz contains data ->\tkept\n",
            "\n",
            "\n",
            "Using these input files: CpG_OT_s_8_cell.deduplicated.txt.gz CpG_CTOT_s_8_cell.deduplicated.txt.gz CpG_CTOB_s_8_cell.deduplicated.txt.gz CpG_OB_s_8_cell.deduplicated.txt.gz CHG_OT_s_8_cell.deduplicated.txt.gz CHG_CTOT_s_8_cell.deduplicated.txt.gz CHG_CTOB_s_8_cell.deduplicated.txt.gz CHG_OB_s_8_cell.deduplicated.txt.gz CHH_OT_s_8_cell.deduplicated.txt.gz CHH_CTOT_s_8_cell.deduplicated.txt.gz CHH_CTOB_s_8_cell.deduplicated.txt.gz CHH_OB_s_8_cell.deduplicated.txt.gz\n",
            "\n",
            "Summary of parameters for bismark2bedGraph conversion:\n",
            "======================================================\n",
            "bedGraph output:\t\ts_8_cell.deduplicated.bedGraph.gz\n",
            "output directory:\t\t><\n",
            "remove whitespaces:\t\tno\n",
            "CX context:\t\t\tno (CpG context only, default)\n",
            "No-header selected:\t\tno\n",
            "Sorting method:\t\t\tUnix sort-based (smaller memory footprint, but slower)\n",
            "Sort buffer size:\t\t2G\n",
            "Coverage threshold:\t\t1\n",
            "=============================================================================\n",
            "Methylation information will now be written into a bedGraph and coverage file\n",
            "=============================================================================\n",
            "\n",
            "Using the following files as Input:\n",
            "/content/CpG_OT_s_8_cell.deduplicated.txt.gz\t/content/CpG_CTOT_s_8_cell.deduplicated.txt.gz\t/content/CpG_CTOB_s_8_cell.deduplicated.txt.gz\t/content/CpG_OB_s_8_cell.deduplicated.txt.gz\n",
            "\n",
            "Writing bedGraph to file: s_8_cell.deduplicated.bedGraph.gz\n",
            "Also writing out a coverage file including counts methylated and unmethylated residues to file: s_8_cell.deduplicated.bismark.cov.gz\n",
            "\n",
            "Now writing methylation information for file >>CpG_OT_s_8_cell.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_OT_s_8_cell.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_CTOT_s_8_cell.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_CTOT_s_8_cell.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_CTOB_s_8_cell.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_CTOB_s_8_cell.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_OB_s_8_cell.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_OB_s_8_cell.deduplicated.txt.gz\n",
            "\n",
            "Collecting temporary chromosome file information... Processing the following input file(s):\n",
            "CpG_OT_s_8_cell.deduplicated.txt.gz.chrchr11.methXtractor.temp\n",
            "\n",
            "Sorting input file CpG_OT_s_8_cell.deduplicated.txt.gz.chrchr11.methXtractor.temp by positions (using -S of 2G)\n",
            "Successfully deleted the temporary input file CpG_OT_s_8_cell.deduplicated.txt.gz.chrchr11.methXtractor.temp\n",
            "\n",
            "Finished BedGraph conversion ...\n",
            "\n",
            "Summary of parameters for genome-wide cytosine report:\n",
            "==============================================================================\n",
            "Coverage infile:\t\t\t\ts_8_cell.deduplicated.bismark.cov.gz\n",
            "Output directory:\t\t\t\t><\n",
            "Parent directory:\t\t\t\t>/content/<\n",
            "Genome directory:\t\t\t\t>/content/mouse_chr/<\n",
            "Coverage threshold\t\t\t\t0 (all positions will be reported, default)\n",
            "CX context:\t\t\t\t\tno (CpG context only, default)\n",
            "Genome coordinates used:\t\t\t1-based (default)\n",
            "GZIP compression:\t\t\t\tyes\n",
            "Split by chromosome:\t\t\t\tno\n",
            "\n",
            "\n",
            "Now reading in and storing sequence information of the genome specified in: /content/mouse_chr/\n",
            "\n",
            "Couldn't find files ending in .fa, trying .fa.gz instead\n",
            "The specified genome folder /content/mouse_chr/ does not contain any sequence files in FastA format (with .fa, .fa.gz, .fasta or .fasta.gz file extensions)\n",
            "\n",
            "\n",
            "Finished generating genome-wide cytosine report\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! bismark_methylation_extractor --gzip --bedGraph --cytosine_report --genome_folder ./mouse_chr s_epiblast.deduplicated.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WvK5LTRYt5AD",
        "outputId": "27b1f387-4b39-4eb5-a64e-4f69de5d9f03"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            " *** Bismark methylation extractor version v0.24.2 ***\n",
            "\n",
            "Trying to determine the type of mapping from the SAM header line of file s_epiblast.deduplicated.bam\n",
            "Treating file(s) as paired-end data (as extracted from @PG line)\n",
            "\n",
            "Setting option '--no_overlap' since this is (normally) the right thing to do for paired-end data\n",
            "\n",
            "\n",
            "Setting core usage to single-threaded (default). Consider using --multicore <int> to speed up the extraction process.\n",
            "\n",
            "Summarising Bismark methylation extractor parameters:\n",
            "===============================================================\n",
            "Bismark paired-end SAM format specified (default)\n",
            "Number of cores to be used: 1\n",
            "Output will be written to the current directory ('/content')\n",
            "\n",
            "\n",
            "Summarising bedGraph parameters:\n",
            "===============================================================\n",
            "Generating additional output in bedGraph and coverage format\n",
            "bedGraph format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage>\n",
            "coverage format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage> <count methylated> <count non-methylated>\n",
            "\n",
            "Using a cutoff of 1 read(s) to report cytosine positions\n",
            "Reporting and sorting cytosine methylation information in CpG context only (default)\n",
            "The bedGraph UNIX sort command will use the following memory setting:\t'2G'. Temporary directory used for sorting is the output directory\n",
            "\n",
            "\n",
            "Summarising genome-wide cytosine methylation report parameters:\n",
            "===============================================================\n",
            "Generating comprehensive genome-wide cytosine report\n",
            "(output format: <Chromosome> <Position> <Strand> <count methylated> <count non-methylated>  <C-context>  <trinucleotide context> )\n",
            "Reporting cytosine methylation in CpG context only (default)\n",
            "Using 1-based genomic coordinates (default)\n",
            "Genome folder was specified as ./mouse_chr/\n",
            "\n",
            "Checking file >>s_epiblast.deduplicated.bam<< for signs of file truncation...\n",
            "\n",
            "\n",
            "Now testing Bismark result file >s_epiblast.deduplicated.bam< for positional sorting (which would be bad...)\t...passed!\n",
            "Writing result file containing methylation information for C in CpG context from the original top strand to CpG_OT_s_epiblast.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the complementary to original top strand to CpG_CTOT_s_epiblast.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the complementary to original bottom strand to CpG_CTOB_s_epiblast.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the original bottom strand to CpG_OB_s_epiblast.deduplicated.txt.gz\n",
            "\n",
            "Writing result file containing methylation information for C in CHG context from the original top strand to CHG_OT_s_epiblast.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the complementary to original top strand to CHG_CTOT_s_epiblast.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the complementary to original bottom strand to CHG_CTOB_s_epiblast.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the original bottom strand to CHG_OB_s_epiblast.deduplicated.txt.gz\n",
            "\n",
            "Writing result file containing methylation information for C in CHH context from the original top strand to CHH_OT_s_epiblast.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the complementary to original top strand to CHH_CTOT_s_epiblast.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the complementary to original bottom strand to CHH_CTOB_s_epiblast.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the original bottom strand to CHH_OB_s_epiblast.deduplicated.txt.gz\n",
            "\n",
            "Now reading in Bismark result file s_epiblast.deduplicated.bam\n",
            "skipping SAM header line:\t@HD\tVN:1.0\tSO:unsorted\n",
            "skipping SAM header line:\t@SQ\tSN:chr11\tLN:121973369\n",
            "skipping SAM header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome_folder /mnt/cache-1/for_minor/mm39_11/ -q --non_directional -parallel 10 -1 SRR3824222_1.fq,SRR5836473_1.fq,SRR5836473_1.fq -2 SRR3824222_2.fq,SRR5836473_2.fq,SRR5836473_2.fq\"\n",
            "skipping SAM header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.21\tCL:/usr/local/bin/samtools view -h --threads 1 SRR3824222_1_bismark_bt2_pe.bam\n",
            "skipping SAM header line:\t@PG\tID:samtools.1\tPN:samtools\tPP:samtools\tVN:1.21\tCL:/usr/local/bin/samtools view -bS --threads 1 -\n",
            "skipping SAM header line:\t@PG\tID:samtools.2\tPN:samtools\tPP:samtools.1\tVN:1.21\tCL:/usr/local/bin/samtools view -h s_epiblast.deduplicated.bam\n",
            "Processed lines: 500000\n",
            "Processed lines: 1000000\n",
            "Processed lines: 1500000\n",
            "Processed lines: 2000000\n",
            "Processed lines: 2500000\n",
            "Processed lines: 3000000\n",
            "Processed lines: 3500000\n",
            "Processed lines: 4000000\n",
            "Processed lines: 4500000\n",
            "Processed lines: 5000000\n",
            "Processed lines: 5500000\n",
            "Processed lines: 6000000\n",
            "Processed lines: 6500000\n",
            "\n",
            "Processed 6846444 lines in total\n",
            "Total number of methylation call strings processed: 13692888\n",
            "\n",
            "Final Cytosine Methylation Report\n",
            "=================================\n",
            "Total number of C's analysed:\t331429490\n",
            "\n",
            "Total methylated C's in CpG context:\t11550199\n",
            "Total methylated C's in CHG context:\t1574513\n",
            "Total methylated C's in CHH context:\t4133143\n",
            "\n",
            "Total C to T conversions in CpG context:\t3463972\n",
            "Total C to T conversions in CHG context:\t71924830\n",
            "Total C to T conversions in CHH context:\t238782833\n",
            "\n",
            "C methylated in CpG context:\t76.9%\n",
            "C methylated in CHG context:\t2.1%\n",
            "C methylated in CHH context:\t1.7%\n",
            "\n",
            "\n",
            "Determining maximum read lengths for M-Bias plots\n",
            "Maximum read length of Read 1: 140\n",
            "Maximum read length of Read 2: 140\n",
            "\n",
            "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
            "Determining maximum read lengths for M-Bias plots\n",
            "Maximum read length of Read 1: 140\n",
            "Maximum read length of Read 2: 140\n",
            "\n",
            "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
            "Deleting unused files ...\n",
            "\n",
            "CpG_OT_s_epiblast.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_CTOT_s_epiblast.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_CTOB_s_epiblast.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_OB_s_epiblast.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_OT_s_epiblast.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_CTOT_s_epiblast.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_CTOB_s_epiblast.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_OB_s_epiblast.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_OT_s_epiblast.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_CTOT_s_epiblast.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_CTOB_s_epiblast.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_OB_s_epiblast.deduplicated.txt.gz contains data ->\tkept\n",
            "\n",
            "\n",
            "Using these input files: CpG_OT_s_epiblast.deduplicated.txt.gz CpG_CTOT_s_epiblast.deduplicated.txt.gz CpG_CTOB_s_epiblast.deduplicated.txt.gz CpG_OB_s_epiblast.deduplicated.txt.gz CHG_OT_s_epiblast.deduplicated.txt.gz CHG_CTOT_s_epiblast.deduplicated.txt.gz CHG_CTOB_s_epiblast.deduplicated.txt.gz CHG_OB_s_epiblast.deduplicated.txt.gz CHH_OT_s_epiblast.deduplicated.txt.gz CHH_CTOT_s_epiblast.deduplicated.txt.gz CHH_CTOB_s_epiblast.deduplicated.txt.gz CHH_OB_s_epiblast.deduplicated.txt.gz\n",
            "\n",
            "Summary of parameters for bismark2bedGraph conversion:\n",
            "======================================================\n",
            "bedGraph output:\t\ts_epiblast.deduplicated.bedGraph.gz\n",
            "output directory:\t\t><\n",
            "remove whitespaces:\t\tno\n",
            "CX context:\t\t\tno (CpG context only, default)\n",
            "No-header selected:\t\tno\n",
            "Sorting method:\t\t\tUnix sort-based (smaller memory footprint, but slower)\n",
            "Sort buffer size:\t\t2G\n",
            "Coverage threshold:\t\t1\n",
            "=============================================================================\n",
            "Methylation information will now be written into a bedGraph and coverage file\n",
            "=============================================================================\n",
            "\n",
            "Using the following files as Input:\n",
            "/content/CpG_OT_s_epiblast.deduplicated.txt.gz\t/content/CpG_CTOT_s_epiblast.deduplicated.txt.gz\t/content/CpG_CTOB_s_epiblast.deduplicated.txt.gz\t/content/CpG_OB_s_epiblast.deduplicated.txt.gz\n",
            "\n",
            "Writing bedGraph to file: s_epiblast.deduplicated.bedGraph.gz\n",
            "Also writing out a coverage file including counts methylated and unmethylated residues to file: s_epiblast.deduplicated.bismark.cov.gz\n",
            "\n",
            "Now writing methylation information for file >>CpG_OT_s_epiblast.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_OT_s_epiblast.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_CTOT_s_epiblast.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_CTOT_s_epiblast.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_CTOB_s_epiblast.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_CTOB_s_epiblast.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_OB_s_epiblast.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_OB_s_epiblast.deduplicated.txt.gz\n",
            "\n",
            "Collecting temporary chromosome file information... Processing the following input file(s):\n",
            "CpG_OT_s_epiblast.deduplicated.txt.gz.chrchr11.methXtractor.temp\n",
            "\n",
            "Sorting input file CpG_OT_s_epiblast.deduplicated.txt.gz.chrchr11.methXtractor.temp by positions (using -S of 2G)\n",
            "Successfully deleted the temporary input file CpG_OT_s_epiblast.deduplicated.txt.gz.chrchr11.methXtractor.temp\n",
            "\n",
            "Finished BedGraph conversion ...\n",
            "\n",
            "Summary of parameters for genome-wide cytosine report:\n",
            "==============================================================================\n",
            "Coverage infile:\t\t\t\ts_epiblast.deduplicated.bismark.cov.gz\n",
            "Output directory:\t\t\t\t><\n",
            "Parent directory:\t\t\t\t>/content/<\n",
            "Genome directory:\t\t\t\t>/content/mouse_chr/<\n",
            "Coverage threshold\t\t\t\t0 (all positions will be reported, default)\n",
            "CX context:\t\t\t\t\tno (CpG context only, default)\n",
            "Genome coordinates used:\t\t\t1-based (default)\n",
            "GZIP compression:\t\t\t\tyes\n",
            "Split by chromosome:\t\t\t\tno\n",
            "\n",
            "\n",
            "Now reading in and storing sequence information of the genome specified in: /content/mouse_chr/\n",
            "\n",
            "Couldn't find files ending in .fa, trying .fa.gz instead\n",
            "The specified genome folder /content/mouse_chr/ does not contain any sequence files in FastA format (with .fa, .fa.gz, .fasta or .fasta.gz file extensions)\n",
            "\n",
            "\n",
            "Finished generating genome-wide cytosine report\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! bismark_methylation_extractor --gzip --bedGraph --cytosine_report --genome_folder ./mouse_chr s_icm.deduplicated.bam"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Mlkjx5Qo5gXH",
        "outputId": "967ceca0-f3e0-4737-f65a-dfb7a22e3092"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            " *** Bismark methylation extractor version v0.24.2 ***\n",
            "\n",
            "Trying to determine the type of mapping from the SAM header line of file s_icm.deduplicated.bam\n",
            "Treating file(s) as paired-end data (as extracted from @PG line)\n",
            "\n",
            "Setting option '--no_overlap' since this is (normally) the right thing to do for paired-end data\n",
            "\n",
            "\n",
            "Setting core usage to single-threaded (default). Consider using --multicore <int> to speed up the extraction process.\n",
            "\n",
            "Summarising Bismark methylation extractor parameters:\n",
            "===============================================================\n",
            "Bismark paired-end SAM format specified (default)\n",
            "Number of cores to be used: 1\n",
            "Output will be written to the current directory ('/content')\n",
            "\n",
            "\n",
            "Summarising bedGraph parameters:\n",
            "===============================================================\n",
            "Generating additional output in bedGraph and coverage format\n",
            "bedGraph format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage>\n",
            "coverage format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage> <count methylated> <count non-methylated>\n",
            "\n",
            "Using a cutoff of 1 read(s) to report cytosine positions\n",
            "Reporting and sorting cytosine methylation information in CpG context only (default)\n",
            "The bedGraph UNIX sort command will use the following memory setting:\t'2G'. Temporary directory used for sorting is the output directory\n",
            "\n",
            "\n",
            "Summarising genome-wide cytosine methylation report parameters:\n",
            "===============================================================\n",
            "Generating comprehensive genome-wide cytosine report\n",
            "(output format: <Chromosome> <Position> <Strand> <count methylated> <count non-methylated>  <C-context>  <trinucleotide context> )\n",
            "Reporting cytosine methylation in CpG context only (default)\n",
            "Using 1-based genomic coordinates (default)\n",
            "Genome folder was specified as ./mouse_chr/\n",
            "\n",
            "Checking file >>s_icm.deduplicated.bam<< for signs of file truncation...\n",
            "\n",
            "\n",
            "Now testing Bismark result file >s_icm.deduplicated.bam< for positional sorting (which would be bad...)\t...passed!\n",
            "Writing result file containing methylation information for C in CpG context from the original top strand to CpG_OT_s_icm.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the complementary to original top strand to CpG_CTOT_s_icm.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the complementary to original bottom strand to CpG_CTOB_s_icm.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CpG context from the original bottom strand to CpG_OB_s_icm.deduplicated.txt.gz\n",
            "\n",
            "Writing result file containing methylation information for C in CHG context from the original top strand to CHG_OT_s_icm.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the complementary to original top strand to CHG_CTOT_s_icm.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the complementary to original bottom strand to CHG_CTOB_s_icm.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHG context from the original bottom strand to CHG_OB_s_icm.deduplicated.txt.gz\n",
            "\n",
            "Writing result file containing methylation information for C in CHH context from the original top strand to CHH_OT_s_icm.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the complementary to original top strand to CHH_CTOT_s_icm.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the complementary to original bottom strand to CHH_CTOB_s_icm.deduplicated.txt.gz\n",
            "Writing result file containing methylation information for C in CHH context from the original bottom strand to CHH_OB_s_icm.deduplicated.txt.gz\n",
            "\n",
            "Now reading in Bismark result file s_icm.deduplicated.bam\n",
            "skipping SAM header line:\t@HD\tVN:1.0\tSO:unsorted\n",
            "skipping SAM header line:\t@SQ\tSN:chr11\tLN:121973369\n",
            "skipping SAM header line:\t@PG\tID:Bismark\tVN:v0.24.2\tCL:\"bismark --genome_folder /mnt/cache-1/for_minor/mm39_11/ -q --non_directional -parallel 10 -1 SRR5836475_1.fq -2 SRR5836475_2.fq\"\n",
            "skipping SAM header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.21\tCL:/usr/local/bin/samtools view -h --threads 1 SRR5836475_1_bismark_bt2_pe.bam\n",
            "skipping SAM header line:\t@PG\tID:samtools.1\tPN:samtools\tPP:samtools\tVN:1.21\tCL:/usr/local/bin/samtools view -bS --threads 1 -\n",
            "skipping SAM header line:\t@PG\tID:samtools.2\tPN:samtools\tPP:samtools.1\tVN:1.21\tCL:/usr/local/bin/samtools view -h s_icm.deduplicated.bam\n",
            "Processed lines: 500000\n",
            "Processed lines: 1000000\n",
            "Processed lines: 1500000\n",
            "Processed lines: 2000000\n",
            "Processed lines: 2500000\n",
            "Processed lines: 3000000\n",
            "Processed lines: 3500000\n",
            "\n",
            "Processed 3791973 lines in total\n",
            "Total number of methylation call strings processed: 7583946\n",
            "\n",
            "Final Cytosine Methylation Report\n",
            "=================================\n",
            "Total number of C's analysed:\t177777254\n",
            "\n",
            "Total methylated C's in CpG context:\t1915876\n",
            "Total methylated C's in CHG context:\t298499\n",
            "Total methylated C's in CHH context:\t1434775\n",
            "\n",
            "Total C to T conversions in CpG context:\t6256205\n",
            "Total C to T conversions in CHG context:\t39157632\n",
            "Total C to T conversions in CHH context:\t128714267\n",
            "\n",
            "C methylated in CpG context:\t23.4%\n",
            "C methylated in CHG context:\t0.8%\n",
            "C methylated in CHH context:\t1.1%\n",
            "\n",
            "\n",
            "Determining maximum read lengths for M-Bias plots\n",
            "Maximum read length of Read 1: 140\n",
            "Maximum read length of Read 2: 140\n",
            "\n",
            "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
            "Determining maximum read lengths for M-Bias plots\n",
            "Maximum read length of Read 1: 140\n",
            "Maximum read length of Read 2: 140\n",
            "\n",
            "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
            "Deleting unused files ...\n",
            "\n",
            "CpG_OT_s_icm.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_CTOT_s_icm.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_CTOB_s_icm.deduplicated.txt.gz contains data ->\tkept\n",
            "CpG_OB_s_icm.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_OT_s_icm.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_CTOT_s_icm.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_CTOB_s_icm.deduplicated.txt.gz contains data ->\tkept\n",
            "CHG_OB_s_icm.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_OT_s_icm.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_CTOT_s_icm.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_CTOB_s_icm.deduplicated.txt.gz contains data ->\tkept\n",
            "CHH_OB_s_icm.deduplicated.txt.gz contains data ->\tkept\n",
            "\n",
            "\n",
            "Using these input files: CpG_OT_s_icm.deduplicated.txt.gz CpG_CTOT_s_icm.deduplicated.txt.gz CpG_CTOB_s_icm.deduplicated.txt.gz CpG_OB_s_icm.deduplicated.txt.gz CHG_OT_s_icm.deduplicated.txt.gz CHG_CTOT_s_icm.deduplicated.txt.gz CHG_CTOB_s_icm.deduplicated.txt.gz CHG_OB_s_icm.deduplicated.txt.gz CHH_OT_s_icm.deduplicated.txt.gz CHH_CTOT_s_icm.deduplicated.txt.gz CHH_CTOB_s_icm.deduplicated.txt.gz CHH_OB_s_icm.deduplicated.txt.gz\n",
            "\n",
            "Summary of parameters for bismark2bedGraph conversion:\n",
            "======================================================\n",
            "bedGraph output:\t\ts_icm.deduplicated.bedGraph.gz\n",
            "output directory:\t\t><\n",
            "remove whitespaces:\t\tno\n",
            "CX context:\t\t\tno (CpG context only, default)\n",
            "No-header selected:\t\tno\n",
            "Sorting method:\t\t\tUnix sort-based (smaller memory footprint, but slower)\n",
            "Sort buffer size:\t\t2G\n",
            "Coverage threshold:\t\t1\n",
            "=============================================================================\n",
            "Methylation information will now be written into a bedGraph and coverage file\n",
            "=============================================================================\n",
            "\n",
            "Using the following files as Input:\n",
            "/content/CpG_OT_s_icm.deduplicated.txt.gz\t/content/CpG_CTOT_s_icm.deduplicated.txt.gz\t/content/CpG_CTOB_s_icm.deduplicated.txt.gz\t/content/CpG_OB_s_icm.deduplicated.txt.gz\n",
            "\n",
            "Writing bedGraph to file: s_icm.deduplicated.bedGraph.gz\n",
            "Also writing out a coverage file including counts methylated and unmethylated residues to file: s_icm.deduplicated.bismark.cov.gz\n",
            "\n",
            "Now writing methylation information for file >>CpG_OT_s_icm.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_OT_s_icm.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_CTOT_s_icm.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_CTOT_s_icm.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_CTOB_s_icm.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_CTOB_s_icm.deduplicated.txt.gz\n",
            "Now writing methylation information for file >>CpG_OB_s_icm.deduplicated.txt.gz<< to individual files for each chromosome\n",
            "Finished writing out individual chromosome files for CpG_OB_s_icm.deduplicated.txt.gz\n",
            "\n",
            "Collecting temporary chromosome file information... Processing the following input file(s):\n",
            "CpG_OT_s_icm.deduplicated.txt.gz.chrchr11.methXtractor.temp\n",
            "\n",
            "Sorting input file CpG_OT_s_icm.deduplicated.txt.gz.chrchr11.methXtractor.temp by positions (using -S of 2G)\n",
            "Successfully deleted the temporary input file CpG_OT_s_icm.deduplicated.txt.gz.chrchr11.methXtractor.temp\n",
            "\n",
            "Finished BedGraph conversion ...\n",
            "\n",
            "Summary of parameters for genome-wide cytosine report:\n",
            "==============================================================================\n",
            "Coverage infile:\t\t\t\ts_icm.deduplicated.bismark.cov.gz\n",
            "Output directory:\t\t\t\t><\n",
            "Parent directory:\t\t\t\t>/content/<\n",
            "Genome directory:\t\t\t\t>/content/mouse_chr/<\n",
            "Coverage threshold\t\t\t\t0 (all positions will be reported, default)\n",
            "CX context:\t\t\t\t\tno (CpG context only, default)\n",
            "Genome coordinates used:\t\t\t1-based (default)\n",
            "GZIP compression:\t\t\t\tyes\n",
            "Split by chromosome:\t\t\t\tno\n",
            "\n",
            "\n",
            "Now reading in and storing sequence information of the genome specified in: /content/mouse_chr/\n",
            "\n",
            "Couldn't find files ending in .fa, trying .fa.gz instead\n",
            "The specified genome folder /content/mouse_chr/ does not contain any sequence files in FastA format (with .fa, .fa.gz, .fasta or .fasta.gz file extensions)\n",
            "\n",
            "\n",
            "Finished generating genome-wide cytosine report\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! ls -lh"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tMSOy45EOJbg",
        "outputId": "71768529-7ef3-4383-f2bb-46a842e40889"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "total 50G\n",
            "-rw-r--r-- 1 root root  7.7K Feb  9 17:16 CHG_CTOB_s_8_cell.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   13K Feb  9 18:45 CHG_CTOB_s_epiblast.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  9.4K Feb  9 19:36 CHG_CTOB_s_icm.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  8.2K Feb  9 17:16 CHG_CTOT_s_8_cell.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   14K Feb  9 18:45 CHG_CTOT_s_epiblast.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   10K Feb  9 19:36 CHG_CTOT_s_icm.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   51M Feb  9 17:16 CHG_OB_s_8_cell.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  157M Feb  9 18:45 CHG_OB_s_epiblast.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   84M Feb  9 19:36 CHG_OB_s_icm.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   51M Feb  9 17:16 CHG_OT_s_8_cell.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  157M Feb  9 18:45 CHG_OT_s_epiblast.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   85M Feb  9 19:36 CHG_OT_s_icm.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  1.1M Feb  9 17:16 CHH_CTOB_s_8_cell.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  1.9M Feb  9 18:45 CHH_CTOB_s_epiblast.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  1.6M Feb  9 19:36 CHH_CTOB_s_icm.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root 1014K Feb  9 17:16 CHH_CTOT_s_8_cell.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  1.7M Feb  9 18:45 CHH_CTOT_s_epiblast.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  1.4M Feb  9 19:36 CHH_CTOT_s_icm.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  133M Feb  9 17:16 CHH_OB_s_8_cell.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  419M Feb  9 18:45 CHH_OB_s_epiblast.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  223M Feb  9 19:36 CHH_OB_s_icm.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  133M Feb  9 17:16 CHH_OT_s_8_cell.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  419M Feb  9 18:45 CHH_OT_s_epiblast.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  223M Feb  9 19:36 CHH_OT_s_icm.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  119M Feb  9 16:32 chr11.fa\n",
            "-rw-r--r-- 1 root root   38M Feb  9 16:32 chr11.fa.gz\n",
            "-rw-r--r-- 1 root root   21K Feb  9 12:32 condacolab_install.log\n",
            "-rw-r--r-- 1 root root  4.2K Feb  9 17:16 CpG_CTOB_s_8_cell.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  7.5K Feb  9 18:45 CpG_CTOB_s_epiblast.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  4.7K Feb  9 19:36 CpG_CTOB_s_icm.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  3.8K Feb  9 17:16 CpG_CTOT_s_8_cell.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  8.0K Feb  9 18:45 CpG_CTOT_s_epiblast.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root  4.6K Feb  9 19:36 CpG_CTOT_s_icm.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   17M Feb  9 17:16 CpG_OB_s_8_cell.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   49M Feb  9 18:45 CpG_OB_s_epiblast.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   28M Feb  9 19:36 CpG_OB_s_icm.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   17M Feb  9 17:16 CpG_OT_s_8_cell.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   49M Feb  9 18:45 CpG_OT_s_epiblast.deduplicated.txt.gz\n",
            "-rw-r--r-- 1 root root   28M Feb  9 19:36 CpG_OT_s_icm.deduplicated.txt.gz\n",
            "drwx------ 6 root root  4.0K Feb  9 15:06 drive\n",
            "drwxr-xr-x 8 root root  4.0K Feb  9 12:06 FastQC\n",
            "drwxr-xr-x 2 root root  4.0K Feb  9 13:11 fastqc_out\n",
            "-rw-r--r-- 1 root root  9.8M Jan 15  2020 fastqc_v0.11.9.zip\n",
            "drwxr-xr-x 2 root root  4.0K Feb  9 16:32 mouse_chr\n",
            "-rw-r--r-- 1 root root  620M Feb  9 15:33 s_8_cell.deduplicated.bam\n",
            "-rw-r--r-- 1 root root  9.8M Feb  9 17:17 s_8_cell.deduplicated.bedGraph.gz\n",
            "-rw-r--r-- 1 root root  9.4M Feb  9 17:17 s_8_cell.deduplicated.bismark.cov.gz\n",
            "-rw-r--r-- 1 root root   22K Feb  9 17:16 s_8_cell.deduplicated.M-bias.txt\n",
            "-rw-r--r-- 1 root root   841 Feb  9 17:16 s_8_cell.deduplicated_splitting_report.txt\n",
            "-rw-r--r-- 1 root root   283 Feb  9 15:33 s_8_cell.deduplication_report.txt\n",
            "drwxr-xr-x 1 root root  4.0K Feb  6 14:19 sample_data\n",
            "-rw-r--r-- 1 root root  1.9G Feb  9 15:49 s_epiblast.deduplicated.bam\n",
            "-rw-r--r-- 1 root root   14M Feb  9 18:49 s_epiblast.deduplicated.bedGraph.gz\n",
            "-rw-r--r-- 1 root root   14M Feb  9 18:49 s_epiblast.deduplicated.bismark.cov.gz\n",
            "-rw-r--r-- 1 root root   24K Feb  9 18:45 s_epiblast.deduplicated.M-bias.txt\n",
            "-rw-r--r-- 1 root root   847 Feb  9 18:45 s_epiblast.deduplicated_splitting_report.txt\n",
            "-rw-r--r-- 1 root root   282 Feb  9 15:49 s_epiblast.deduplication_report.txt\n",
            "-rw-r--r-- 1 root root  1.1G Feb  9 15:55 s_icm.deduplicated.bam\n",
            "-rw-r--r-- 1 root root   12M Feb  9 19:38 s_icm.deduplicated.bedGraph.gz\n",
            "-rw-r--r-- 1 root root   12M Feb  9 19:38 s_icm.deduplicated.bismark.cov.gz\n",
            "-rw-r--r-- 1 root root   23K Feb  9 19:36 s_icm.deduplicated.M-bias.txt\n",
            "-rw-r--r-- 1 root root   839 Feb  9 19:36 s_icm.deduplicated_splitting_report.txt\n",
            "-rw-r--r-- 1 root root   282 Feb  9 15:55 s_icm.deduplication_report.txt\n",
            "-rw-r--r-- 1 root root  528M Feb  9 15:11 sorted_8cell.bam\n",
            "-rw-r--r-- 1 root root  203K Feb  9 15:24 sorted_8cell.bam.bai\n",
            "-rw-r--r-- 1 root root  1.3G Feb  9 15:18 sorted_epiblast.bam\n",
            "-rw-r--r-- 1 root root  349K Feb  9 15:25 sorted_epiblast.bam.bai\n",
            "-rw-r--r-- 1 root root  777M Feb  9 15:21 sorted_icm.bam\n",
            "-rw-r--r-- 1 root root  341K Feb  9 15:25 sorted_icm.bam.bai\n",
            "-rw------- 1 root root  1.9G Feb  9 15:07 SRR3824222_1_bismark_bt2_pe.bam\n",
            "-rw------- 1 root root  1.8K Feb  9 15:07 SRR3824222_1_bismark_bt2_PE_report.txt\n",
            "-rw------- 1 root root  769M Feb  9 15:06 SRR5836473_1_bismark_bt2_pe.bam\n",
            "-rw------- 1 root root  1.8K Feb  9 15:07 SRR5836473_1_bismark_bt2_PE_report.txt\n",
            "-rw-r--r-- 1 root root   32G Feb  9 12:26 SRR5836473_1.fastq\n",
            "-rw-r--r-- 1 root root  6.4G Feb  9 12:26 SRR5836473_1.fastq.gz\n",
            "-rw------- 1 root root  1.2G Feb  9 15:07 SRR5836475_1_bismark_bt2_pe.bam\n",
            "-rw------- 1 root root  1.8K Feb  9 15:07 SRR5836475_1_bismark_bt2_PE_report.txt\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! bismark2report -mbias_report s_8_cell.deduplicated.M-bias.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7XqUdXIaOQEV",
        "outputId": "76f09c1d-3710-4222-c15a-7b8e697fd3f8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Found 3 alignment reports in current directory. Now trying to figure out whether there are corresponding optional reports\n",
            "\n",
            "Writing Bismark HTML report to >> SRR3824222_1_bismark_bt2_PE_report.html <<\n",
            "\n",
            "==============================================================================================================\n",
            "Using the following alignment report:\t\t> SRR3824222_1_bismark_bt2_PE_report.txt <\n",
            "Processing alignment report SRR3824222_1_bismark_bt2_PE_report.txt ...\n",
            "Complete\n",
            "\n",
            "No deduplication report file specified, skipping this step\n",
            "No splitting report file specified, skipping this step\n",
            "Using the following M-bias report:\t\t> s_8_cell.deduplicated.M-bias.txt <\n",
            "Processing M-bias report s_8_cell.deduplicated.M-bias.txt ...\n",
            "Complete\n",
            "\n",
            "No nucleotide coverage report file specified, skipping this step\n",
            "==============================================================================================================\n",
            "\n",
            "\n",
            "\n",
            "Writing Bismark HTML report to >> SRR5836473_1_bismark_bt2_PE_report.html <<\n",
            "\n",
            "==============================================================================================================\n",
            "Using the following alignment report:\t\t> SRR5836473_1_bismark_bt2_PE_report.txt <\n",
            "Processing alignment report SRR5836473_1_bismark_bt2_PE_report.txt ...\n",
            "Complete\n",
            "\n",
            "No deduplication report file specified, skipping this step\n",
            "No splitting report file specified, skipping this step\n",
            "No M-bias report file specified, skipping this step\n",
            "No nucleotide coverage report file specified, skipping this step\n",
            "==============================================================================================================\n",
            "\n",
            "\n",
            "\n",
            "Writing Bismark HTML report to >> SRR5836475_1_bismark_bt2_PE_report.html <<\n",
            "\n",
            "==============================================================================================================\n",
            "Using the following alignment report:\t\t> SRR5836475_1_bismark_bt2_PE_report.txt <\n",
            "Processing alignment report SRR5836475_1_bismark_bt2_PE_report.txt ...\n",
            "Complete\n",
            "\n",
            "No deduplication report file specified, skipping this step\n",
            "No splitting report file specified, skipping this step\n",
            "No M-bias report file specified, skipping this step\n",
            "No nucleotide coverage report file specified, skipping this step\n",
            "==============================================================================================================\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Гистограммы"
      ],
      "metadata": {
        "id": "ODvda9urQM0P"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install pyGenomeTracks\n",
        "!pip install -U ipykernel"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "5d3PCwmvQN4m",
        "outputId": "849eb452-d704-4d69-d5be-fd90de07baf3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting pyGenomeTracks\n",
            "  Downloading pyGenomeTracks-3.9-py3-none-any.whl.metadata (51 kB)\n",
            "Collecting numpy>=1.20 (from pyGenomeTracks)\n",
            "  Downloading numpy-2.2.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (62 kB)\n",
            "Collecting matplotlib<3.9,>=3.1.1 (from pyGenomeTracks)\n",
            "  Downloading matplotlib-3.8.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.8 kB)\n",
            "Collecting intervaltree>=2.1.0 (from pyGenomeTracks)\n",
            "  Downloading intervaltree-3.1.0.tar.gz (32 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting pyBigWig>=0.3.16 (from pyGenomeTracks)\n",
            "  Downloading pyBigWig-0.3.24-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (19 kB)\n",
            "Collecting future>=0.17.0 (from pyGenomeTracks)\n",
            "  Downloading future-1.0.0-py3-none-any.whl.metadata (4.0 kB)\n",
            "Collecting hicmatrix>=15 (from pyGenomeTracks)\n",
            "  Downloading HiCMatrix-17.2-py3-none-any.whl.metadata (2.0 kB)\n",
            "Collecting pysam>=0.14 (from pyGenomeTracks)\n",
            "  Downloading pysam-0.23.0-cp311-cp311-manylinux_2_28_x86_64.whl.metadata (1.6 kB)\n",
            "Collecting pytest (from pyGenomeTracks)\n",
            "  Downloading pytest-8.3.4-py3-none-any.whl.metadata (7.5 kB)\n",
            "Collecting gffutils>=0.9 (from pyGenomeTracks)\n",
            "  Downloading gffutils-0.13-py3-none-any.whl.metadata (1.5 kB)\n",
            "Collecting pybedtools>=0.8.1 (from pyGenomeTracks)\n",
            "  Downloading pybedtools-0.11.0.tar.gz (12.5 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m12.5/12.5 MB\u001b[0m \u001b[31m80.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: tqdm>=4.20 in /usr/local/lib/python3.11/site-packages (from pyGenomeTracks) (4.67.1)\n",
            "Collecting bx-python>=0.8.13 (from pyGenomeTracks)\n",
            "  Downloading bx_python-0.13.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.1 kB)\n",
            "Collecting pyfaidx>=0.1.3 (from pyGenomeTracks)\n",
            "  Downloading pyfaidx-0.8.1.3-py3-none-any.whl.metadata (25 kB)\n",
            "Collecting pyparsing (from bx-python>=0.8.13->pyGenomeTracks)\n",
            "  Downloading pyparsing-3.2.1-py3-none-any.whl.metadata (5.0 kB)\n",
            "Collecting argh>=0.26.2 (from gffutils>=0.9->pyGenomeTracks)\n",
            "  Downloading argh-0.31.3-py3-none-any.whl.metadata (7.4 kB)\n",
            "Collecting argcomplete>=1.9.4 (from gffutils>=0.9->pyGenomeTracks)\n",
            "  Downloading argcomplete-3.5.3-py3-none-any.whl.metadata (16 kB)\n",
            "Collecting simplejson (from gffutils>=0.9->pyGenomeTracks)\n",
            "  Downloading simplejson-3.19.3-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.2 kB)\n",
            "Collecting scipy>=1.2 (from hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading scipy-1.15.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)\n",
            "Collecting tables>=3.5 (from hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading tables-3.10.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.0 kB)\n",
            "Collecting pandas>=0.25 (from hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading pandas-2.2.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (89 kB)\n",
            "Collecting cooler>=0.8.9 (from hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading cooler-0.10.3-py3-none-any.whl.metadata (9.5 kB)\n",
            "Collecting sortedcontainers<3.0,>=2.0 (from intervaltree>=2.1.0->pyGenomeTracks)\n",
            "  Downloading sortedcontainers-2.4.0-py2.py3-none-any.whl.metadata (10 kB)\n",
            "Collecting contourpy>=1.0.1 (from matplotlib<3.9,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading contourpy-1.3.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (5.4 kB)\n",
            "Collecting cycler>=0.10 (from matplotlib<3.9,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)\n",
            "Collecting fonttools>=4.22.0 (from matplotlib<3.9,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading fonttools-4.56.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (101 kB)\n",
            "Collecting kiwisolver>=1.3.1 (from matplotlib<3.9,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading kiwisolver-1.4.8-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.2 kB)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.11/site-packages (from matplotlib<3.9,>=3.1.1->pyGenomeTracks) (24.2)\n",
            "Collecting pillow>=8 (from matplotlib<3.9,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading pillow-11.1.0-cp311-cp311-manylinux_2_28_x86_64.whl.metadata (9.1 kB)\n",
            "Collecting python-dateutil>=2.7 (from matplotlib<3.9,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)\n",
            "Collecting importlib-metadata (from pyfaidx>=0.1.3->pyGenomeTracks)\n",
            "  Downloading importlib_metadata-8.6.1-py3-none-any.whl.metadata (4.7 kB)\n",
            "Collecting iniconfig (from pytest->pyGenomeTracks)\n",
            "  Downloading iniconfig-2.0.0-py3-none-any.whl.metadata (2.6 kB)\n",
            "Requirement already satisfied: pluggy<2,>=1.5 in /usr/local/lib/python3.11/site-packages (from pytest->pyGenomeTracks) (1.5.0)\n",
            "Collecting asciitree (from cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading asciitree-0.3.3.tar.gz (4.0 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting click>=7 (from cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading click-8.1.8-py3-none-any.whl.metadata (2.3 kB)\n",
            "Collecting cytoolz (from cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading cytoolz-1.0.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.6 kB)\n",
            "Collecting h5py>=2.5 (from cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading h5py-3.12.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.5 kB)\n",
            "Collecting multiprocess (from cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading multiprocess-0.70.17-py311-none-any.whl.metadata (7.2 kB)\n",
            "Collecting pyyaml (from cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading PyYAML-6.0.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.1 kB)\n",
            "Collecting pytz>=2020.1 (from pandas>=0.25->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading pytz-2025.1-py2.py3-none-any.whl.metadata (22 kB)\n",
            "Collecting tzdata>=2022.7 (from pandas>=0.25->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading tzdata-2025.1-py2.py3-none-any.whl.metadata (1.4 kB)\n",
            "Collecting six>=1.5 (from python-dateutil>=2.7->matplotlib<3.9,>=3.1.1->pyGenomeTracks)\n",
            "  Downloading six-1.17.0-py2.py3-none-any.whl.metadata (1.7 kB)\n",
            "Collecting numexpr>=2.6.2 (from tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading numexpr-2.10.2-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl.metadata (8.1 kB)\n",
            "Collecting py-cpuinfo (from tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading py_cpuinfo-9.0.0-py3-none-any.whl.metadata (794 bytes)\n",
            "Collecting blosc2>=2.3.0 (from tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading blosc2-3.0.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
            "Collecting typing-extensions>=4.4.0 (from tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading typing_extensions-4.12.2-py3-none-any.whl.metadata (3.0 kB)\n",
            "Collecting zipp>=3.20 (from importlib-metadata->pyfaidx>=0.1.3->pyGenomeTracks)\n",
            "  Downloading zipp-3.21.0-py3-none-any.whl.metadata (3.7 kB)\n",
            "Collecting ndindex (from blosc2>=2.3.0->tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading ndindex-1.9.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.4 kB)\n",
            "Collecting msgpack (from blosc2>=2.3.0->tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading msgpack-1.1.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (8.4 kB)\n",
            "Collecting httpx (from blosc2>=2.3.0->tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading httpx-0.28.1-py3-none-any.whl.metadata (7.1 kB)\n",
            "Collecting toolz>=0.8.0 (from cytoolz->cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading toolz-1.0.0-py3-none-any.whl.metadata (5.1 kB)\n",
            "Collecting dill>=0.3.9 (from multiprocess->cooler>=0.8.9->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading dill-0.3.9-py3-none-any.whl.metadata (10 kB)\n",
            "Collecting anyio (from httpx->blosc2>=2.3.0->tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading anyio-4.8.0-py3-none-any.whl.metadata (4.6 kB)\n",
            "Requirement already satisfied: certifi in /usr/local/lib/python3.11/site-packages (from httpx->blosc2>=2.3.0->tables>=3.5->hicmatrix>=15->pyGenomeTracks) (2024.12.14)\n",
            "Collecting httpcore==1.* (from httpx->blosc2>=2.3.0->tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading httpcore-1.0.7-py3-none-any.whl.metadata (21 kB)\n",
            "Requirement already satisfied: idna in /usr/local/lib/python3.11/site-packages (from httpx->blosc2>=2.3.0->tables>=3.5->hicmatrix>=15->pyGenomeTracks) (3.10)\n",
            "Collecting h11<0.15,>=0.13 (from httpcore==1.*->httpx->blosc2>=2.3.0->tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading h11-0.14.0-py3-none-any.whl.metadata (8.2 kB)\n",
            "Collecting sniffio>=1.1 (from anyio->httpx->blosc2>=2.3.0->tables>=3.5->hicmatrix>=15->pyGenomeTracks)\n",
            "  Downloading sniffio-1.3.1-py3-none-any.whl.metadata (3.9 kB)\n",
            "Downloading pyGenomeTracks-3.9-py3-none-any.whl (127 kB)\n",
            "Downloading bx_python-0.13.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (6.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.0/6.0 MB\u001b[0m \u001b[31m91.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading future-1.0.0-py3-none-any.whl (491 kB)\n",
            "Downloading gffutils-0.13-py3-none-any.whl (1.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m40.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading HiCMatrix-17.2-py3-none-any.whl (47 kB)\n",
            "Downloading matplotlib-3.8.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m11.6/11.6 MB\u001b[0m \u001b[31m83.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading numpy-2.2.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.4 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m16.4/16.4 MB\u001b[0m \u001b[31m91.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pyBigWig-0.3.24-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (185 kB)\n",
            "Downloading pyfaidx-0.8.1.3-py3-none-any.whl (28 kB)\n",
            "Downloading pysam-0.23.0-cp311-cp311-manylinux_2_28_x86_64.whl (26.0 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m26.0/26.0 MB\u001b[0m \u001b[31m61.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pytest-8.3.4-py3-none-any.whl (343 kB)\n",
            "Downloading argcomplete-3.5.3-py3-none-any.whl (43 kB)\n",
            "Downloading argh-0.31.3-py3-none-any.whl (44 kB)\n",
            "Downloading contourpy-1.3.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (326 kB)\n",
            "Downloading cooler-0.10.3-py3-none-any.whl (109 kB)\n",
            "Downloading cycler-0.12.1-py3-none-any.whl (8.3 kB)\n",
            "Downloading fonttools-4.56.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.9 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.9/4.9 MB\u001b[0m \u001b[31m75.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading kiwisolver-1.4.8-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.4 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.4/1.4 MB\u001b[0m \u001b[31m47.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pandas-2.2.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m13.1/13.1 MB\u001b[0m \u001b[31m94.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pillow-11.1.0-cp311-cp311-manylinux_2_28_x86_64.whl (4.5 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.5/4.5 MB\u001b[0m \u001b[31m77.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pyparsing-3.2.1-py3-none-any.whl (107 kB)\n",
            "Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)\n",
            "Downloading scipy-1.15.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (40.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m40.6/40.6 MB\u001b[0m \u001b[31m43.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading sortedcontainers-2.4.0-py2.py3-none-any.whl (29 kB)\n",
            "Downloading tables-3.10.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (7.5 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m7.5/7.5 MB\u001b[0m \u001b[31m85.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading importlib_metadata-8.6.1-py3-none-any.whl (26 kB)\n",
            "Downloading iniconfig-2.0.0-py3-none-any.whl (5.9 kB)\n",
            "Downloading simplejson-3.19.3-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (144 kB)\n",
            "Downloading blosc2-3.0.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.4 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.4/4.4 MB\u001b[0m \u001b[31m88.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading click-8.1.8-py3-none-any.whl (98 kB)\n",
            "Downloading h5py-3.12.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.5 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.5/5.5 MB\u001b[0m \u001b[31m89.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading numexpr-2.10.2-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl (398 kB)\n",
            "Downloading pytz-2025.1-py2.py3-none-any.whl (507 kB)\n",
            "Downloading six-1.17.0-py2.py3-none-any.whl (11 kB)\n",
            "Downloading typing_extensions-4.12.2-py3-none-any.whl (37 kB)\n",
            "Downloading tzdata-2025.1-py2.py3-none-any.whl (346 kB)\n",
            "Downloading zipp-3.21.0-py3-none-any.whl (9.6 kB)\n",
            "Downloading cytoolz-1.0.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.1/2.1 MB\u001b[0m \u001b[31m62.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading multiprocess-0.70.17-py311-none-any.whl (144 kB)\n",
            "Downloading py_cpuinfo-9.0.0-py3-none-any.whl (22 kB)\n",
            "Downloading PyYAML-6.0.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (762 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m763.0/763.0 kB\u001b[0m \u001b[31m20.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading dill-0.3.9-py3-none-any.whl (119 kB)\n",
            "Downloading toolz-1.0.0-py3-none-any.whl (56 kB)\n",
            "Downloading httpx-0.28.1-py3-none-any.whl (73 kB)\n",
            "Downloading httpcore-1.0.7-py3-none-any.whl (78 kB)\n",
            "Downloading msgpack-1.1.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (403 kB)\n",
            "Downloading ndindex-1.9.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (506 kB)\n",
            "Downloading anyio-4.8.0-py3-none-any.whl (96 kB)\n",
            "Downloading h11-0.14.0-py3-none-any.whl (58 kB)\n",
            "Downloading sniffio-1.3.1-py3-none-any.whl (10 kB)\n",
            "Building wheels for collected packages: intervaltree, pybedtools, asciitree\n",
            "  Building wheel for intervaltree (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for intervaltree: filename=intervaltree-3.1.0-py2.py3-none-any.whl size=26096 sha256=ab6fa9348a365d454fe2a127555b1b414b6ab297e3061f613cb937e0b0f72471\n",
            "  Stored in directory: /root/.cache/pip/wheels/31/d7/d9/eec6891f78cac19a693bd40ecb8365d2f4613318c145ec9816\n",
            "  Building wheel for pybedtools (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pybedtools: filename=pybedtools-0.11.0-cp311-cp311-linux_x86_64.whl size=12749692 sha256=5f80c8dfcb078404e09e7736b2a8681c44026b10ce569d3e833463cf152352a2\n",
            "  Stored in directory: /root/.cache/pip/wheels/29/a9/64/b72ced6183351e9fb889c0fc86eda39fab286a4b8838443314\n",
            "  Building wheel for asciitree (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for asciitree: filename=asciitree-0.3.3-py3-none-any.whl size=5033 sha256=652771301a710d6c95604c6760b8ff2ff669a70266055c769363141cb99c0b13\n",
            "  Stored in directory: /root/.cache/pip/wheels/71/c1/da/23077eb3b87d24d6f3852ed1ed1a1ac2d3c885ad6ebd2b4a07\n",
            "Successfully built intervaltree pybedtools asciitree\n",
            "Installing collected packages: sortedcontainers, pytz, py-cpuinfo, asciitree, zipp, tzdata, typing-extensions, toolz, sniffio, six, simplejson, pyyaml, pysam, pyparsing, pyBigWig, pillow, numpy, ndindex, msgpack, kiwisolver, intervaltree, iniconfig, h11, future, fonttools, dill, cycler, click, argh, argcomplete, scipy, python-dateutil, pytest, pybedtools, numexpr, multiprocess, importlib-metadata, httpcore, h5py, cytoolz, contourpy, bx-python, anyio, pyfaidx, pandas, matplotlib, httpx, gffutils, cooler, blosc2, tables, hicmatrix, pyGenomeTracks\n",
            "Successfully installed anyio-4.8.0 argcomplete-3.5.3 argh-0.31.3 asciitree-0.3.3 blosc2-3.0.0 bx-python-0.13.0 click-8.1.8 contourpy-1.3.1 cooler-0.10.3 cycler-0.12.1 cytoolz-1.0.1 dill-0.3.9 fonttools-4.56.0 future-1.0.0 gffutils-0.13 h11-0.14.0 h5py-3.12.1 hicmatrix-17.2 httpcore-1.0.7 httpx-0.28.1 importlib-metadata-8.6.1 iniconfig-2.0.0 intervaltree-3.1.0 kiwisolver-1.4.8 matplotlib-3.8.4 msgpack-1.1.0 multiprocess-0.70.17 ndindex-1.9.2 numexpr-2.10.2 numpy-2.2.2 pandas-2.2.3 pillow-11.1.0 py-cpuinfo-9.0.0 pyBigWig-0.3.24 pyGenomeTracks-3.9 pybedtools-0.11.0 pyfaidx-0.8.1.3 pyparsing-3.2.1 pysam-0.23.0 pytest-8.3.4 python-dateutil-2.9.0.post0 pytz-2025.1 pyyaml-6.0.2 scipy-1.15.1 simplejson-3.19.3 six-1.17.0 sniffio-1.3.1 sortedcontainers-2.4.0 tables-3.10.2 toolz-1.0.0 typing-extensions-4.12.2 tzdata-2025.1 zipp-3.21.0\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "PIL",
                  "cycler",
                  "importlib_metadata",
                  "kiwisolver",
                  "numexpr",
                  "pytz",
                  "six",
                  "zipp"
                ]
              },
              "id": "21e1f8f34909430c983137dc99b72617"
            }
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting ipykernel\n",
            "  Downloading ipykernel-6.29.5-py3-none-any.whl.metadata (6.3 kB)\n",
            "Collecting comm>=0.1.1 (from ipykernel)\n",
            "  Downloading comm-0.2.2-py3-none-any.whl.metadata (3.7 kB)\n",
            "Collecting debugpy>=1.6.5 (from ipykernel)\n",
            "  Downloading debugpy-1.8.12-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (1.3 kB)\n",
            "Collecting ipython>=7.23.1 (from ipykernel)\n",
            "  Downloading ipython-8.32.0-py3-none-any.whl.metadata (5.0 kB)\n",
            "Collecting jupyter-client>=6.1.12 (from ipykernel)\n",
            "  Downloading jupyter_client-8.6.3-py3-none-any.whl.metadata (8.3 kB)\n",
            "Collecting jupyter-core!=5.0.*,>=4.12 (from ipykernel)\n",
            "  Downloading jupyter_core-5.7.2-py3-none-any.whl.metadata (3.4 kB)\n",
            "Collecting matplotlib-inline>=0.1 (from ipykernel)\n",
            "  Downloading matplotlib_inline-0.1.7-py3-none-any.whl.metadata (3.9 kB)\n",
            "Collecting nest-asyncio (from ipykernel)\n",
            "  Downloading nest_asyncio-1.6.0-py3-none-any.whl.metadata (2.8 kB)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.11/site-packages (from ipykernel) (24.2)\n",
            "Collecting psutil (from ipykernel)\n",
            "  Downloading psutil-6.1.1-cp36-abi3-manylinux_2_12_x86_64.manylinux2010_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (22 kB)\n",
            "Collecting pyzmq>=24 (from ipykernel)\n",
            "  Downloading pyzmq-26.2.1-cp311-cp311-manylinux_2_28_x86_64.whl.metadata (6.2 kB)\n",
            "Collecting tornado>=6.1 (from ipykernel)\n",
            "  Downloading tornado-6.4.2-cp38-abi3-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (2.5 kB)\n",
            "Collecting traitlets>=5.4.0 (from ipykernel)\n",
            "  Downloading traitlets-5.14.3-py3-none-any.whl.metadata (10 kB)\n",
            "Collecting decorator (from ipython>=7.23.1->ipykernel)\n",
            "  Downloading decorator-5.1.1-py3-none-any.whl.metadata (4.0 kB)\n",
            "Collecting jedi>=0.16 (from ipython>=7.23.1->ipykernel)\n",
            "  Downloading jedi-0.19.2-py2.py3-none-any.whl.metadata (22 kB)\n",
            "Collecting pexpect>4.3 (from ipython>=7.23.1->ipykernel)\n",
            "  Downloading pexpect-4.9.0-py2.py3-none-any.whl.metadata (2.5 kB)\n",
            "Collecting prompt_toolkit<3.1.0,>=3.0.41 (from ipython>=7.23.1->ipykernel)\n",
            "  Downloading prompt_toolkit-3.0.50-py3-none-any.whl.metadata (6.6 kB)\n",
            "Collecting pygments>=2.4.0 (from ipython>=7.23.1->ipykernel)\n",
            "  Downloading pygments-2.19.1-py3-none-any.whl.metadata (2.5 kB)\n",
            "Collecting stack_data (from ipython>=7.23.1->ipykernel)\n",
            "  Downloading stack_data-0.6.3-py3-none-any.whl.metadata (18 kB)\n",
            "Requirement already satisfied: typing_extensions>=4.6 in /usr/local/lib/python3.11/site-packages (from ipython>=7.23.1->ipykernel) (4.12.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/site-packages (from jupyter-client>=6.1.12->ipykernel) (2.9.0.post0)\n",
            "Requirement already satisfied: platformdirs>=2.5 in /usr/local/lib/python3.11/site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel) (4.3.6)\n",
            "Collecting parso<0.9.0,>=0.8.4 (from jedi>=0.16->ipython>=7.23.1->ipykernel)\n",
            "  Downloading parso-0.8.4-py2.py3-none-any.whl.metadata (7.7 kB)\n",
            "Collecting ptyprocess>=0.5 (from pexpect>4.3->ipython>=7.23.1->ipykernel)\n",
            "  Downloading ptyprocess-0.7.0-py2.py3-none-any.whl.metadata (1.3 kB)\n",
            "Collecting wcwidth (from prompt_toolkit<3.1.0,>=3.0.41->ipython>=7.23.1->ipykernel)\n",
            "  Downloading wcwidth-0.2.13-py2.py3-none-any.whl.metadata (14 kB)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/site-packages (from python-dateutil>=2.8.2->jupyter-client>=6.1.12->ipykernel) (1.17.0)\n",
            "Collecting executing>=1.2.0 (from stack_data->ipython>=7.23.1->ipykernel)\n",
            "  Downloading executing-2.2.0-py2.py3-none-any.whl.metadata (8.9 kB)\n",
            "Collecting asttokens>=2.1.0 (from stack_data->ipython>=7.23.1->ipykernel)\n",
            "  Downloading asttokens-3.0.0-py3-none-any.whl.metadata (4.7 kB)\n",
            "Collecting pure-eval (from stack_data->ipython>=7.23.1->ipykernel)\n",
            "  Downloading pure_eval-0.2.3-py3-none-any.whl.metadata (6.3 kB)\n",
            "Downloading ipykernel-6.29.5-py3-none-any.whl (117 kB)\n",
            "Downloading comm-0.2.2-py3-none-any.whl (7.2 kB)\n",
            "Downloading debugpy-1.8.12-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m44.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading ipython-8.32.0-py3-none-any.whl (825 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m825.5/825.5 kB\u001b[0m \u001b[31m29.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading jupyter_client-8.6.3-py3-none-any.whl (106 kB)\n",
            "Downloading jupyter_core-5.7.2-py3-none-any.whl (28 kB)\n",
            "Downloading matplotlib_inline-0.1.7-py3-none-any.whl (9.9 kB)\n",
            "Downloading pyzmq-26.2.1-cp311-cp311-manylinux_2_28_x86_64.whl (874 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m874.5/874.5 kB\u001b[0m \u001b[31m27.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading tornado-6.4.2-cp38-abi3-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (437 kB)\n",
            "Downloading traitlets-5.14.3-py3-none-any.whl (85 kB)\n",
            "Downloading nest_asyncio-1.6.0-py3-none-any.whl (5.2 kB)\n",
            "Downloading psutil-6.1.1-cp36-abi3-manylinux_2_12_x86_64.manylinux2010_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (287 kB)\n",
            "Downloading jedi-0.19.2-py2.py3-none-any.whl (1.6 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.6/1.6 MB\u001b[0m \u001b[31m47.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pexpect-4.9.0-py2.py3-none-any.whl (63 kB)\n",
            "Downloading prompt_toolkit-3.0.50-py3-none-any.whl (387 kB)\n",
            "Downloading pygments-2.19.1-py3-none-any.whl (1.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.2/1.2 MB\u001b[0m \u001b[31m42.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading decorator-5.1.1-py3-none-any.whl (9.1 kB)\n",
            "Downloading stack_data-0.6.3-py3-none-any.whl (24 kB)\n",
            "Downloading asttokens-3.0.0-py3-none-any.whl (26 kB)\n",
            "Downloading executing-2.2.0-py2.py3-none-any.whl (26 kB)\n",
            "Downloading parso-0.8.4-py2.py3-none-any.whl (103 kB)\n",
            "Downloading ptyprocess-0.7.0-py2.py3-none-any.whl (13 kB)\n",
            "Downloading pure_eval-0.2.3-py3-none-any.whl (11 kB)\n",
            "Downloading wcwidth-0.2.13-py2.py3-none-any.whl (34 kB)\n",
            "Installing collected packages: wcwidth, pure-eval, ptyprocess, traitlets, tornado, pyzmq, pygments, psutil, prompt_toolkit, pexpect, parso, nest-asyncio, executing, decorator, debugpy, asttokens, stack_data, matplotlib-inline, jupyter-core, jedi, comm, jupyter-client, ipython, ipykernel\n",
            "Successfully installed asttokens-3.0.0 comm-0.2.2 debugpy-1.8.12 decorator-5.1.1 executing-2.2.0 ipykernel-6.29.5 ipython-8.32.0 jedi-0.19.2 jupyter-client-8.6.3 jupyter-core-5.7.2 matplotlib-inline-0.1.7 nest-asyncio-1.6.0 parso-0.8.4 pexpect-4.9.0 prompt_toolkit-3.0.50 psutil-6.1.1 ptyprocess-0.7.0 pure-eval-0.2.3 pygments-2.19.1 pyzmq-26.2.1 stack_data-0.6.3 tornado-6.4.2 traitlets-5.14.3 wcwidth-0.2.13\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "matplotlib_inline",
                  "pexpect",
                  "prompt_toolkit",
                  "tornado",
                  "wcwidth"
                ]
              },
              "id": "dbbea1e5ea91440fbe783038fa8ed54a"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! conda install -c bioconda ucsc-fetchchromsizes"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QrvF9hxKQYU0",
        "outputId": "b661d244-8dec-4059-bb41-2236ba3e9cd8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Channels:\n",
            " - bioconda\n",
            " - conda-forge\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Solving environment: | \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "    current version: 24.11.3\n",
            "    latest version: 25.1.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c conda-forge conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - ucsc-fetchchromsizes\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    libopenssl-static-3.4.0    |       hb9d3cd8_1         2.5 MB  conda-forge\n",
            "    libpng-1.6.46              |       h943b412_0         285 KB  conda-forge\n",
            "    mysql-connector-c-6.1.11   |    h659d440_1008         1.2 MB  conda-forge\n",
            "    ucsc-fetchchromsizes-469   |       h664eb37_1           9 KB  bioconda\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:         4.0 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  libopenssl-static  conda-forge/linux-64::libopenssl-static-3.4.0-hb9d3cd8_1 \n",
            "  libpng             conda-forge/linux-64::libpng-1.6.46-h943b412_0 \n",
            "  mysql-connector-c  conda-forge/linux-64::mysql-connector-c-6.1.11-h659d440_1008 \n",
            "  ucsc-fetchchromsi~ bioconda/linux-64::ucsc-fetchchromsizes-469-h664eb37_1 \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages:\n",
            "libopenssl-static-3. | 2.5 MB    | :   0% 0/1 [00:00<?, ?it/s]\n",
            "mysql-connector-c-6. | 1.2 MB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\n",
            "\n",
            "libpng-1.6.46        | 285 KB    | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "ucsc-fetchchromsizes | 9 KB      | :   0% 0/1 [00:00<?, ?it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "libopenssl-static-3. | 2.5 MB    | :   1% 0.006176530240644917/1 [00:00<00:17, 17.53s/it]\n",
            "mysql-connector-c-6. | 1.2 MB    | :   1% 0.01291599165947316/1 [00:00<00:10, 11.01s/it]\u001b[A\n",
            "\n",
            "libpng-1.6.46        | 285 KB    | : 100% 1.0/1 [00:00<00:00,  1.83s/it]                \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "ucsc-fetchchromsizes | 9 KB      | : 100% 1.0/1 [00:00<00:00,  7.19it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "ucsc-fetchchromsizes | 9 KB      | : 100% 1.0/1 [00:00<00:00,  7.19it/s]\u001b[A\u001b[A\u001b[A\n",
            "\n",
            "\n",
            "ucsc-fetchchromsizes | 9 KB      | : 100% 1.0/1 [00:00<00:00,  7.19it/s]\u001b[A\u001b[A\u001b[A\n",
            "libopenssl-static-3. | 2.5 MB    | : 100% 1.0/1 [00:00<00:00,  4.55it/s]\n",
            "\n",
            "libpng-1.6.46        | 285 KB    | : 100% 1.0/1 [00:00<00:00,  4.30it/s]\u001b[A\u001b[A\n",
            "\n",
            "libpng-1.6.46        | 285 KB    | : 100% 1.0/1 [00:00<00:00,  4.30it/s]\u001b[A\u001b[A\n",
            "mysql-connector-c-6. | 1.2 MB    | : 100% 1.0/1 [00:00<00:00,  2.51it/s]\u001b[A\n",
            "                                                                        \n",
            "                                                                        \u001b[A\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\n",
            "\n",
            "\n",
            "                                                                        \u001b[A\u001b[A\u001b[A\n",
            "Preparing transaction: - \b\bdone\n",
            "Verifying transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Executing transaction: - \b\bdone\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! fetchChromSizes mm10 > m.chrom.sizes"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "b_oPFJ9NRnmb",
        "outputId": "32e8a793-68e5-470c-89ca-3e9eac4163b9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO: trying CURL  for database mm10\n",
            "url: http://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/mm10.chrom.sizes\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import requests\n",
        "from google.colab import files\n",
        "\n",
        "url = \"https://hgdownload.soe.ucsc.edu/goldenPath/mm39/bigZips/mm39.chrom.sizes\"\n",
        "response = requests.get(url)\n",
        "\n",
        "with open(\"mm39.chrom.sizes\", \"wb\") as file:\n",
        "    file.write(response.content)\n",
        "\n",
        "files.download(\"mm39.chrom.sizes\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "cP348GIAWnRW",
        "outputId": "d77ef75a-4a3d-4059-fd14-0c477cf53c24"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ],
            "application/javascript": [
              "download(\"download_9f4a250c-68d3-4c97-9931-2be9869f8aba\", \"mm39.chrom.sizes\", 1350)"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! wget https://hgdownload-test.gi.ucsc.edu/goldenPath/mm10/bigZips/mm10.chrom.sizes"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qTJbbdXIYcyK",
        "outputId": "0c685cb7-c357-41b3-b5ff-e23312ed1a1c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2025-02-09 20:23:47--  https://hgdownload-test.gi.ucsc.edu/goldenPath/mm10/bigZips/mm10.chrom.sizes\n",
            "Resolving hgdownload-test.gi.ucsc.edu (hgdownload-test.gi.ucsc.edu)... 128.114.198.32, 2607:f5f0:136:1::26\n",
            "Connecting to hgdownload-test.gi.ucsc.edu (hgdownload-test.gi.ucsc.edu)|128.114.198.32|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1405 (1.4K)\n",
            "Saving to: ‘mm10.chrom.sizes’\n",
            "\n",
            "mm10.chrom.sizes    100%[===================>]   1.37K  --.-KB/s    in 0s      \n",
            "\n",
            "2025-02-09 20:23:47 (339 MB/s) - ‘mm10.chrom.sizes’ saved [1405/1405]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! conda install -c bioconda ucsc-bedgraphtobigwig"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ezjNJvmhRo29",
        "outputId": "dfb6fbf5-f7fc-4819-85eb-aa76c633f71b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Channels:\n",
            " - bioconda\n",
            " - conda-forge\n",
            "Platform: linux-64\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Solving environment: | \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "    current version: 24.11.3\n",
            "    latest version: 25.1.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c conda-forge conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - ucsc-bedgraphtobigwig\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    ucsc-bedgraphtobigwig-472  |       h664eb37_2         2.0 MB  bioconda\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:         2.0 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  ucsc-bedgraphtobi~ bioconda/linux-64::ucsc-bedgraphtobigwig-472-h664eb37_2 \n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages:\n",
            "                                                                        \n",
            "Preparing transaction: - \b\bdone\n",
            "Verifying transaction: | \b\b/ \b\bdone\n",
            "Executing transaction: \\ \b\bdone\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! apt-get install bedtools"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "CvmoBeJ0SXQM",
        "outputId": "c9ef97d7-c92d-4198-d12b-cf4bae9b64d5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following NEW packages will be installed:\n",
            "  bedtools\n",
            "0 upgraded, 1 newly installed, 0 to remove and 19 not upgraded.\n",
            "Need to get 563 kB of archives.\n",
            "After this operation, 1,548 kB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 bedtools amd64 2.30.0+dfsg-2ubuntu0.1 [563 kB]\n",
            "Fetched 563 kB in 1s (618 kB/s)\n",
            "Selecting previously unselected package bedtools.\n",
            "(Reading database ... 124926 files and directories currently installed.)\n",
            "Preparing to unpack .../bedtools_2.30.0+dfsg-2ubuntu0.1_amd64.deb ...\n",
            "Unpacking bedtools (2.30.0+dfsg-2ubuntu0.1) ...\n",
            "Setting up bedtools (2.30.0+dfsg-2ubuntu0.1) ...\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!gunzip s_8_cell.deduplicated.bedGraph.gz\n",
        "!gunzip s_epiblast.deduplicated.bedGraph.gz\n",
        "!gunzip s_icm.deduplicated.bedGraph.gz"
      ],
      "metadata": {
        "id": "Lx-eV9xRTDif"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "! head s_8_cell.deduplicated.bedGraph"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "b-R2Fer0TJNg",
        "outputId": "b5dc2cf4-a18c-4cf5-dfc8-f31aed94d3db"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "track type=bedGraph\n",
            "chr11\t3050030\t3050031\t50\n",
            "chr11\t3050047\t3050048\t66.6666666666667\n",
            "chr11\t3050048\t3050049\t50\n",
            "chr11\t3050223\t3050224\t87.5\n",
            "chr11\t3050224\t3050225\t53.3333333333333\n",
            "chr11\t3050285\t3050286\t66.6666666666667\n",
            "chr11\t3050286\t3050287\t33.3333333333333\n",
            "chr11\t3050309\t3050310\t62.5\n",
            "chr11\t3050310\t3050311\t50\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! bedtools genomecov -bg -ibam s_8_cell.deduplicated.bam -g mm10.chrom.sizes  >  s_8_cell.cover.bedGraph\n",
        "! bedtools genomecov -bg -ibam s_epiblast.deduplicated.bam -g mm10.chrom.sizes  >  s_epiblast.cover.bedGraph\n",
        "! bedtools genomecov -bg -ibam s_icm.deduplicated.bam -g mm10.chrom.sizes  >  s_icm.cover.bedGraph"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sb84aixgT83v",
        "outputId": "bd2dfaf3-5d3b-448d-f14d-0a2fbcf9c274"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "*****\n",
            "*****WARNING: Genome (-g) files are ignored when BAM input is provided. \n",
            "*****\n",
            "\n",
            "*****\n",
            "*****WARNING: Genome (-g) files are ignored when BAM input is provided. \n",
            "*****\n",
            "\n",
            "*****\n",
            "*****WARNING: Genome (-g) files are ignored when BAM input is provided. \n",
            "*****\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! bedGraphToBigWig   s_8_cell.cover.bedGraph mm39.chrom.sizes cell8_coverage.bigWig.bw\n",
        "! bedGraphToBigWig   s_icm.cover.bedGraph mm39.chrom.sizes icm_coverage.bigWig.bw\n",
        "! bedGraphToBigWig   s_epiblast.cover.bedGraph mm39.chrom.sizes epiblast_coverage.bigWig.bw"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qGGKHrHKWeWP",
        "outputId": "144e5258-84d9-4b2e-f82c-18c62bc2cf54"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Couldn't open mm39.chrom.sizes\n",
            "Couldn't open mm39.chrom.sizes\n",
            "Couldn't open mm39.chrom.sizes\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt"
      ],
      "metadata": {
        "id": "mbxhB5IYVUMq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!gunzip s_8_cell.deduplicated.bismark.cov.gz\n",
        "!gunzip s_epiblast.deduplicated.bismark.cov.gz\n",
        "!gunzip s_icm.deduplicated.bismark.cov.gz"
      ],
      "metadata": {
        "id": "uQTHIUvPVzy4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "s_8_cell = pd.read_csv('s_8_cell.deduplicated.bismark.cov', sep = '\\t', names = ['chr', 'start','stop', 'percent of methylation', 'met', 'nonmet'])\n",
        "epiblast = pd.read_csv('s_epiblast.deduplicated.bismark.cov', sep = '\\t', names = ['chr', 'start','stop', 'percent of methylation', 'met', 'nonmet'])\n",
        "icm = pd.read_csv('s_icm.deduplicated.bismark.cov', sep = '\\t', names = ['chr', 'start','stop', 'percent of methylation', 'met', 'nonmet'])"
      ],
      "metadata": {
        "id": "2c1vF2BwVcUf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "sns.histplot(data = s_8_cell, x = 'percent of methylation', bins = 20, kde = True, kde_kws = {'bw_adjust': 2})\n",
        "plt.title('8 cell')\n",
        "plt.savefig('8_cell.png')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 472
        },
        "id": "Aw8gZJl_XDTE",
        "outputId": "709110be-f01a-47c4-c4c1-5d9d1003a5cc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlUAAAHHCAYAAACWQK1nAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAaflJREFUeJzt3XlcVOX+B/DPzMAMm8Mqm4CgooC7qIhaaZFk1s20zczI9dYFU7m5VS7ZYtl1S0nrVuq9aaa/bNPUFFNTERXEFXHDQGXYYdiXmef3B3Kuo6iAg8Pyeb9e87oy5zvPfOdcgg/nPOc5MiGEABERERHdF7mpGyAiIiJqDhiqiIiIiIyAoYqIiIjICBiqiIiIiIyAoYqIiIjICBiqiIiIiIyAoYqIiIjICBiqiIiIiIyAoYqIiIjICBiqiIhMbO/evZDJZNi7d6/03GuvvQZvb2+T9UREdcdQRUQt1oULF/DSSy/Bw8MDVlZW8PPzw4IFC1BcXGzq1oioCTIzdQNERKaQmpqKvn37wtbWFhEREXBwcEBMTAzmzZuHuLg4/Pzzz6ZukYiaGIYqImqR/vvf/yIvLw8HDhxA586dAQCTJk2CXq/Hf/7zH+Tm5sLe3t7EXRJRU8LTf0TUImm1WgCAi4uLwfNubm6Qy+VQKpX3HCM2NhZPPvkk7O3tYW1tjW7dumH58uUGNefOncNzzz0HBwcHWFhYoHfv3vjll1+M90GIqNFgqCKiFmnQoEEAgPHjxyMhIQGpqan4/vvvsWrVKrz55puwtra+6+t37dqFhx9+GGfPnsWUKVOwePFiDB48GFu3bpVqzpw5g379+iExMRGzZs3C4sWLYW1tjeHDh+PHH39syI9HRCYgE0IIUzdBRGQKH3zwAT766COUlJRIz73zzjv44IMP7vo6nU4HX19f6PV6JCQkwM7OTtomhIBMJgMAhISEICMjA0ePHoVKpZK2Dxw4EJmZmTh//jyAqqv/Bg8ejD/++EMKe6+99hr27t2LK1euGO8DE1GD4pEqImqxvL298fDDD+PLL7/EDz/8gHHjxuGjjz7CypUr7/q648ePIzk5GVOnTjUIVACkQJWTk4M9e/bghRdeQEFBAbKyspCVlYXs7GyEhobiwoULuHbtWkN9NCIyAU5UJ6IWaePGjZg0aRLOnz8PDw8PAMCIESOg1+sxc+ZMjBo1Co6OjjW+9tKlSwCALl263HH8ixcvQgiBOXPmYM6cOTXWZGRkoE2bNvf5SYiosWCoIqIW6fPPP0fPnj2lQFXtb3/7G9auXYvjx48jJCSk3uPr9XoAwFtvvYXQ0NAaazp06FDv8Ymo8WGoIqIWKT09vcYlEyoqKgAAlZWVd3xt+/btAQCnT5++Y/Bq164dAMDc3Py+whkRNR2cU0VELVLHjh1x/PhxabJ4te+++w5yuRzdunW742t79eoFHx8fLFu2DHl5eQbbqq/9cXZ2xqBBg/DFF18gLS3ttjEyMzPv/0MQUaPCI1VE1CJNnz4d27dvx0MPPYSIiAg4Ojpi69at2L59OyZMmAB3d/c7vlYul2PVqlV4+umn0aNHD4wdOxZubm44d+4czpw5g507dwIAoqKiMHDgQHTt2hUTJ05Eu3btkJ6ejpiYGFy9ehUnTpx4UB+XiB4AhioiapEefvhhHDp0CPPnz8fnn3+O7Oxs+Pj44MMPP8SMGTPu+frQ0FD88ccfeO+997B48WLo9Xq0b98eEydOlGoCAgJw7NgxvPfee1i7di2ys7Ph7OyMnj17Yu7cuQ358YjIBLhOFREREZERcE4VERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAdepeoD0ej2uX7+OVq1aSXeyJyIiosZNCIGCggK4u7tDLr/z8SiGqgfo+vXr8PT0NHUbREREVA+pqam33YT9ZgxVD1CrVq0AVP2folarTdwNERER1YZWq4Wnp6f0e/xOGKoeoOpTfmq1mqGKiIioibnX1B1OVCciIiIyAoYqIiIiIiMwaajS6XSYM2cOfHx8YGlpifbt2+P999/Hzfd4FkJg7ty5cHNzg6WlJUJCQnDhwgWDcXJycjB69Gio1WrY2dlh/PjxKCwsNKg5efIkHnroIVhYWMDT0xOLFi26rZ/NmzfDz88PFhYW6Nq1K3777TeD7bXphYiIiFomk4aqTz75BKtWrcLKlSuRmJiITz75BIsWLcKKFSukmkWLFuGzzz7D6tWrERsbC2tra4SGhqK0tFSqGT16NM6cOYNdu3Zh69at2L9/PyZNmiRt12q1GDJkCNq2bYu4uDh8+umnmD9/Pr788kup5tChQxg1ahTGjx+P48ePY/jw4Rg+fDhOnz5dp16IiIiohRImNGzYMDFu3DiD50aMGCFGjx4thBBCr9cLV1dX8emnn0rb8/LyhEqlEt99950QQoizZ88KAOLo0aNSzfbt24VMJhPXrl0TQgjx+eefC3t7e1FWVibVzJw5U3Tq1En6+oUXXhDDhg0z6CUoKEj8/e9/r3Uv95Kfny8AiPz8/FrVExERkenV9ve3SY9U9e/fH9HR0Th//jwA4MSJEzhw4ACGDh0KAEhOToZGo0FISIj0GltbWwQFBSEmJgYAEBMTAzs7O/Tu3VuqCQkJgVwuR2xsrFTz8MMPQ6lUSjWhoaFISkpCbm6uVHPz+1TXVL9PbXq5VVlZGbRarcGDiIiImieTLqkwa9YsaLVa+Pn5QaFQQKfT4cMPP8To0aMBABqNBgDg4uJi8DoXFxdpm0ajgbOzs8F2MzMzODg4GNT4+PjcNkb1Nnt7e2g0mnu+z716udXChQvx3nvv1WJPEBERUVNn0iNVmzZtwvr167FhwwbEx8dj3bp1+Ne//oV169aZsi2jmT17NvLz86VHamqqqVsiIiKiBmLSI1XTp0/HrFmz8NJLLwEAunbtir/++gsLFy5EWFgYXF1dAQDp6elwc3OTXpeeno4ePXoAAFxdXZGRkWEwbmVlJXJycqTXu7q6Ij093aCm+ut71dy8/V693EqlUkGlUtVuZxAREVGTZtIjVcXFxbfdmFChUECv1wMAfHx84OrqiujoaGm7VqtFbGwsgoODAQDBwcHIy8tDXFycVLNnzx7o9XoEBQVJNfv370dFRYVUs2vXLnTq1An29vZSzc3vU11T/T616YWIiIhasAc0cb5GYWFhok2bNmLr1q0iOTlZbNmyRTg5OYkZM2ZINR9//LGws7MTP//8szh58qR45plnhI+PjygpKZFqnnjiCdGzZ08RGxsrDhw4IHx9fcWoUaOk7Xl5ecLFxUWMGTNGnD59WmzcuFFYWVmJL774Qqo5ePCgMDMzE//6179EYmKimDdvnjA3NxenTp2qUy93w6v/iIiImp7a/v42aajSarViypQpwsvLS1hYWIh27dqJd955x2DpA71eL+bMmSNcXFyESqUSjz32mEhKSjIYJzs7W4waNUrY2NgItVotxo4dKwoKCgxqTpw4IQYOHChUKpVo06aN+Pjjj2/rZ9OmTaJjx45CqVSKzp07i23bthlsr00vd8NQRURE1PTU9ve3TIibli+nBqXVamFra4v8/HzeUJmIiKiJqO3vb977j4iIiMgITHr1HxlPSkoKsrKyGmRsJycneHl5NcjYREREzQVDVTOQkpICP39/lBQXN8j4llZWOJeYyGBFRER0FwxVzUBWVhZKiosxeuancPFqb9Sx01MuYf0n05GVlcVQRUREdBcMVc2Ii1d7ePh2NnUbRERELRInqhMREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZAUMVERERkREwVBEREREZgUlDlbe3N2Qy2W2P8PBwAEBpaSnCw8Ph6OgIGxsbjBw5Eunp6QZjpKSkYNiwYbCysoKzszOmT5+OyspKg5q9e/eiV69eUKlU6NChA9auXXtbL1FRUfD29oaFhQWCgoJw5MgRg+216YWIiIhaLpOGqqNHjyItLU167Nq1CwDw/PPPAwCmTZuGX3/9FZs3b8a+fftw/fp1jBgxQnq9TqfDsGHDUF5ejkOHDmHdunVYu3Yt5s6dK9UkJydj2LBhGDx4MBISEjB16lRMmDABO3fulGq+//57REZGYt68eYiPj0f37t0RGhqKjIwMqeZevRAREVHLJhNCCFM3UW3q1KnYunUrLly4AK1Wi9atW2PDhg147rnnAADnzp2Dv78/YmJi0K9fP2zfvh1PPfUUrl+/DhcXFwDA6tWrMXPmTGRmZkKpVGLmzJnYtm0bTp8+Lb3PSy+9hLy8POzYsQMAEBQUhD59+mDlypUAAL1eD09PT0yePBmzZs1Cfn7+PXupDa1WC1tbW+Tn50OtVhttv8XHxyMwMBCRUVvg4dvZaOMCwNULZ7AkfATi4uLQq1cvo45NRETUFNT293ejmVNVXl6Ob7/9FuPGjYNMJkNcXBwqKioQEhIi1fj5+cHLywsxMTEAgJiYGHTt2lUKVAAQGhoKrVaLM2fOSDU3j1FdUz1GeXk54uLiDGrkcjlCQkKkmtr0UpOysjJotVqDBxERETVPjSZU/fTTT8jLy8Nrr70GANBoNFAqlbCzszOoc3FxgUajkWpuDlTV26u33a1Gq9WipKQEWVlZ0Ol0NdbcPMa9eqnJwoULYWtrKz08PT3vvSOIiIioSWo0oerrr7/G0KFD4e7ubupWjGb27NnIz8+XHqmpqaZuiYiIiBqImakbAIC//voLu3fvxpYtW6TnXF1dUV5ejry8PIMjROnp6XB1dZVqbr1Kr/qKvJtrbr1KLz09HWq1GpaWllAoFFAoFDXW3DzGvXqpiUqlgkqlquVeICIioqasURypWrNmDZydnTFs2DDpucDAQJibmyM6Olp6LikpCSkpKQgODgYABAcH49SpUwZX6e3atQtqtRoBAQFSzc1jVNdUj6FUKhEYGGhQo9frER0dLdXUphciIiJq2Ux+pEqv12PNmjUICwuDmdn/2rG1tcX48eMRGRkJBwcHqNVqTJ48GcHBwdLVdkOGDEFAQADGjBmDRYsWQaPR4N1330V4eLh0hOj111/HypUrMWPGDIwbNw579uzBpk2bsG3bNum9IiMjERYWht69e6Nv375YtmwZioqKMHbs2Fr3QkRERC2byUPV7t27kZKSgnHjxt22benSpZDL5Rg5ciTKysoQGhqKzz//XNquUCiwdetWvPHGGwgODoa1tTXCwsKwYMECqcbHxwfbtm3DtGnTsHz5cnh4eOCrr75CaGioVPPiiy8iMzMTc+fOhUajQY8ePbBjxw6Dyev36oWIiIhatka1TlVzx3WqiIiImp4mt04VERERUVPGUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBCYPVdeuXcMrr7wCR0dHWFpaomvXrjh27Ji0XQiBuXPnws3NDZaWlggJCcGFCxcMxsjJycHo0aOhVqthZ2eH8ePHo7Cw0KDm5MmTeOihh2BhYQFPT08sWrTotl42b94MPz8/WFhYoGvXrvjtt98MttemFyIiImqZTBqqcnNzMWDAAJibm2P79u04e/YsFi9eDHt7e6lm0aJF+Oyzz7B69WrExsbC2toaoaGhKC0tlWpGjx6NM2fOYNeuXdi6dSv279+PSZMmSdu1Wi2GDBmCtm3bIi4uDp9++inmz5+PL7/8Uqo5dOgQRo0ahfHjx+P48eMYPnw4hg8fjtOnT9epFyIiImqZZEIIYao3nzVrFg4ePIg///yzxu1CCLi7u+Of//wn3nrrLQBAfn4+XFxcsHbtWrz00ktITExEQEAAjh49it69ewMAduzYgSeffBJXr16Fu7s7Vq1ahXfeeQcajQZKpVJ6759++gnnzp0DALz44osoKirC1q1bpffv168fevTogdWrV9eql3vRarWwtbVFfn4+1Gp1/XfcLeLj4xEYGIjIqC3w8O1stHEB4OqFM1gSPgJxcXHo1auXUccmIiJqCmr7+9ukR6p++eUX9O7dG88//zycnZ3Rs2dP/Pvf/5a2JycnQ6PRICQkRHrO1tYWQUFBiImJAQDExMTAzs5OClQAEBISArlcjtjYWKnm4YcflgIVAISGhiIpKQm5ublSzc3vU11T/T616eVWZWVl0Gq1Bg8iIiJqnkwaqi5fvoxVq1bB19cXO3fuxBtvvIE333wT69atAwBoNBoAgIuLi8HrXFxcpG0ajQbOzs4G283MzODg4GBQU9MYN7/HnWpu3n6vXm61cOFC2NraSg9PT8977RIiIiJqokwaqvR6PXr16oWPPvoIPXv2xKRJkzBx4kSsXr3alG0ZzezZs5Gfny89UlNTTd0SERERNRCThio3NzcEBAQYPOfv74+UlBQAgKurKwAgPT3doCY9PV3a5urqioyMDIPtlZWVyMnJMaipaYyb3+NONTdvv1cvt1KpVFCr1QYPIiIiap5MGqoGDBiApKQkg+fOnz+Ptm3bAgB8fHzg6uqK6OhoabtWq0VsbCyCg4MBAMHBwcjLy0NcXJxUs2fPHuj1egQFBUk1+/fvR0VFhVSza9cudOrUSbrSMDg42OB9qmuq36c2vRAREVHLZdJQNW3aNBw+fBgfffQRLl68iA0bNuDLL79EeHg4AEAmk2Hq1Kn44IMP8Msvv+DUqVN49dVX4e7ujuHDhwOoOrL1xBNPYOLEiThy5AgOHjyIiIgIvPTSS3B3dwcAvPzyy1AqlRg/fjzOnDmD77//HsuXL0dkZKTUy5QpU7Bjxw4sXrwY586dw/z583Hs2DFERETUuhciIiJqucxM+eZ9+vTBjz/+iNmzZ2PBggXw8fHBsmXLMHr0aKlmxowZKCoqwqRJk5CXl4eBAwdix44dsLCwkGrWr1+PiIgIPPbYY5DL5Rg5ciQ+++wzabutrS1+//13hIeHIzAwEE5OTpg7d67BWlb9+/fHhg0b8O677+Ltt9+Gr68vfvrpJ3Tp0qVOvRAREVHLZNJ1qloarlNFRETU9DSJdaqIiIiImguGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjMGmomj9/PmQymcHDz89P2l5aWorw8HA4OjrCxsYGI0eORHp6usEYKSkpGDZsGKysrODs7Izp06ejsrLSoGbv3r3o1asXVCoVOnTogLVr197WS1RUFLy9vWFhYYGgoCAcOXLEYHtteiEiIqKWy+RHqjp37oy0tDTpceDAAWnbtGnT8Ouvv2Lz5s3Yt28frl+/jhEjRkjbdTodhg0bhvLychw6dAjr1q3D2rVrMXfuXKkmOTkZw4YNw+DBg5GQkICpU6diwoQJ2Llzp1Tz/fffIzIyEvPmzUN8fDy6d++O0NBQZGRk1LoXIiIiatlMHqrMzMzg6uoqPZycnAAA+fn5+Prrr7FkyRI8+uijCAwMxJo1a3Do0CEcPnwYAPD777/j7Nmz+Pbbb9GjRw8MHToU77//PqKiolBeXg4AWL16NXx8fLB48WL4+/sjIiICzz33HJYuXSr1sGTJEkycOBFjx45FQEAAVq9eDSsrK3zzzTe17oWIiIhaNpOHqgsXLsDd3R3t2rXD6NGjkZKSAgCIi4tDRUUFQkJCpFo/Pz94eXkhJiYGABATE4OuXbvCxcVFqgkNDYVWq8WZM2ekmpvHqK6pHqO8vBxxcXEGNXK5HCEhIVJNbXqpSVlZGbRarcGDiIiImieThqqgoCCsXbsWO3bswKpVq5CcnIyHHnoIBQUF0Gg0UCqVsLOzM3iNi4sLNBoNAECj0RgEqurt1dvuVqPValFSUoKsrCzodLoaa24e41691GThwoWwtbWVHp6enrXbMURERNTkmJnyzYcOHSr9u1u3bggKCkLbtm2xadMmWFpamrAz45g9ezYiIyOlr7VaLYMVERFRM2Xy0383s7OzQ8eOHXHx4kW4urqivLwceXl5BjXp6elwdXUFALi6ut52BV711/eqUavVsLS0hJOTExQKRY01N49xr15qolKpoFarDR5ERETUPDWqUFVYWIhLly7Bzc0NgYGBMDc3R3R0tLQ9KSkJKSkpCA4OBgAEBwfj1KlTBlfp7dq1C2q1GgEBAVLNzWNU11SPoVQqERgYaFCj1+sRHR0t1dSmFyIiImrZTHr676233sLTTz+Ntm3b4vr165g3bx4UCgVGjRoFW1tbjB8/HpGRkXBwcIBarcbkyZMRHByMfv36AQCGDBmCgIAAjBkzBosWLYJGo8G7776L8PBwqFQqAMDrr7+OlStXYsaMGRg3bhz27NmDTZs2Ydu2bVIfkZGRCAsLQ+/evdG3b18sW7YMRUVFGDt2LADUqhciIiJq2Uwaqq5evYpRo0YhOzsbrVu3xsCBA3H48GG0bt0aALB06VLI5XKMHDkSZWVlCA0Nxeeffy69XqFQYOvWrXjjjTcQHBwMa2trhIWFYcGCBVKNj48Ptm3bhmnTpmH58uXw8PDAV199hdDQUKnmxRdfRGZmJubOnQuNRoMePXpgx44dBpPX79ULERERtWwyIYQwdRMthVarha2tLfLz8406vyo+Ph6BgYGIjNoCD9/ORhsXAK5eOIMl4SMQFxeHXr16GXVsIiKipqC2v78b1ZwqIiIioqaKoYqIiIjICBiqiIiIiIyAoYqIiIjICBiqiIiIiIyAoYqIiIjICBiqiIiIiIyAoYqIiIjICBiqiIiIiIyAoYqIiIjICBiqiIiIiIyAoYqIiIjICBiqiIiIiIyAoYqIiIjICBiqiIiIiIyAoYqIiIjICOoVqtq1a4fs7Ozbns/Ly0O7du3uuykiIiKipqZeoerKlSvQ6XS3PV9WVoZr167dd1NERERETY1ZXYp/+eUX6d87d+6Era2t9LVOp0N0dDS8vb2N1hwRERE1PykpKcjKyjL6uE5OTvDy8jL6uLVVp1A1fPhwAIBMJkNYWJjBNnNzc3h7e2Px4sVGa46IiIial5SUFPj5+6OkuNjoY1taWeFcYqLJglWdQpVerwcA+Pj44OjRo3BycmqQpoiIiKh5ysrKQklxMUbP/BQuXu2NNm56yiWs/2Q6srKymkaoqpacnGzsPoiIiKgFcfFqDw/fzqZuw6jqFaoAIDo6GtHR0cjIyJCOYFX75ptv7rsxIiIioqakXqHqvffew4IFC9C7d2+4ublBJpMZuy8iIiKiJqVeoWr16tVYu3YtxowZY+x+iIiIiJqkeq1TVV5ejv79+xu7FyIiIqImq16hasKECdiwYYOxeyEiIiJqsup1+q+0tBRffvkldu/ejW7dusHc3Nxg+5IlS4zSHBEREVFTUa9QdfLkSfTo0QMAcPr0aYNtnLRORERELVG9QtUff/xh7D6IiIiImrR6zakiIiIiIkP1OlI1ePDgu57m27NnT70bIiIiImqK6hWqqudTVauoqEBCQgJOnz59242WiYiIiFqCeoWqpUuX1vj8/PnzUVhYeF8NERERETVFRp1T9corr9T7vn8ff/wxZDIZpk6dKj1XWlqK8PBwODo6wsbGBiNHjkR6errB61JSUjBs2DBYWVnB2dkZ06dPR2VlpUHN3r170atXL6hUKnTo0AFr16697f2joqLg7e0NCwsLBAUF4ciRIwbba9MLERERtVxGDVUxMTGwsLCo8+uOHj2KL774At26dTN4ftq0afj111+xefNm7Nu3D9evX8eIESOk7TqdDsOGDUN5eTkOHTqEdevWYe3atZg7d65Uk5ycjGHDhmHw4MFISEjA1KlTMWHCBOzcuVOq+f777xEZGYl58+YhPj4e3bt3R2hoKDIyMmrdCxEREbVs9Tr9d2uYEEIgLS0Nx44dw5w5c+o0VmFhIUaPHo1///vf+OCDD6Tn8/Pz8fXXX2PDhg149NFHAQBr1qyBv78/Dh8+jH79+uH333/H2bNnsXv3bri4uKBHjx54//33MXPmTMyfPx9KpRKrV6+Gj48PFi9eDADw9/fHgQMHsHTpUoSGhgKoWqx04sSJGDt2LICqextu27YN33zzDWbNmlWrXoiIiKhlq9eRKltbW4OHg4MDBg0ahN9++w3z5s2r01jh4eEYNmwYQkJCDJ6Pi4tDRUWFwfN+fn7w8vJCTEwMgKojY127doWLi4tUExoaCq1WizNnzkg1t44dGhoqjVFeXo64uDiDGrlcjpCQEKmmNr3UpKysDFqt1uBBREREzVO9jlStWbPGKG++ceNGxMfH4+jRo7dt02g0UCqVsLOzM3jexcUFGo1Gqrk5UFVvr952txqtVouSkhLk5uZCp9PVWHPu3Lla91KThQsX4r333rvjdiIiImo+6hWqqsXFxSExMREA0LlzZ/Ts2bPWr01NTcWUKVOwa9eues3Dagpmz56NyMhI6WutVgtPT08TdkREREQNpV6hKiMjAy+99BL27t0rHb3Jy8vD4MGDsXHjRrRu3fqeY8TFxSEjIwO9evWSntPpdNi/fz9WrlyJnTt3ory8HHl5eQZHiNLT0+Hq6goAcHV1ve0qveor8m6uufUqvfT0dKjValhaWkKhUEChUNRYc/MY9+qlJiqVCiqV6p77goiIiJq+es2pmjx5MgoKCnDmzBnk5OQgJycHp0+fhlarxZtvvlmrMR577DGcOnUKCQkJ0qN3794YPXq09G9zc3NER0dLr0lKSkJKSgqCg4MBAMHBwTh16pTBVXq7du2CWq1GQECAVHPzGNU11WMolUoEBgYa1Oj1ekRHR0s1gYGB9+yFiIiIWrZ6HanasWMHdu/eDX9/f+m5gIAAREVFYciQIbUao1WrVujSpYvBc9bW1nB0dJSeHz9+PCIjI+Hg4AC1Wo3JkycjODhYutpuyJAhCAgIwJgxY7Bo0SJoNBq8++67CA8Pl44Qvf7661i5ciVmzJiBcePGYc+ePdi0aRO2bdsmvW9kZCTCwsLQu3dv9O3bF8uWLUNRUZF0NaCtre09eyEiIqKWrV6hSq/Xw9zc/Lbnzc3Nodfr77upakuXLoVcLsfIkSNRVlaG0NBQfP7559J2hUKBrVu34o033kBwcDCsra0RFhaGBQsWSDU+Pj7Ytm0bpk2bhuXLl8PDwwNfffWVtJwCALz44ovIzMzE3LlzodFo0KNHD+zYscNg8vq9eiEiIqKWrV6h6tFHH8WUKVPw3Xffwd3dHQBw7do1TJs2DY899li9m9m7d6/B1xYWFoiKikJUVNQdX9O2bVv89ttvdx130KBBOH78+F1rIiIiEBERccfttemFiIiIWq56zalauXIltFotvL290b59e7Rv3x4+Pj7QarVYsWKFsXskIiIiavTqdaTK09MT8fHx2L17t7SWk7+//22LbBIRERG1FHU6UrVnzx4EBARAq9VCJpPh8ccfx+TJkzF58mT06dMHnTt3xp9//tlQvRIRERE1WnUKVcuWLcPEiROhVqtv22Zra4u///3vWLJkidGaIyIiImoq6hSqTpw4gSeeeOKO24cMGYK4uLj7boqIiIioqalTqEpPT69xKYVqZmZmyMzMvO+miIiIiJqaOoWqNm3a4PTp03fcfvLkSbi5ud13U0RERERNTZ1C1ZNPPok5c+agtLT0tm0lJSWYN28ennrqKaM1R0RERNRU1GlJhXfffRdbtmxBx44dERERgU6dOgEAzp07h6ioKOh0OrzzzjsN0igRERFRY1anUOXi4oJDhw7hjTfewOzZsyGEAADIZDKEhoYiKirK4NYuRERERC1FnRf/rL4tTG5uLi5evAghBHx9fWFvb98Q/RERERE1CfVaUR0A7O3t0adPH2P2QkRERNRk1evef0RERERkiKGKiIiIyAgYqoiIiIiMgKGKiIiIyAgYqoiIiIiMgKGKiIiIyAgYqoiIiIiMgKGKiIiIyAgYqoiIiIiMgKGKiIiIyAgYqoiIiIiMgKGqGTh2vRROf5sBTYkMer0wdTtEREQtEkNVM7D7cjGs/R/GwUxzfH0wGYlpWlO3RERE1OIwVDUDzwe0gvbYL1DKBYrLdfj9bDoupBeYui0iIqIWhaGqGWjvYI7c6C8xrE0FuraxBQDsPJOO1JxiE3dGRETUcjBUNSNyGTCoU2t0cLaBTghsO5WG4vJKU7dFRETUIjBUNTNymQyhnV3Q2kaFsko9jiTnmLolIiKiFoGhqhkyk8vxkK8TAODUtXzkFpebuCMiIqLmj6GqmfJ0sIK3oxX0Ajh4McvU7RARETV7DFXN2MAOTpABuJRZBE1+qanbISIiatYYqpoxRxsV/FxbAQBOXM0zbTNERETNHENVM9fNww4AcCGjECXlOtM2Q0RE1IyZNFStWrUK3bp1g1qthlqtRnBwMLZv3y5tLy0tRXh4OBwdHWFjY4ORI0ciPT3dYIyUlBQMGzYMVlZWcHZ2xvTp01FZabiMwN69e9GrVy+oVCp06NABa9euva2XqKgoeHt7w8LCAkFBQThy5IjB9tr00hi5qFVwbqWCTi9wliutExERNRiThioPDw98/PHHiIuLw7Fjx/Doo4/imWeewZkzZwAA06ZNw6+//orNmzdj3759uH79OkaMGCG9XqfTYdiwYSgvL8ehQ4ewbt06rF27FnPnzpVqkpOTMWzYMAwePBgJCQmYOnUqJkyYgJ07d0o133//PSIjIzFv3jzEx8eje/fuCA0NRUZGhlRzr14aK5lMJi0IeupaPoTgvQGJiIgagklD1dNPP40nn3wSvr6+6NixIz788EPY2Njg8OHDyM/Px9dff40lS5bg0UcfRWBgINasWYNDhw7h8OHDAIDff/8dZ8+exbfffosePXpg6NCheP/99xEVFYXy8qplBFavXg0fHx8sXrwY/v7+iIiIwHPPPYelS5dKfSxZsgQTJ07E2LFjERAQgNWrV8PKygrffPMNANSql8ask2srKM3kyC+pQApXWSciImoQjWZOlU6nw8aNG1FUVITg4GDExcWhoqICISEhUo2fnx+8vLwQExMDAIiJiUHXrl3h4uIi1YSGhkKr1UpHu2JiYgzGqK6pHqO8vBxxcXEGNXK5HCEhIVJNbXqpSVlZGbRarcHDFMwVcvjfmLDOU4BEREQNw+Sh6tSpU7CxsYFKpcLrr7+OH3/8EQEBAdBoNFAqlbCzszOod3FxgUajAQBoNBqDQFW9vXrb3Wq0Wi1KSkqQlZUFnU5XY83NY9yrl5osXLgQtra20sPT07N2O6UBdLoRqpKzilCp05usDyIioubK5KGqU6dOSEhIQGxsLN544w2EhYXh7Nmzpm7LKGbPno38/HzpkZqaarJeXNUWaGVhhgqdwF88BUhERGR0ZqZuQKlUokOHDgCAwMBAHD16FMuXL8eLL76I8vJy5OXlGRwhSk9Ph6urKwDA1dX1tqv0qq/Iu7nm1qv00tPToVarYWlpCYVCAYVCUWPNzWPcq5eaqFQqqFSqOuyNhiOTydDB2QbHU/JwPr0A7VvbmLolIiKiZsXkR6pupdfrUVZWhsDAQJibmyM6OlralpSUhJSUFAQHBwMAgoODcerUKYOr9Hbt2gW1Wo2AgACp5uYxqmuqx1AqlQgMDDSo0ev1iI6Olmpq00tT4OtcFaR4CpCIiMj4THqkavbs2Rg6dCi8vLxQUFCADRs2YO/evdi5cydsbW0xfvx4REZGwsHBAWq1GpMnT0ZwcDD69esHABgyZAgCAgIwZswYLFq0CBqNBu+++y7Cw8OlI0Svv/46Vq5ciRkzZmDcuHHYs2cPNm3ahG3btkl9REZGIiwsDL1790bfvn2xbNkyFBUVYezYsQBQq16agupTgAWllbiSXYwOzjxaRUREZCwmDVUZGRl49dVXkZaWBltbW3Tr1g07d+7E448/DgBYunQp5HI5Ro4cibKyMoSGhuLzzz+XXq9QKLB161a88cYbCA4OhrW1NcLCwrBgwQKpxsfHB9u2bcO0adOwfPlyeHh44KuvvkJoaKhU8+KLLyIzMxNz586FRqNBjx49sGPHDoPJ6/fqpSm4+RTghYwChioiIiIjMmmo+vrrr++63cLCAlFRUYiKirpjTdu2bfHbb7/ddZxBgwbh+PHjd62JiIhARETEffXSFPjeCFVXsouh0wso5DJTt0RERNQsNLo5VdSwXNQWsDCXo7xSD01+qanbISIiajYYqloYuUyGto7WAIDk7CITd0NERNR8MFS1QN6OVgCAKwxVRERERsNQ1QJVH6nKLixHQWmFibshIiJqHhiqWiBLcwXcbC0AAFeyubo6ERGRMTBUtVDeN45W/cVTgEREREbBUNVCVc+rSskpRqWeq6sTERHdL4aqFqp1KxUszRWo0AkurUBERGQEDFUtlEwmg6eDJQAgNafExN0QERE1fQxVLZinQ9UpwNRcTlYnIiK6XwxVLZiXfVWoSteWoryS86qIiIjuB0NVC6a2NIetpTn0AriWx1OARERE94OhqoXztK+eV8VTgERERPeDoaqF47wqIiIi42CoauE8bhypyiosR3F5pYm7ISIiaroYqlo4K6UZnGyUAICruZxXRUREVF8MVQQPu6pTgJysTkREVH8MVQR3+6qbKzNUERER1R9DFaGNXdW8quzCcpRW6EzcDRERUdPEUEWwUprB3socAHCdR6uIiIjqhaGKAPzvaBVPARIREdUPQxUBANrYM1QRERHdD4YqAgC43zhSlVFQxvsAEhER1QNDFQEA1BbmaGVhBiGAtHwerSIiIqorhiqSVM+rup5XauJOiIiImh6GKpJwsjoREVH9MVSRpHqyukZbikod51URERHVBUMVSewszWGlVECnF0jXlpm6HSIioiaFoYokMplMugrwGierExER1QlDFRmQJqvnMlQRERHVBUMVGZBCVX4J9Hph4m6IiIiaDoYqMuBoo4TKTI4KnUBmIedVERER1ZaZqRugxkV+Y15VclYRruWVwNnUDRE1ISkpKcjKyjL6uE5OTvDy8jL6uERkXAxVdBt3O4uqUJVbAmcrU3dD1DSkpKTAz98fJcXFRh/b0soK5xITGayIGjmThqqFCxdiy5YtOHfuHCwtLdG/f3988skn6NSpk1RTWlqKf/7zn9i4cSPKysoQGhqKzz//HC4uLlJNSkoK3njjDfzxxx+wsbFBWFgYFi5cCDOz/328vXv3IjIyEmfOnIGnpyfeffddvPbaawb9REVF4dNPP4VGo0H37t2xYsUK9O3bt069NAf/W1m9BD0sTdwMURORlZWFkuJijJ75KVy82htt3PSUS1j/yXRkZWUxVBE1ciYNVfv27UN4eDj69OmDyspKvP322xgyZAjOnj0La2trAMC0adOwbds2bN68Gba2toiIiMCIESNw8OBBAIBOp8OwYcPg6uqKQ4cOIS0tDa+++irMzc3x0UcfAQCSk5MxbNgwvP7661i/fj2io6MxYcIEuLm5ITQ0FADw/fffIzIyEqtXr0ZQUBCWLVuG0NBQJCUlwdnZuVa9NBfOrSxgJpehtFKPgkpTd0PUtLh4tYeHb2dTt0FEJmDSULVjxw6Dr9euXQtnZ2fExcXh4YcfRn5+Pr7++mts2LABjz76KABgzZo18Pf3x+HDh9GvXz/8/vvvOHv2LHbv3g0XFxf06NED77//PmbOnIn58+dDqVRi9erV8PHxweLFiwEA/v7+OHDgAJYuXSqFqiVLlmDixIkYO3YsAGD16tXYtm0bvvnmG8yaNatWvTQXCrkMrrYWuJpbgqxSXstARERUG43qN2Z+fj4AwMHBAQAQFxeHiooKhISESDV+fn7w8vJCTEwMACAmJgZdu3Y1OAUXGhoKrVaLM2fOSDU3j1FdUz1GeXk54uLiDGrkcjlCQkKkmtr00py421ad98sqk5m4EyIioqah0UxU1+v1mDp1KgYMGIAuXboAADQaDZRKJezs7AxqXVxcoNFopJpb5zRVf32vGq1Wi5KSEuTm5kKn09VYc+7cuVr3cquysjKUlf1vWQKtVnuv3dBouNtZAACyyxpV7iYiImq0Gs1vzPDwcJw+fRobN240dStGs3DhQtja2koPT09PU7dUa262lpDJgGKdDIpWrU3dDhERUaPXKEJVREQEtm7dij/++AMeHh7S866urigvL0deXp5BfXp6OlxdXaWa9PT027ZXb7tbjVqthqWlJZycnKBQKGqsuXmMe/Vyq9mzZyM/P196pKam1mJvNA5KMzla26gAACqPABN3Q0RE1PiZNFQJIRAREYEff/wRe/bsgY+Pj8H2wMBAmJubIzo6WnouKSkJKSkpCA4OBgAEBwfj1KlTyMjIkGp27doFtVqNgIAAqebmMaprqsdQKpUIDAw0qNHr9YiOjpZqatPLrVQqFdRqtcGjKam+ubKFJ69kIiIiuheTzqkKDw/Hhg0b8PPPP6NVq1bS3CRbW1tYWlrC1tYW48ePR2RkJBwcHKBWqzF58mQEBwdLV9sNGTIEAQEBGDNmDBYtWgSNRoN3330X4eHhUKmqjrS8/vrrWLlyJWbMmIFx48Zhz5492LRpE7Zt2yb1EhkZibCwMPTu3Rt9+/bFsmXLUFRUJF0NWJtemps2dpZISM2DyoOhioiIGj+5hQ3KKk1331qThqpVq1YBAAYNGmTw/Jo1a6SFOZcuXQq5XI6RI0caLLhZTaFQYOvWrXjjjTcQHBwMa2trhIWFYcGCBVKNj48Ptm3bhmnTpmH58uXw8PDAV199JS2nAAAvvvgiMjMzMXfuXGg0GvTo0QM7duwwmLx+r16am+rJ6srWbVFQpjdxN0RERIaEEDifXohLmYW4nmMOzykbkZBehprPHzU8k4YqIe6dJi0sLBAVFYWoqKg71rRt2xa//fbbXccZNGgQjh8/fteaiIgIRERE3FcvzYmV0gw2ZgKFlTKcyy7HI6ZuiIiI6IaconLsOZeBa3klN56pWgIozYSrVjeKierUeDmpqo5QJWaWm7gTIiKiKleyi7DhSAqu5ZXATC5DH297DHSuQOrylzDcz8ZkfTWadaqocXJSCVwpAs5mMVQREZHp/ZVdhK0n06DTC3g5WOExP2eoLc1x9UI69KWFJu2NoYruytGi6kjV5dwKlJTrYKlUmLgjIiJqqa7lleDXG4GqfWtrDO3iBoW88dz5g6f/6K6sFUBlQTYq9UBCap6p2yEiohaqpEKHHac10OkFfJwaX6ACGKroHmQyoOxq1T0Uj17JMXE3RETUEgkhEJ2YjsKySthZmeOJzq6NLlABDFVUC2VXzwJgqCIiItM4dS0flzKLoJDJMLSLK5RmjTO+NM6uqFEpTa06UhX/Vy4qdVyvioiIHpyiskocvJgNABjQwRHOrSxM3NGdMVTRPVVk/QUrcxmKynVITCswdTtERNSCHLyUhXKdHi5qFXp42pm6nbtiqKJ7E3r4OSkBALHJ2SZuhoiIWorreSXSH/ODOjpDJmt886huxlBFtRIghSrOqyIiooYnhMC+85kAgAA3NVxtG+9pv2oMVVQrXZyrQtWR5Bzo9aa7WSUREbUMFzMKkVFQBqVCjgEdHE3dTq0wVFGttLM3h5VSgfySCpzTcF4VERE1HL0QOHy56sxITy87WCmbxlrlDFVUK2ZyGQLb2gPgvCoiImpY5zUFyCkuh8pMjp5edqZup9YYqqjW+rWrOvwae5nzqoiIqGHo9AKHb8zfDWxrD5VZ07k9GkMV1Vq/dg4Aqo5UcV4VERE1hPPpBcgvqYCluQLdPexM3U6dMFRRrXVtYwdLcwVyiytwIcO0dwInIqLmRwiBuL9yAVTNpWqsK6ffSdPqlkxKaSbnvCoiImowV7KLkV1UDqVCjm5tbE3dTp0xVFGdVJ8CjLnEUEVERMZ17K+quVRd2qihMm86c6mqMVRRnQS3r5qsfvgy51UREZHxpOWX4HpeKeQyoKenvanbqReGKqqTbh52sFZWzas6m6Y1dTtERNRMxKfkAQD8XNWwsWga61LdiqGK6sRcIUdfn6pTgIcuZZm4GyIiag4KSitwKbPqAqimtC7VrRiqqM4GdHACABy8yHlVRER0/05dy4cQQBs7SzjZqEzdTr0xVFGd9W9fFaqOJOegvFJv4m6IiKgpq9Tpcfpa1XSS7h5N74q/mzFUUZ35ubaCg7USJRU6JKTmmbodIiJqwi5kFKKkQgcblRnat7YxdTv3haGK6kwul0lXAR68yHlVRERUP0II6Y/zrh62kMtlpm3oPjFUUb0MuHEKkJPViYiovtK1ZcgoKINCLkMXd7Wp27lvDFVULwNvTFY/npKHgtIKE3dDRERNUcLVPABARxcbWCmb5jIKN2OoonrxcrSCt6MVKvWCq6sTEVGdleqAC+kFANDkbpx8JwxVVG+PdGwNANh3PtPEnRARUVOTXCiHXgCuagu4qC1M3Y5RMFRRvT3S6X+hSgjesoaIiGpJrsDlwqp7+3X3bNrLKNyMoYrqrV87RygVclzNLcHlrCJTt0NERE2EVcf+KNXJYKVUwNe5lanbMRqGKqo3K6WZdMuafUk8BUhERLXTKvApAEDXNrZQNPFlFG7GUEX3hfOqiIioLi7nVsDCozNkEOjSpvmc+gMYqug+Vc+rik3ORmmFzsTdEBFRY7f9YtV0kTZWetiomv4yCjczaajav38/nn76abi7u0Mmk+Gnn34y2C6EwNy5c+Hm5gZLS0uEhITgwoULBjU5OTkYPXo01Go17OzsMH78eBQWFhrUnDx5Eg899BAsLCzg6emJRYsW3dbL5s2b4efnBwsLC3Tt2hW//fZbnXtpiXydbeBma4HSCj2XViAiorvKLSrHnyklAIAOrZrfvWNNGqqKiorQvXt3REVF1bh90aJF+Oyzz7B69WrExsbC2toaoaGhKC0tlWpGjx6NM2fOYNeuXdi6dSv279+PSZMmSdu1Wi2GDBmCtm3bIi4uDp9++inmz5+PL7/8Uqo5dOgQRo0ahfHjx+P48eMYPnw4hg8fjtOnT9epl5ZIJpPhMX9nAMCuxHQTd0NERI3ZxqOpKNcBZZqLcFA2v6vGTRqqhg4dig8++ADPPvvsbduEEFi2bBneffddPPPMM+jWrRv+85//4Pr169IRrcTEROzYsQNfffUVgoKCMHDgQKxYsQIbN27E9evXAQDr169HeXk5vvnmG3Tu3BkvvfQS3nzzTSxZskR6r+XLl+OJJ57A9OnT4e/vj/fffx+9evXCypUra91LS/Z4gCsAYPfZdOj1ze8/EiIiun+VOj3+G3MFAFAQ9wtkzWd+uqTRzqlKTk6GRqNBSEiI9JytrS2CgoIQExMDAIiJiYGdnR169+4t1YSEhEAulyM2Nlaqefjhh6FUKqWa0NBQJCUlITc3V6q5+X2qa6rfpza9tGT92jnARmWGjIIynLyWb+p2iIioEdp5Jh3X80uhVslRlPinqdtpEI02VGk0GgCAi4uLwfMuLi7SNo1GA2dnZ4PtZmZmcHBwMKipaYyb3+NONTdvv1cvNSkrK4NWqzV4NEcqM4V0FeDuszwFSEREt1t7KBkAENreCtA1z3vGNtpQ1RwsXLgQtra20sPT09PULTWYxwOqAucuhioiIrrF6Wv5OHolF2ZyWVWoaqYabahyda2ap5OebvhLOj09Xdrm6uqKjIwMg+2VlZXIyckxqKlpjJvf4041N2+/Vy81mT17NvLz86VHamrqPT510zWoU2so5DIkpRcgJbvY1O0QPRDX8kqw+2w6ov64iO9OF8B24Mu4WCDH9bwSlFc2vyubiOprzcErAIBh3dzgYKkwbTMNqNGGKh8fH7i6uiI6Olp6TqvVIjY2FsHBwQCA4OBg5OXlIS4uTqrZs2cP9Ho9goKCpJr9+/ejouJ/hxp37dqFTp06wd7eXqq5+X2qa6rfpza91ESlUkGtVhs8mis7KyX6eletrv772TufEiVq6rILy/Dv/Zfx5PI/MeDjPZjwn2P4dGcSNp8thN2Al3Ei1wyb467ii/2XsPXkdVzOLOQFHNSiZRaU4dcTVRePvdbf27TNNDCThqrCwkIkJCQgISEBQNWE8ISEBKSkpEAmk2Hq1Kn44IMP8Msvv+DUqVN49dVX4e7ujuHDhwMA/P398cQTT2DixIk4cuQIDh48iIiICLz00ktwd3cHALz88stQKpUYP348zpw5g++//x7Lly9HZGSk1MeUKVOwY8cOLF68GOfOncP8+fNx7NgxREREAECteiFgaNeqo3ZbT6aZuBMi4ysorcCSXefx0KI/8OFviTibpoWZXAY/11YY3sMdQztYoSB+G9wsqxY01AvgUmYRfj2ZhvVHUnA5s5A3HqcWaUNsCsp1evTwtENPL3tTt9OgTLqU6bFjxzB48GDp6+qgExYWhrVr12LGjBkoKirCpEmTkJeXh4EDB2LHjh2wsLCQXrN+/XpERETgscceg1wux8iRI/HZZ59J221tbfH7778jPDwcgYGBcHJywty5cw3Wsurfvz82bNiAd999F2+//TZ8fX3x008/oUuXLlJNbXpp6YZ2ccP8X84gITUPqTnF8HRovufNqWX5/YwGb/94GlmFZQCAzu5qvNTXC091dYO9ddWVxfHx8Vg9cRX6D38cHr6dkFVYhsQ0Lc5e1yKnqBy/nkyDp4MlQvxdoLYwN+XHIXpgyiv1+Db2LwDA2AHepm3mATBpqBo0aNBd/3KTyWRYsGABFixYcMcaBwcHbNiw4a7v061bN/z5590v33z++efx/PPP31cvLV3rVioEt3fEwYvZ+PXkdfxjUAdTt0Q3pKSkICsry+jjOjk5wcvLy+jjNhaFZZWY+9NpbDl+DQDQzska00M74YkurpDdY5EdJxsVHvJtjb7eDjj2Vy6Op+YhNacE62NTMLhja/i5Nd/pAETVfjuVhsyCMji3UmFoFzdTt9PgmtdNd8jknurmjoMXs7H1RBpDVSORkpICP39/lBQb/wICSysrnEtMbJbB6nJmISb9Nw4XMwohlwGTHm6PaY/7QmVWt0m2KnMFBnRwQoC7Gr+fSYdGW4qdZ6v+9yHfqgs8iJojIQTWHKxaRmFMv7ZQmjXaadxGw1BFRvVEZ1fM+ek0zqZpcSmzEO1b25i6pRYvKysLJcXFGD3zU7h4tTfauOkpl7D+k+nIyspqdqFq3/lMRKyPR0FZJVzUKkS93Au9b1yIUV/2Vko8H+iB2Cs5OJKcgxNX85FdVI5hXd1gYd58r4ailuvw5arvc5WZHKOCmtfPiDthqCKjsrdWYqCvE/YmZWLriTRMCfE1dUt0g4tXe3j4djZ1G43eD3FXMfOHk6jUC/Rua4/PX+kF51bGmTspl8sQ3M4Rzq1U2HlGg6u5Jfi/uKt4poc7WnGeFTUzq/ddAgA839sDTjYqE3fzYDT/Y3H0wD3drerKyx+PX+XVTtSkfLHvEv65+QQq9QLDe7hjw8R+RgtUN2vf2gYv9PaEtUqB7KJybI67ityicqO/D5GpnL2uxb7zmVWnzh8y3hHyxo6hiozuiS6usFYqcCW7GEeSc0zdDlGtrIi+gIXbzwEA/v5wOyx5oUeDzgFxslHhhUBP2FmZo6C0Ev8XfxU5DFbUTHyxv+oo1ZNd3eDl2HKuBOfpPzI6a5UZnu7ujo1HU/H9sVQEtXM0dUtEd7V89wUs3X0eADA9tBPCBz+YiyzUluZ4PtADPx6/hqzCcvwQfxUje3nAwVp57xc3AbzqtGVKyS6W1it8/ZGWc5QKYKiiBvJCH09sPJqK306lYf7fOnNdHmqUhBBYtvsClkdfAADMfMIPbwx6sL8ErJRmGNHTA1uOX21WwYpXnbZcK/+4AJ1e4OGOrdGlja2p23mgGKqoQfT0tIOvsw0uZBTil4TreKVfW1O3RGRACIElu85jxZ6LAIDZQ/3wdxP9VW2pVDS7YMWrTlumlOxi/BBfta7blMda3oVKDFXUIGQyGV7s44kPtiVi07FUhipqVIQQ+HRnEj7fWzXv491h/pjwUDuT9lRTsHqul4dJezIGXnXastx8lCqwbfO+JU1NOFGdGsyzPdtAqZDj5NV8xKfkmrodIgBVgerjHeekQDXnqQCTB6pq1cHKyUaJ4nIdfjh+FYUV934dUWPQ0o9SAQxV1IAcbVT4W4+q5RW+PpBs4m6IqgLVh9sS8cW+ywCA+U8HYPxAHxN3ZchSqcCzPdvA0VqJojId9meYw8zWxdRtEd3Tkl1J0OkFHvJ1apFHqQCGKmpg1b+wtp9Kw9Vc409YJaotvV5gzs+n8dWNgL/gmc54bUDjClTVrJRmeLZnGzhYKVGik8Fl1EJkFFWaui2iOzp9LR8/JVwHUHXBR0vFUEUNyt9NjYEdnKAXwLpDV0zdDrVQOr3AjB9O4tvDKZDJgI9HdMWrwd6mbuuurFVmGNGrDWzMBMxsnTF3bw6u5ZWYui2i2wgh8NFviQCA4T3cW9wVfzdjqKIGN/6hqqMBG4+koqCUE0TowarQ6TFl43H8X9xVKOQyLH2hB17q2zSuGrNWmeFh5wpU5FxHRpEOL6yOQXJWkanbIjKw73wmDl3KhlIhx1uhnUzdjkkxVFGDe8S3NTo426CgrBJrD14xdTvUgpRW6PDGt/HYejIN5goZol7uieE925i6rTqxNAPSN74NNxsFruWV4PnVMTh7XWvqtogAAOWVeizYehYA8NoAb3jYt5zV02vCUEUNTi6XYfKjVStUf/nnZeQX82gVNbzswjK8/O/D2J2YDpWZHF+O6Y0nuriZuq160RVk4cNHHRHgpkZWYRle/DIGx67wFlBkev/+8zIuZxbByUb1wO5E0JgxVNED8XQ3d/i5tkJBaaV0TyiihnIxoxAjVh1CfEoe1BZmWDeuLwb7OZu6rftiZ6HAd5P6oY+3PQpKK/HK17HYm5Rh6raoBUvNKcaKPVV3I3h3mD9sLXnnDIYqeiDkchkiH+8IAFhz8AoyC8pM3BE1V9tPpeGZlQfwV3YxPOwtseUf/dGvmdx/0tbSHP8ZF4RBnVqjtEKPCeuO4bsjKaZui1ogIQTe+/UMSiv06NfOAc/cWD6npeOK6vTAPB7ggu6edjiRmoelu8/jo2e7mrqlFksIgYLSSmQUlCG3uBzakgoUllWitEKP0goddEJALwRkkMFcIYPSTA4rpRmsVQo4WCnhaKNCuc7Un8JQaYUOn+w4hzU35u0F+Thg5cu90LqVyrSNGZmlUoEvx/TGW5tP4JcT1zF7yykkpmkx56kAmCv4dzI9GD/EX8PuxAyYK2R4/5kukMlkpm6pUWCoogdGJpPh7aF+ePHLw/juSAqeD/RAT6+WuUDcgyYzU0FTIkPyhUxkaMuQWViGskp9HUe59eiiEh4R32LBvmw8nn8R/do5oGsbOyjNHvwv9viUXLy1+QQuZ1ZdGTfp4XaYEdoJZs00ZCjN5Fj+Ug90dLHBv34/j//E/IUTV/Ox/MUe8HayNnV71MxdzS3G/F/OAACmPd4Rvi6tTNxR48FQRQ9UUDtHjOzlgR/ir+KdH0/jl4gBzfYXnykJIXBOU4D95zOxLT4bnlO+w8FMcwB5Uo1cVrXqvaO1EmpLc7RSmcFSqYCFmQIKhQxyGSBE1ZIE5ZV6FJXpUFBWgZyicmQVliO/pBwKazskpJcjYWcSAMDSXIHe3vYY0MEJD/k6wd9VDbm84f6CTdeWYtGOJPwQfxUA4NxKhU+e64bBnZr2/KnakMlkiHjUF51c1fjnpgScSM3DsM/+xLtPBeDF3p4Nut+p5dLrBd7afAKFZZUIbGuPvz9smpuQN1YMVfTAvf2kH3YnpuNsmhZrD11pNPdda+pyisrx54VM7D+fhT8vZCLjpnlrMjMlrBQC7Vxs4WZridatVHCwVkJxH794rySdweoPZ2DOsm9wrcIKhy9nI7e4An9eyMKfF7Lw8XbAyUaFh3yrAtZAXyc4t7IwxkdFSnYxvjpwGZuOpaK0ouqI28heHpjzlD/srJRGeY+m4vEAF+yY+jCmfp+AI8k5mL3lFDYfS8X7w7ugs7tpFmHML67AqfQyqPuOQHyOAsdPXEdReSWKy3UoLtdBCAEAEDfqzeQyWCnNYKVU3HiYwd7KXAr9VkoFTy81Est2n8fhyzmwUiqw5IXu9/UzpDliqKIHztFGhVlD/TB7yyl8ujMJAzo4wd9Nbeq2mpwKnR7xf+Vi/40gdfp6PoT433ZLcwX6tXNAO8syLAgfhckLVsCzo/HuIWcmB8o1F/GkrzV69eoFvV7gQkYhDl3KwoELWYi5nI2swjL8ePwafjxedZNVP9dWCGxrjx6edujpZY92Tta1PqKSoS3FnnMZ2HoyDYcuZUF/47MGtrXHnKcC0MPTzmifralxt7PEdxP7Yc3BZCzddR7xKXl4asUBPNnVDVMe80XHBjw9k1dcjtPXtDh1LR+nr+Xj1LV8pORU3ZLKfvA4JBcCKLz7gqUVOoH8kgrkl9S83IqFmRxONiq42lrAvFgGuSV/XpjCzjMafLbnIgDgw2e7oK0jTzXfiqGKTOLF3p74/YwGfyRlInxDPH6NGAhrVcv+dkxJSUFWVtYdtwshkF6kQ4KmDAmaMpzKKEdJpTCo8bYzQw8XFXq4quDvpIS5QobExAuozLmGhv5DXy6XoZNrK3RybYWxA3xQVqlD/F952H8hE39eyMTpa1qc0xTgnKYA62OrrlhrZWEGf1c1PB2s0MbOAmpLc1irzFCp06OsUg9Nfimu5pbg9PV8XM01vEXLIx1b4+8Pt0Nwe0cexQCgkMsw4aF2GNbNDR9uS8TWk2nYduMR3M4RL/bxREiAC2zu47+z3KJynLoRnKoD1K3/v1RzsVbgcvx+BPbpBzc3V1irzKSjUNU5WgYZIKv6A6HqKFbV0ayiskrkFJUju6gc+cUVKK3U42peCa7mlQAwh+ebGxD+WwYGXjqBvj4O6NfOER72lvw+aEAX0gvwz00nAABjB3jj2Z4eJu6ocWrZv8XIZORyGRa/0ANPLv8TlzOL8O5Pp7Hkhe4t9odiSkoK/Pz9UVJseNNphY0DLLy6waJtd1i07QYzW8MjTbrifJQmH0dJcjxKrxzHX0W52HeH9ygsLGyg7mumMlMguL0jgts7YuYTfsgqLEPs5RwkpOYiITUPp67lo6C0Ekeu5OBILRaylMuAzu62GNrVFU91dYeXY8teuflO3GwtsfLlXggfrMXy3Rew86wGMZezEXM5G2ZyGXp726OvtwP83NRo19oaTjYq2FmaQy6TQQAoKK1AXnEFruaWIDmrEJezipCcVYQL6YV3vPdgW0crdGljiy7utujaxhZd2qhx+dxpBM5diM6Pb4GHh909+77TQtyVOj1yiyuQXlCKtLxSpGbmo6BShrRCHTbHXcXmuKs3PrcF+vo4oK+PA4J8HNC+tU2L/XlibKk5xRjz9REUllUiyMcBbz/pb+qWGi2GKjIZB2slPhvVEy99GYMfj1+Dl4MVpt1Yy6qlycrKQklxMZ6buRRyJ29klsqRWSpHQaXhLwUZBBxVAi4WerhYCtiZW0Lm1x9A/zuOnXhkH7avW47S0tIG/hR352SjwrBubhjWrWpV8wqdHufTC3Apswgp2UXQaEtRUFqJojKdtIxDaxsVPOwt0d7ZBj087dDKgosL1pa/mxqrxwTiWl4JNh9LxU/Hr+FKdjEOX87B4cv1X43dx8kaXdrYomsbNbq426JzG9sGXfTRTCFH61YqtG6lQhd3W1y9kIVl/3wVqzbvRI7CAUeSs3Hyaj7S8kvxc8J1/JxwHQDgaK1EH28HBLWrClp+rmrO/6mHzIIyjPk6FhptKXydbbDqlUAu3XEXDFVkUn19HPDeM10w56fTWB59AQ7WSoT19zZ1Ww9EeaUeZ9O0iP8rF9EnctHmjW9wFM7ALWcAnVup4GlvBU8HS7jbWdb5B1p6SuNcwd5cIUdnd1uTTaZuKdrYWWJqSEdMDemIK1lF0qnYRI0WV3NLkFtcbjAXDwCslAq42VrAx8kG7Vpbw8fJGu2crOHvroa6EQRbfWkhertboFcvPwBAcXklElLyEJucg9jkbBxPyUN2UTl2nNFgxxkNgKpTzX28HaSjWV3b2Nb439K9TsPXl5OTE7y8msaNvKtdzS1G2DdHcCW7GG3sLPHf8UFwsG5ZF4LUFUMVmdyYfm2RXViGZbsvYP6vZyCEwGsDfEzdllEJIXAtrwSnr+XjeEoe4lNycfJqvsFaUWZqZwACTjYquNtawtPBCh72lrAwV5iucWpWvJ2sb1vHqlKnR2FZpRSsrFVmJllr7H5YKc3Qv4MT+ndwAgCUVepw6mo+YpNzcCQ5B3F/5aKgtBJ7zmVgz7mqW/tYmivQq60d+no7oqeXHTq7q1GUk17jaXhjsLSywrnExCYTrM5ptAj75gjStWVws7XAtxOC4GprnKt3mzOGKmoUpjzmi/ySCqw5eAXzfz2Lq7klePtJ/ya51o4QAldzSwwm9J6+lo/cGm4kbWdljp6ednAzL8HSd97ExOnz4dOpZZ4CJdMwU8ib3TIUKjMFens7oLe3A8IHVwXHxLQCxCZn40hy1Ry+vOIKHLyYjYMXs6XXOVnJYfNEJHr7tkcbJzvYKQUsFbjvizzSUy5h/SfTkZWV1SRC1S8nruPtLadQWFYJX2cbrBvXF+52lqZuq0lgqKJGQSaTYe5TAWjdSoVFO5Lw1YFkJGq0WPRcd7RpxP8xCyGQklNscEn56ev5yKshQJnduDqum4ctennZo1fbqiUFZDIZ4uPjsTDlJMyb1gECoibBTCFHVw9bdPWwxYSH2knLfxxJzkZscg5OX8vHlexiZBXrYeXbDykAUm6cAVSZyeForYSDjRKO1lXrZjnaKGGlbH6/PrWlFfhoWyI2Hk0FUHWrpy/GBDa70N2Qmt93BTVZMpkM/xjUAW3sLDHzh5M4eDEboUv3Y8YTnfBSHy+Tn5Ior9TjYkYhEtO0OKfR4sx1LU5fy4e2tPK2WnNFVYCquhKq6oqoTq6toDLjqbza4twWaig3L/8xJtgbQNVVjz/tO4bwdz5Gj2fGowiWyCkuR1mlHtfzS3E93/BCD0tzBRyslXC0VsLeWgl7K3PYWynRysKsyV11KITAzwnX8eFvicgsKINMBkwe3AFvPubLO17UEUMVNTrP9GiDbh52+OemBMSn5GHuz2fw5f7LiBjcAU93d2/w9ayEEMgsLMO5tAIkpmlvhKgCXMwoRKVe3FavVMjh59bK4JLyjq42DFD34U5LTBhDU5vbQg9GKwtzdG6tQkHcL+gz7jV4+LaVlnPILiqrWjer8MbaWSUVKKnQ4VpeyW3LTCjkMthZVgUsOytz2FsrUVEmg9zCxkSf7M50eoFtp9Lw+R8XcU5TAKDq6s4Pn+2C/u2dTNxd08RQRY2Sj5M1Nr/eH+tj/8KKPRdxNbcEs7acwoKtZzG0ixsGdWqN/u0d4Wijqvd7lFbocCW7CJczq9bhuZRZKP37Tis7t7IwQzsHFdpYCbS1M0d7e3N4qs1grpABqACQhYqMLJzJqFsviYmJ9f4czVH1EhOjZ34KFy/j3Vusqc1teRAa6ohgc/ievnk5h5tV6PTIvbE4aXZROfKKy5FbXIH84gro9EJ6/n/M4TllI1754Ro8o7Pg3soMbjZmcLZWSA87Cznk9TzCVZejr3q9wNk0LX49cR0/JVxDurbqdlbWSgX+MbgDJjzkwz8I7wNDFTVaCrkMrwZ74/lAT3x7+C+sj/0LV7KL8UP8VekGum3sLOHrYgMvBys42ahgb2UOpZkcZnI5Sit1KCnXobSi6n5jucXl0OSXIl1bhnRt6S0/9AzJZICPozX83dTwc20Ffzc1/N3VqMzPgH9AQIMcQQEe/AKdjZ2LV3t4+HY2dRvNVkMeEazWHL+nzRVyOKst4Kw2vBpOLwQKSiuRW1yO3KJy5BVXILe4HJl5BSgVZijWyZGUXYGk7Nv/aBOVFajUZqJSm47K/Azo8jNQqc2ErigXuqI86IvzoCvOB4T+ttfe6ehrpU6P9IIyXMkqwpnr+Th5NR8xl7INfvbZW5lj7AAfhAV7w9bK9MtlNHUMVXUUFRWFTz/9FBqNBt27d8eKFSvQt29fU7fVrFkqFZj4cDtMeMgHx/7KxY7TGhy8mIVzmoIaD7/XhdrCDO1a26Cdk/WN9Xhs4ONUtS6PpfL2v9biL2c3yBGUxrJAJ7UsDXVEEGiZ39NymQy2luawtTSH9033xYuL/gUblryLwRPmwsmnMwoqgcJKGYpvPEp0AMzMYe7gDnMH97u8g4C5rOq+m2YywEwuoC8vQXrKZby+KQmW1lehFwJFZZXIK65ATlF5jVMWrJUKPOTbGs/2aoPBnZxNPl+1OWGoqoPvv/8ekZGRWL16NYKCgrBs2TKEhoYiKSkJzs7Opm6v2ZPJZOjj7YA+3g4Aqm7kej69EOfTC6DJL0VWYRnySypQXqlHhV7AwkwOK6UClkoFLMwVsLU0h6vaAi62FnBpZQFXWwvYW5nXa1KpsY+gNNYFOqllaIgjgvyeNiQqyuDh2ho9etx+ixe9XqCwrBLa0gpoSytRUHLjf0srUFyhQ3FZ1RF3ARkqBFChq36lDIA1LLy64mxWOZB1+9F3c4UMHvZW8HNthc7uavTxdkBPL3sGqQbCUFUHS5YswcSJEzF27FgAwOrVq7Ft2zZ88803mDVrlom7a3nsrJTS6si3+t88EQGg8sajDEAhUASUFgFXNMCVOr5nc5gnYmwNsU8aej+z54Yb80FpTvtDLpdBbWkO9V1u96MXAiXlOpTr9CivrHpU6PRIv5aKbd/8C29OngxPDw/IZYDKTIZWSjnUKjnsLeQ33Z6nAMgrwOm8v2rVF6+UrTuGqloqLy9HXFwcZs+eLT0nl8sREhKCmJgYE3b2YDTUD5uysjKoVPWfbF6TtLQ0PPf88ygtqf9pwXtpjvNE6kqbkwkAeOWVVxrsPYy9n9lzzZrS93NL3R9ymQzWKjNY3/J86eV0FJ87gI/DDxj9PVUWFvjh//4Pbm5uRh23KYf5e2GoqqWsrCzodDq4uLgYPO/i4oJz587V+JqysjKUlZVJX+fn5wMAtFqtUXur/gFw9cIZlJUYd8LplcTjABr2B1hDCf7bK3D18DbqmCnnTyFu98+4cu6kUW8qWn2qRHPlPC5ZWxlt3IYc+8rZqu+NPkNfhIePr9HGBRpuP7NnQw3VM8Dvu5s15H/fDbU/0q6cx+HfNuGpp54y2pi3On/ymFF/Z2VeTQZQ9TvR2L9nq8cTt94o81aCauXatWsCgDh06JDB89OnTxd9+/at8TXz5s0TqDr/xAcffPDBBx98NPFHamrqXbMCj1TVkpOTExQKBdLT0w2eT09Ph6ura42vmT17NiIjI6Wv9Xo9cnJy4OjoaNQVd7VaLTw9PZGamgq1Wm20cel23NcPBvfzg8H9/GBwPz8YDbmfhRAoKCiAu/vdrs7k6b9aUyqVCAwMRHR0NIYPHw6gKiRFR0cjIiKixteoVKrb5gvZ2dk1WI9qtZr/wT4g3NcPBvfzg8H9/GBwPz8YDbWfbW1t71nDUFUHkZGRCAsLQ+/evdG3b18sW7YMRUVF0tWARERE1HIxVNXBiy++iMzMTMydOxcajQY9evTAjh07bpu8TkRERC0PQ1UdRURE3PF0n6moVCrMmzfP6EsT0O24rx8M7ucHg/v5weB+fjAaw36WCXGv6wOJiIiI6F64Tj0RERGRETBUERERERkBQxURERGRETBUERERERkBQ1UzEBUVBW9vb1hYWCAoKAhHjhwxdUtN2sKFC9GnTx+0atUKzs7OGD58OJKSkgxqSktLER4eDkdHR9jY2GDkyJG3rbZPdfPxxx9DJpNh6tSp0nPcz8Zx7do1vPLKK3B0dISlpSW6du2KY8eOSduFEJg7dy7c3NxgaWmJkJAQXLhwwYQdNz06nQ5z5syBj48PLC0t0b59e7z//vsG94rjfq6f/fv34+mnn4a7uztkMhl++ukng+212a85OTkYPXo01Go17OzsMH78+Aa5cTZDVRP3/fffIzIyEvPmzUN8fDy6d++O0NBQZGRkmLq1Jmvfvn0IDw/H4cOHsWvXLlRUVGDIkCEoKiqSaqZNm4Zff/0Vmzdvxr59+3D9+nWMGDHChF03bUePHsUXX3yBbt26GTzP/Xz/cnNzMWDAAJibm2P79u04e/YsFi9eDHt7e6lm0aJF+Oyzz7B69WrExsbC2toaoaGhKC0tNWHnTcsnn3yCVatWYeXKlUhMTMQnn3yCRYsWYcWKFVIN93P9FBUVoXv37oiKiqpxe2326+jRo3HmzBns2rULW7duxf79+zFp0iTjN3v/txomU+rbt68IDw+XvtbpdMLd3V0sXLjQhF01LxkZGQKA2LdvnxBCiLy8PGFubi42b94s1SQmJgoAIiYmxlRtNlkFBQXC19dX7Nq1SzzyyCNiypQpQgjuZ2OZOXOmGDhw4B236/V64erqKj799FPpuby8PKFSqcR33333IFpsFoYNGybGjRtn8NyIESPE6NGjhRDcz8YCQPz444/S17XZr2fPnhUAxNGjR6Wa7du3C5lMJq5du2bU/nikqgkrLy9HXFwcQkJCpOfkcjlCQkIQExNjws6al/z8fACAg4MDACAuLg4VFRUG+93Pzw9eXl7c7/UQHh6OYcOGGexPgPvZWH755Rf07t0bzz//PJydndGzZ0/8+9//lrYnJydDo9EY7GdbW1sEBQVxP9dB//79ER0djfPnzwMATpw4gQMHDmDo0KEAuJ8bSm32a0xMDOzs7NC7d2+pJiQkBHK5HLGxsUbthyuqN2FZWVnQ6XS33SbHxcUF586dM1FXzYter8fUqVMxYMAAdOnSBQCg0WigVCpvuzm2i4sLNBqNCbpsujZu3Ij4+HgcPXr0tm3cz8Zx+fJlrFq1CpGRkXj77bdx9OhRvPnmm1AqlQgLC5P2ZU0/R7ifa2/WrFnQarXw8/ODQqGATqfDhx9+iNGjRwMA93MDqc1+1Wg0cHZ2NthuZmYGBwcHo+97hiqiuwgPD8fp06dx4MABU7fS7KSmpmLKlCnYtWsXLCwsTN1Os6XX69G7d2989NFHAICePXvi9OnTWL16NcLCwkzcXfOxadMmrF+/Hhs2bEDnzp2RkJCAqVOnwt3dnfu5BeHpvybMyckJCoXitquh0tPT4erqaqKumo+IiAhs3boVf/zxBzw8PKTnXV1dUV5ejry8PIN67ve6iYuLQ0ZGBnr16gUzMzOYmZlh3759+Oyzz2BmZgYXFxfuZyNwc3NDQECAwXP+/v5ISUkBAGlf8ufI/Zk+fTpmzZqFl156CV27dsWYMWMwbdo0LFy4EAD3c0OpzX51dXW97eKtyspK5OTkGH3fM1Q1YUqlEoGBgYiOjpae0+v1iI6ORnBwsAk7a9qEEIiIiMCPP/6IPXv2wMfHx2B7YGAgzM3NDfZ7UlISUlJSuN/r4LHHHsOpU6eQkJAgPXr37o3Ro0dL/+Z+vn8DBgy4bUmQ8+fPo23btgAAHx8fuLq6GuxnrVaL2NhY7uc6KC4uhlxu+CtVoVBAr9cD4H5uKLXZr8HBwcjLy0NcXJxUs2fPHuj1egQFBRm3IaNOe6cHbuPGjUKlUom1a9eKs2fPikmTJgk7Ozuh0WhM3VqT9cYbbwhbW1uxd+9ekZaWJj2Ki4ulmtdff114eXmJPXv2iGPHjong4GARHBxswq6bh5uv/hOC+9kYjhw5IszMzMSHH34oLly4INavXy+srKzEt99+K9V8/PHHws7OTvz888/i5MmT4plnnhE+Pj6ipKTEhJ03LWFhYaJNmzZi69atIjk5WWzZskU4OTmJGTNmSDXcz/VTUFAgjh8/Lo4fPy4AiCVLlojjx4+Lv/76SwhRu/36xBNPiJ49e4rY2Fhx4MAB4evrK0aNGmX0XhmqmoEVK1YILy8voVQqRd++fcXhw4dN3VKTBqDGx5o1a6SakpIS8Y9//EPY29sLKysr8eyzz4q0tDTTNd1M3BqquJ+N49dffxVdunQRKpVK+Pn5iS+//NJgu16vF3PmzBEuLi5CpVKJxx57TCQlJZmo26ZJq9WKKVOmCC8vL2FhYSHatWsn3nnnHVFWVibVcD/Xzx9//FHjz+SwsDAhRO32a3Z2thg1apSwsbERarVajB07VhQUFBi9V5kQNy33SkRERET1wjlVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVERERkBAxVREREREbAUEVEdAfz58+Hi4sLZDIZfvrpJ1O3g0GDBmHq1KkmH8OY4xA1JwxVRNSsGCsAJSYm4r333sMXX3yBtLQ0DB069P6bq6W9e/dCJpPddjNpU7hTL1u2bMH7779vmqaIGikzUzdARM2fTqeDTCa77YazjdmlS5cAAM888wxkMpmJu2l8HBwcTN0CUaPTdH7CEdEDMWjQIERERCAiIgK2trZwcnLCnDlzcPMdrcrKyvDWW2+hTZs2sLa2RlBQEPbu3SttX7t2Lezs7PDLL78gICAAKpUKKSkpKCsrw8yZM+Hp6QmVSoUOHTrg66+/ll53+vRpDB06FDY2NnBxccGYMWOQlZVl0Nubb76JGTNmwMHBAa6urpg/f7603dvbGwDw7LPPQiaTSV/X5NSpU3j00UdhaWkJR0dHTJo0CYWFhQCqTvs9/fTTAAC5XH7HUFV9FGfnzp3o2bMnLC0t8eijjyIjIwPbt2+Hv78/1Go1Xn75ZRQXF0uv0+v1WLhwIXx8fGBpaYnu3bvj//7v/wAAV65cweDBgwEA9vb2kMlkeO211wxee6fPP27cODz11FMGPVZUVMDZ2dlgP9/sv//9L3r37o1WrVrB1dUVL7/8MjIyMu7Zy62n/3Jzc/Hqq6/C3t4eVlZWGDp0KC5cuCBtr/6e2LlzJ/z9/WFjY4MnnngCaWlpNfZF1CQZ/W6CRNSkPfLII8LGxkZMmTJFnDt3Tnz77bfCysrK4Ca8EyZMEP379xf79+8XFy9eFJ9++qlQqVTi/PnzQggh1qxZI8zNzUX//v3FwYMHxblz50RRUZF44YUXhKenp9iyZYu4dOmS2L17t9i4caMQQojc3FzRunVrMXv2bJGYmCji4+PF448/LgYPHmzQm1qtFvPnzxfnz58X69atEzKZTPz+++9CCCEyMjKkm1+npaWJjIyMGj9jYWGhcHNzEyNGjBCnTp0S0dHRwsfHR7pBa0FBgVizZo0AINLS0u54E+fqG73269dPHDhwQMTHx4sOHTqIRx55RAwZMkTEx8eL/fv3C0dHR/Hxxx9Lr/vggw+En5+f2LFjh7h06ZJYs2aNUKlUYu/evaKyslL88MMPAoBISkoSaWlpIi8vr1af/+DBg0KhUIjr169L77VlyxZhbW0t3Tz21ptWf/311+K3334Tly5dEjExMSI4OFgMHTpUCCHu2cvN4/ztb38T/v7+Yv/+/SIhIUGEhoaKDh06iPLycoPviZCQEHH06FERFxcn/P39xcsvv1zjviVqihiqiMjAI488Ivz9/YVer5eemzlzpvD39xdCCPHXX38JhUIhrl27ZvC6xx57TMyePVsIIaRAkpCQIG1PSkoSAMSuXbtqfN/3339fDBkyxOC51NRU6Rd6dW8DBw40qOnTp4+YOXOm9DUA8eOPP971M3755ZfC3t5eFBYWSs9t27ZNyOVyodFohBBC/Pjjj+Jef3dWh6rdu3dLzy1cuFAAEJcuXZKe+/vf/y5CQ0OFEEKUlpYKKysrcejQIYOxxo8fL0aNGmUwbm5urkFNbT5/QECA+OSTT6Svn376afHaa68ZjHFzGLrV0aNHBQAphN2tl+pxzp8/LwCIgwcPStuzsrKEpaWl2LRpkxDif98TFy9elGqioqKEi4vLHXshamp4+o+IbtOvXz+DU17BwcG4cOECdDodTp06BZ1Oh44dO8LGxkZ67Nu3T5qHBABKpRLdunWTvk5ISIBCocAjjzxS43ueOHECf/zxh8GYfn5+AGAw7s1jAoCbm5t0uqq2EhMT0b17d1hbW0vPDRgwAHq9HklJSXUa69aeXFxcYGVlhXbt2hk8V93jxYsXUVxcjMcff9zgs/7nP/8x+Jy1eS/g9s8/YcIErFmzBgCQnp6O7du3Y9y4cXccLy4uDk8//TS8vLzQqlUr6f+flJSUWnzyKomJiTAzM0NQUJD0nKOjIzp16oTExETpOSsrK7Rv3/6OvRM1dZyoTkR1UlhYCIVCgbi4OCgUCoNtNjY20r8tLS0NgpmlpeU9x3366afxySef3LbNzc1N+re5ubnBNplMBr1eX6fPYGw39ySTye7aY/W8rW3btqFNmzYGdSqVqk7vdevYAPDqq69i1qxZiImJwaFDh+Dj44OHHnqoxrGKiooQGhqK0NBQrF+/Hq1bt0ZKSgpCQ0NRXl5+z17qqqbexU1z9YiaOoYqIrpNbGyswdeHDx+Gr68vFAoFevbsCZ1Oh4yMjDv+sq5J165dodfrsW/fPoSEhNy2vVevXvjhhx/g7e0NM7P6/2gyNzeHTqe7a42/vz/Wrl2LoqIi6WjVwYMHIZfL0alTp3q/d23cPHH/TkftlEolANzzc9TE0dERw4cPx5o1axATE4OxY8fesfbcuXPIzs7Gxx9/DE9PTwDAsWPH6tyLv78/KisrERsbi/79+wMAsrOzkZSUhICAgDp/BqKmiqf/iOg2KSkpiIyMRFJSEr777jusWLECU6ZMAQB07NgRo0ePxquvvootW7YgOTkZR44cwcKFC7Ft27Y7junt7Y2wsDCMGzcOP/30E5KTk7F3715s2rQJABAeHo6cnByMGjUKR48exaVLl7Bz506MHTu2TuHC29sb0dHR0Gg0yM3NrbFm9OjRsLCwQFhYGE6fPo0//vgDkydPxpgxY+Di4lKHPVV3rVq1wltvvYVp06Zh3bp1uHTpEuLj47FixQqsW7cOANC2bVvIZDJs3boVmZmZ0tGt2powYQLWrVuHxMREhIWF3bHOy8sLSqUSK1aswOXLl/HLL7/ctvZUbXrx9fXFM888g4kTJ+LAgQM4ceIEXnnlFbRp0wbPPPNMnXonasoYqojoNq+++ipKSkrQt29fhIeHY8qUKZg0aZK0fc2aNXj11Vfxz3/+E506dcLw4cNx9OhReHl53XXcVatW4bnnnsM//vEP+Pn5YeLEiSgqKgIAuLu74+DBg9DpdBgyZAi6du2KqVOnws7Ork7rWy1evBi7du2Cp6cnevbsWWONlZUVdu7ciZycHPTp0wfPPfccHnvsMaxcubLW73M/3n//fcyZMwcLFy6Ev78/nnjiCWzbtg0+Pj4AgDZt2uC9997DrFmz4OLigoiIiDqNHxISAjc3N4SGhsLd3f2Oda1bt8batWuxefNmBAQE4OOPP8a//vUvg5ra9rJmzRoEBgbiqaeeQnBwMIQQ+O2332475UfUnMkET2gT0U0GDRqEHj16YNmyZaZuheqpsLAQbdq0wZo1azBixAhTt0PUYnBOFRFRM6HX65GVlYXFixfDzs4Of/vb30zdElGLwlBFRNRMpKSkwMfHBx4eHli7du19Tfgnorrj6T8iIiIiI+BEdSIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIjYKgiIiIiMgKGKiIiIiIj+H/WNUFefZRIPwAAAABJRU5ErkJggg==\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "sns.histplot(data = epiblast, x = 'percent of methylation', bins = 20, kde = True, kde_kws = {'bw_adjust': 2})\n",
        "plt.title('epiblast')\n",
        "plt.savefig('epiblast.png')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 472
        },
        "id": "CwhZBbQnXKJh",
        "outputId": "05201093-85a7-4c82-c34a-655efdfbabd0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjcAAAHHCAYAAABDUnkqAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAUvtJREFUeJzt3XlcVPX+P/DXmRWGfR0WQRBRcSdNxRa1SLKybPt1y9I0bZMyvW1WLtUt7FuafcvyW7e07s2rLeqtNM1IKnMHzQ1QcRlFtpGdgQFmPr8/kCkSFHBgZg6v5+MxD5wzn3PmPSdyXn4+n3M+khBCgIiIiEgmFI4ugIiIiMieGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYbojI6T344IOIioqyPT958iQkScJbb711yX0XLFgASZI6sDoicjYMN0RE7WQymbBgwQKkpaU5uhQi+hOVowsgIrqUjz76CFar1dFlXMBkMuHll18GAIwePdqxxRCRDcMNETk9tVrt6BKIyIVwWIqIOkRubi6mTp0KvV4PrVaLfv364ZNPPrG9npaWBkmSsHr1arzwwgsICQmBh4cHbr31Vpw+fbrJsf465+bP3n77bXTv3h3u7u4YNWoUDh48eMnali9fjuuuuw7BwcHQarXo27cvPvjggwva7dmzB0lJSQgMDIS7uzuio6MxdepUAA3zfoKCggAAL7/8MiRJgiRJWLBgQSvPEBF1FPbcEJHdFRQUYMSIEZAkCcnJyQgKCsL333+Phx56COXl5XjqqadsbV977TVIkoTnnnsOhYWFWLJkCRITE7Fv3z64u7tf9H0+++wzVFRUYMaMGaipqcE777yD6667DgcOHIBer29xvw8++AD9+vXDrbfeCpVKhW+//RaPP/44rFYrZsyYAQAoLCzE2LFjERQUhOeffx6+vr44efIk1qxZAwAICgrCBx98gMceewy333477rjjDgDAwIEDL/PsEdFlE0REdvbQQw+J0NBQYTQam2z/29/+Jnx8fITJZBJbtmwRAER4eLgoLy+3tfniiy8EAPHOO+/Ytk2ePFl0797d9vzEiRMCgHB3dxdnzpyxbd+5c6cAIGbNmmXbNn/+fPHXv+pMJtMFNSclJYkePXrYnq9du1YAELt3727xcxYVFQkAYv78+S2fDCLqdByWIiK7EkLg66+/xvjx4yGEgNFotD2SkpJQVlaGjIwMW/tJkybBy8vL9vyuu+5CaGgoNmzYcMn3mjBhAsLDw23Phw0bhuHDh19y3z/3CJWVlcFoNGLUqFE4fvw4ysrKAAC+vr4AgO+++w51dXWt+uxE5By6dLj55ZdfMH78eISFhUGSJKxbt67NxxBC4K233kKvXr2g1WoRHh6O1157zf7FErmIoqIilJaW4sMPP0RQUFCTx5QpUwA0DPk0io2NbbK/JEno2bMnTp48ecn3+uu+ANCrV69L7vvbb78hMTERHh4e8PX1RVBQEF544QUAsIWbUaNG4c4778TLL7+MwMBA3HbbbVi+fDnMZvMl6yIix+rSc26qqqowaNAgTJ061TZe3lYzZ87EDz/8gLfeegsDBgxAcXExiouL7VwpketovGT7/vvvx+TJk5ttM3DgQBw+fLgzy7LJycnB9ddfjz59+mDx4sWIiIiARqPBhg0b8Pbbb9vqlyQJX331FXbs2IFvv/0WmzZtwtSpU7Fo0SLs2LEDnp6eDqmfiC6tS4ebcePGYdy4cS2+bjab8eKLL+I///kPSktL0b9/f7zxxhu2+1lkZmbigw8+wMGDB9G7d28AQHR0dGeUTuS0goKC4OXlBYvFgsTExBbbNYabo0ePNtkuhMCxY8daNTH3r/sCwJEjR1q8sgoAvv32W5jNZnzzzTeIjIy0bd+yZUuz7UeMGIERI0bgtddew8qVKzFx4kSsWrUK06ZN452PiZxUlx6WupTk5GRs374dq1atwv79+3H33XfjxhtvtP2F+u2336JHjx747rvvEB0djaioKEybNo09N9SlKZVK3Hnnnfj666+bvSy7qKioyfPGK54affXVV8jLy7voPzwarVu3Drm5ubbnu3btws6dOy+6r1KpBNAQohqVlZVh+fLlTdqVlJQ0aQMAgwcPBgDb0JROpwMAlJaWXrJWIuo8Xbrn5mIMBgOWL18Og8GAsLAwAMDTTz+NjRs3Yvny5Xj99ddx/PhxnDp1Cl9++SU+++wzWCwWzJo1C3fddRd++uknB38CIsdZuHAhtmzZguHDh2P69Ono27cviouLkZGRgR9//LHJPwD8/f1x9dVXY8qUKSgoKMCSJUvQs2dPTJ8+/ZLv07NnT1x99dV47LHHYDabsWTJEgQEBODZZ59tcZ+xY8dCo9Fg/PjxeOSRR1BZWYmPPvoIwcHByMvLs7X79NNP8f777+P2229HTEwMKioq8NFHH8Hb2xs33XQTgIaJyX379sXq1avRq1cv+Pv7o3///ujfv/9lnD0iulwMNy04cOAALBYLevXq1WS72WxGQEAAgIa5BWazGZ999pmt3ccff4whQ4YgOzvbNlRF1NXo9Xrs2rULr7zyCtasWYP3338fAQEB6NevH954440mbV944QXs378fKSkpqKiowPXXX4/333/f1ityMZMmTYJCocCSJUtQWFiIYcOG4b333kNoaGiL+/Tu3RtfffUVXnrpJTz99NMICQnBY489hqCgINsN+oCGCcW7du3CqlWrUFBQAB8fHwwbNgyff/55k+Hnf/7zn3jiiScwa9Ys1NbWYv78+Qw3RA4mib/2u3ZRkiRh7dq1mDBhAgBg9erVmDhxIg4dOmTrxm7k6emJkJAQzJ8/H6+//nqTy0Srq6uh0+nwww8/4IYbbujMj0DkUtLS0jBmzBh8+eWXuOuuuxxdDhHJCHtuWhAfHw+LxYLCwkJcc801zba56qqrUF9fj5ycHMTExABomMwIAN27d++0WomIiOgPXTrcVFZW4tixY7bnJ06cwL59++Dv749evXph4sSJmDRpEhYtWoT4+HgUFRUhNTUVAwcOxM0334zExERcccUVmDp1KpYsWWK7dfsNN9xwwXAWERERdY4ufbXUnj17EB8fj/j4eADA7NmzER8fj3nz5gFoWFxv0qRJ+Pvf/47evXtjwoQJ2L17t+3yUYVCgW+//RaBgYG49tprcfPNNyMuLg6rVq1y2GciIiLq6jjnhoiIiGSlS/fcEBERkfww3BAREZGsdLkJxVarFWfPnoWXlxdvnU5EROQihBCoqKhAWFgYFIqL9810uXBz9uxZREREOLoMIiIiaofTp0+jW7duF23T5cKNl5cXgIaT4+3t7eBqiIiIqDXKy8sRERFh+x6/mC4XbhqHory9vRluiIiIXExrppRwQjERERHJCsMNERERyQrDDREREckKww0RERHJCsMNERERyQrDDREREckKww0RERHJCsMNERERyQrDDREREckKww0RERHJCsMNERERyQrDDREREckKww0RERHJCsMNERERyYrK0QUQERFR5zMYDDAajR1y7MDAQERGRnbIsVuD4YaIiKiLMRgM6BMXh2qTqUOO767TISsz02EBh+GGiIioizEajag2mTDxuTehj4yx67ELDDn4/I1nYDQaGW6IiIioc+kjY9Attp+jy7A7TigmIiIiWWG4ISIiIllhuCEiIiJZYbghIiIiWWG4ISIiIllhuCEiIiJZYbghIiIiWWG4ISIiIllxaLj55ZdfMH78eISFhUGSJKxbt+6S+6SlpeGKK66AVqtFz549sWLFig6vk4iIiFyHQ8NNVVUVBg0ahKVLl7aq/YkTJ3DzzTdjzJgx2LdvH5566ilMmzYNmzZt6uBKiYiIyFU4dPmFcePGYdy4ca1uv2zZMkRHR2PRokUAgLi4OGzduhVvv/02kpKSOqpMIiIiciEuNedm+/btSExMbLItKSkJ27dvd1BFRERE5GxcauHM/Px86PX6Jtv0ej3Ky8tRXV0Nd3f3C/Yxm80wm8225+Xl5R1eJxERETmOS/XctEdKSgp8fHxsj4iICEeXRERERB3IpcJNSEgICgoKmmwrKCiAt7d3s702ADBnzhyUlZXZHqdPn+6MUomIiMhBXGpYKiEhARs2bGiybfPmzUhISGhxH61WC61W29GlERERkZNwaM9NZWUl9u3bh3379gFouNR73759MBgMABp6XSZNmmRr/+ijj+L48eN49tlnkZWVhffffx9ffPEFZs2a5YjyiYiIyAk5NNzs2bMH8fHxiI+PBwDMnj0b8fHxmDdvHgAgLy/PFnQAIDo6GuvXr8fmzZsxaNAgLFq0CP/85z95GTgRERHZOHRYavTo0RBCtPh6c3cfHj16NPbu3duBVREREZErc6kJxURERESXwnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLi8HCzdOlSREVFwc3NDcOHD8euXbsu2n7JkiXo3bs33N3dERERgVmzZqGmpqaTqiUiIiJn59Bws3r1asyePRvz589HRkYGBg0ahKSkJBQWFjbbfuXKlXj++ecxf/58ZGZm4uOPP8bq1avxwgsvdHLlRERE5KwcGm4WL16M6dOnY8qUKejbty+WLVsGnU6HTz75pNn227Ztw1VXXYX77rsPUVFRGDt2LO69995L9vYQERFR1+GwcFNbW4v09HQkJib+UYxCgcTERGzfvr3ZfUaOHIn09HRbmDl+/Dg2bNiAm266qcX3MZvNKC8vb/IgIiIi+VI56o2NRiMsFgv0en2T7Xq9HllZWc3uc99998FoNOLqq6+GEAL19fV49NFHLzoslZKSgpdfftmutRMREZHzcviE4rZIS0vD66+/jvfffx8ZGRlYs2YN1q9fj1dffbXFfebMmYOysjLb4/Tp051YMREREXU2h/XcBAYGQqlUoqCgoMn2goIChISENLvP3Llz8cADD2DatGkAgAEDBqCqqgoPP/wwXnzxRSgUF2Y1rVYLrVZr/w9ARERETslhPTcajQZDhgxBamqqbZvVakVqaioSEhKa3cdkMl0QYJRKJQBACNFxxRIREZHLcFjPDQDMnj0bkydPxtChQzFs2DAsWbIEVVVVmDJlCgBg0qRJCA8PR0pKCgBg/PjxWLx4MeLj4zF8+HAcO3YMc+fOxfjx420hh4iIiLo2h4abe+65B0VFRZg3bx7y8/MxePBgbNy40TbJ2GAwNOmpeemllyBJEl566SXk5uYiKCgI48ePx2uvveaoj0BEREROxqHhBgCSk5ORnJzc7GtpaWlNnqtUKsyfPx/z58/vhMqIiIjIFbnU1VJEREREl8JwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLi8HCzdOlSREVFwc3NDcOHD8euXbsu2r60tBQzZsxAaGgotFotevXqhQ0bNnRStUREROTsVI5889WrV2P27NlYtmwZhg8fjiVLliApKQnZ2dkIDg6+oH1tbS1uuOEGBAcH46uvvkJ4eDhOnToFX1/fzi+eiIiInJJDw83ixYsxffp0TJkyBQCwbNkyrF+/Hp988gmef/75C9p/8sknKC4uxrZt26BWqwEAUVFRnVkyERERObl2DUv16NED586du2B7aWkpevTo0apj1NbWIj09HYmJiX8Uo1AgMTER27dvb3afb775BgkJCZgxYwb0ej369++P119/HRaLpcX3MZvNKC8vb/IgIiIi+WpXuDl58mSzgcJsNiM3N7dVxzAajbBYLNDr9U226/V65OfnN7vP8ePH8dVXX8FisWDDhg2YO3cuFi1ahH/84x8tvk9KSgp8fHxsj4iIiFbVR0RERK6pTcNS33zzje3PmzZtgo+Pj+25xWJBampqhw4TWa1WBAcH48MPP4RSqcSQIUOQm5uLN998E/Pnz292nzlz5mD27Nm25+Xl5Qw4REREMtamcDNhwgQAgCRJmDx5cpPX1Go1oqKisGjRolYdKzAwEEqlEgUFBU22FxQUICQkpNl9QkNDoVaroVQqbdvi4uKQn5+P2tpaaDSaC/bRarXQarWtqomIiIhcX5uGpaxWK6xWKyIjI1FYWGh7brVaYTabkZ2djVtuuaVVx9JoNBgyZAhSU1ObHD81NRUJCQnN7nPVVVfh2LFjsFqttm1HjhxBaGhos8GGiIiIup52zbk5ceIEAgMDL/vNZ8+ejY8++giffvopMjMz8dhjj6Gqqsp29dSkSZMwZ84cW/vHHnsMxcXFmDlzJo4cOYL169fj9ddfx4wZMy67FiIiIpKHdl8KnpqaitTUVFsPzp998sknrTrGPffcg6KiIsybNw/5+fkYPHgwNm7caJtkbDAYoFD8kb8iIiKwadMmzJo1CwMHDkR4eDhmzpyJ5557rr0fg4iIiOzIKgBAcmgN7Qo3L7/8Ml555RUMHToUoaGhkKT2f4jk5GQkJyc3+1paWtoF2xISErBjx452vx8RERF1jJKqWqQVqOA19FaH1tGucLNs2TKsWLECDzzwgL3rISIiIhcjhMD+M2XYesyIeqsC3sPvhLleOKyeds25qa2txciRI+1dCxEREbmg9FMlSDtShHqrQLCbFfmfzoJW5bihqXaFm2nTpmHlypX2roWIiIhcTL3FigxDKQAgoUcArg6qh6XywlUMOlO7hqVqamrw4Ycf4scff8TAgQNt6zw1Wrx4sV2KIyIiIueWlV+B6joLvNxUGNrdD2dz8hxdUvvCzf79+zF48GAAwMGDB5u8djmTi4mIiMh1CCGw93QpAGBwhC8UCufIAO0KN1u2bLF3HURERORiDMUmFFfVQqNUoF+Yt6PLsWnXnBsiIiKixl6bvmHe0KqUF2/cidrVczNmzJiLDj/99NNP7S6IiIiInF9tvRWni00AgIHhPpdo3bnaFW4a59s0qqurw759+3Dw4MELFtQkIiIi+cktrYZVAN5uKvjq1JfeoRO1K9y8/fbbzW5fsGABKisrL6sgIiIicn6G8702kf46p7uYyK5zbu6///5WrytFREREruvP4cbZ2DXcbN++HW5ubvY8JBERETmZypp6FFfVQgIQ4YThpl3DUnfccUeT50II5OXlYc+ePZg7d65dCiMiIiLn1NhrE+ythZvaea6SatSucOPj03RWtEKhQO/evfHKK69g7NixdimMiIiInJMzD0kB7Qw3y5cvt3cdRERE5AKEELZw093fw8HVNK9d4aZReno6MjMzAQD9+vVDfHy8XYoiIiIi51RcVYvqOgtUCgkhPs45z7Zd4aawsBB/+9vfkJaWBl9fXwBAaWkpxowZg1WrViEoKMieNRIREZGTKKwwAwCCvbRQOslaUn/VrqulnnjiCVRUVODQoUMoLi5GcXExDh48iPLycjz55JP2rpGIiIicREF5DQBA7+2cvTZAO3tuNm7ciB9//BFxcXG2bX379sXSpUs5oZiIiEjGCsrP99x4ax1cScva1XNjtVqhVl94q2W1Wg2r1XrZRREREZHzsVgFiiobwo0z99y0K9xcd911mDlzJs6ePWvblpubi1mzZuH666+3W3FERETkPIqramGxCmhUCvi6O9d6Un/WrnDz3nvvoby8HFFRUYiJiUFMTAyio6NRXl6Od9991941EhERkRNonG8T7KV1uvWk/qxdc24iIiKQkZGBH3/8EVlZWQCAuLg4JCYm2rU4IiIich4FFc4/mRhoY8/NTz/9hL59+6K8vBySJOGGG27AE088gSeeeAJXXnkl+vXrh19//bWjaiUiIiIHKjw/mVjv5byTiYE2hpslS5Zg+vTp8Pb2vuA1Hx8fPPLII1i8eLHdiiMiIiLnUG+xwugCk4mBNoab33//HTfeeGOLr48dOxbp6emXXRQRERE5F2NlLawCcFcr4eV2WQscdLg2hZuCgoJmLwFvpFKpUFRUdNlFERERkXMprHCNycRAG8NNeHg4Dh482OLr+/fvR2ho6GUXRURERM7lXFUtACDQyefbAG0MNzfddBPmzp2LmpqaC16rrq7G/Pnzccstt9itOCIiInIOxZUN4SbAQ+PgSi6tTYNmL730EtasWYNevXohOTkZvXv3BgBkZWVh6dKlsFgsePHFFzukUCIiInIMIQSMVQ2TiQM8ZRZu9Ho9tm3bhsceewxz5syBEAIAIEkSkpKSsHTpUuj1+g4plIiIiBzDVGtBTZ0VEgB/nczCDQB0794dGzZsQElJCY4dOwYhBGJjY+Hn59cR9REREZGDNc638dGpoVK2a3GDTtXua7n8/Pxw5ZVX2rMWIiIickLnzt/fxhXm2wDtXFuKiIiIuo7GnpsAT+e/UgpguCEiIqJLOHf+SqlA9twQERGRqxNC4JztSin23BAREZGLq6ipR51FQClJ8HFveZUCZ8JwQ0RERC1qnG/j56GGUuHcyy40YrghIiKiFv1xpZRrDEkBDDdERER0EY09N/4ucGfiRgw3RERE1KIS0/lw4wJ3Jm7EcENERETNEkKgpKoOAOCnc43JxADDDREREbXAVGtBraVhTSkfhhsiIiJydY1DUt7uaqgUrhMZXKdSIiIi6lQlpoYhKV8X6rUBGG6IiIioBY09N34uNJkYYLghIiKiFpSaXG8yMcBwQ0RERC0oqWLPDREREcmExSpQVtPYc8NwQ0RERC6urLoOQgBqpQQPrdLR5bQJww0RERFd4M+TiSXJNRbMbMRwQ0RERBdoDDeudhk4wHBDREREzfhj2QXXmm8DMNwQERFRM1z1HjcAww0RERE1w1XvcQMw3BAREdFf1NRZUF1nAQD4suemfZYuXYqoqCi4ublh+PDh2LVrV6v2W7VqFSRJwoQJEzq2QCIioi6kcUjKU6uCRuUUUaFNHF7x6tWrMXv2bMyfPx8ZGRkYNGgQkpKSUFhYeNH9Tp48iaeffhrXXHNNJ1VKRETUNbjqgpmNHB5uFi9ejOnTp2PKlCno27cvli1bBp1Oh08++aTFfSwWCyZOnIiXX34ZPXr06MRqiYiI5M9Vl11o5NBwU1tbi/T0dCQmJtq2KRQKJCYmYvv27S3u98orryA4OBgPPfTQJd/DbDajvLy8yYOIiIha9seVUuy5aTOj0QiLxQK9Xt9ku16vR35+frP7bN26FR9//DE++uijVr1HSkoKfHx8bI+IiIjLrpuIiEjO/rhSij03Ha6iogIPPPAAPvroIwQGBrZqnzlz5qCsrMz2OH36dAdXSURE5LqsQqC0+ny48XDNcKNy5JsHBgZCqVSioKCgyfaCggKEhIRc0D4nJwcnT57E+PHjbdusVisAQKVSITs7GzExMU320Wq10Gq1HVA9ERGR/FTU1MNiFVAqJHi5OTQmtJtDe240Gg2GDBmC1NRU2zar1YrU1FQkJCRc0L5Pnz44cOAA9u3bZ3vceuutGDNmDPbt28chJyIiostkW1PKXQ2Fiy2Y2cjhkWz27NmYPHkyhg4dimHDhmHJkiWoqqrClClTAACTJk1CeHg4UlJS4Obmhv79+zfZ39fXFwAu2E5ERERt13illKteBg44Qbi55557UFRUhHnz5iE/Px+DBw/Gxo0bbZOMDQYDFAqXmhpERETkskpcfDIx4AThBgCSk5ORnJzc7GtpaWkX3XfFihX2L4iIiKiLcuUFMxuxS4SIiIhsbJeBe7jusBTDDREREQEAauutqDTXA2DPDREREclAaXXDkJSbWgE3tdLB1bQfww0REREBAEqqXH8yMcBwQ0REROfJYTIxwHBDRERE57n6gpmNGG6IiIgIwJ+vlGLPDREREbk4IQSHpYiIiEg+qmotqLMISAC83Z3iHr/txnBDREREtjWlvN3VULn4skeuXT0RERHZhW2+jYtPJgYYboiIiAjyuQwcYLghIiIiMNwQERGRzJScH5by5bAUERERuTqLVaC8Wh73uAEYboiIiLq8suo6CABqpQQPjesumNmI4YaIiKiL+/N8G0mSHFzN5WO4ISIi6uIa73Ejh8nEAMMNERFRlyenycQAww0REVGXJ6fLwAGGGyIioi7vj9XA2XNDRERELq6mzoLqOgsA9twQERGRDDQOSXlqVVAr5REL5PEpiIiIqF3kNpkYYLghIiLq0uR2GTjAcENERNSl/XGlFHtuiIiISAb+uFKKPTdERETk4oQAShsXzOSwFBEREbk6k6VhRXClQoKXm8rR5dgNww0REVEXVVHXsEimr7saChksmNmI4YaIiKiLsoUbGU0mBhhuiIiIuqzK+oZwI6f5NgDDDRERUZfV2HMjpyulAIYbIiKiLqv8fLjxZ7ghIiIiV6dw84LZej7cyGxYSj7XfTkJg8EAo9Fo9+MGBgYiMjLS7sclIqKuSR3Y8J3i7aaCRiWvvg6GGzsyGAzoExeHapPJ7sd21+mQlZnJgENERHbRGG7kNiQFMNzYldFoRLXJhInPvQl9ZIzdjltgyMHnbzwDo9HIcENERHahDuwOAAjw1Dq4EvtjuOkA+sgYdIvt5+gyiIiIWqQ533MTIMOeG3kNshEREVGryHlYiuGGiIioiymrsUDp4QuA4YaIiIhk4Ex5PQBApxRQK+UXBeT3iYiIiOiiDOfDjbdaOLiSjsFwQ0RE1MWcLjsfbjQMN0RERCQDp9lzQ0RERHIhhGC4ISIiIvkoqjCj3GyFsFrgpWK4ISIiIhd3KK8cAFBXfAYyW1LKRqYfi4iIiJpz+Oz5cFNwwsGVdByGGyIioi6kMdzUFh53cCUdh+GGiIioCzmcx3BDREREMlFprsfJc1UAgNpCDksRERGRi8vOL4cQgL+7AlZTmaPL6TAMN0RERF1E43ybKF+1gyvpWAw3REREXUTjfJtoX5WDK+lYDDdERERdRGPPTTR7bjre0qVLERUVBTc3NwwfPhy7du1qse1HH32Ea665Bn5+fvDz80NiYuJF2xMRERFQb7EiK78CAIelOtzq1asxe/ZszJ8/HxkZGRg0aBCSkpJQWFjYbPu0tDTce++92LJlC7Zv346IiAiMHTsWubm5nVw5ERGR6zhurIK53gqdRokQT6Wjy+lQDg83ixcvxvTp0zFlyhT07dsXy5Ytg06nwyeffNJs+88//xyPP/44Bg8ejD59+uCf//wnrFYrUlNTO7lyIiIi1/H76VIAQL8wbygkybHFdDCHhpva2lqkp6cjMTHRtk2hUCAxMRHbt29v1TFMJhPq6urg7+/fUWUSERG5vL3nw80VkX6OLaQTOHS6tNFohMVigV6vb7Jdr9cjKyurVcd47rnnEBYW1iQg/ZnZbIbZbLY9Ly8vb3/BRERELmqvoRQAEB/pC9RWO7SWjubwYanLsXDhQqxatQpr166Fm5tbs21SUlLg4+Nje0RERHRylURERI5VZa5Hdn7DP+7ju0DPjUPDTWBgIJRKJQoKCppsLygoQEhIyEX3feutt7Bw4UL88MMPGDhwYIvt5syZg7KyMtvj9OnTdqmdiIjIVew/UwarAMJ83KD3br4zQE4cOiyl0WgwZMgQpKamYsKECQBgmxycnJzc4n7/8z//g9deew2bNm3C0KFDL/oeWq0WWq3WnmUTEZGLMhgMMBqNdj9uYGAgIiMj7X5ce9l7ugRA1+i1ARwcbgBg9uzZmDx5MoYOHYphw4ZhyZIlqKqqwpQpUwAAkyZNQnh4OFJSUgAAb7zxBubNm4eVK1ciKioK+fn5AABPT094eno67HMQEZFzMxgM6BMXh2qTye7HdtfpkJWZ6bQBp8l8my7A4eHmnnvuQVFREebNm4f8/HwMHjwYGzdutE0yNhgMUCj+GD374IMPUFtbi7vuuqvJcebPn48FCxZ0ZulERORCjEYjqk0mTHzuTegjY+x23AJDDj5/4xkYjUanDDdCCIYbR0hOTm5xGCotLa3J85MnT3Z8QUREJFv6yBh0i+3n6DI6zZmSahgrzVArJfQL83F0OZ3Cpa+WIiIiootrvL9N31BvuKnlfWfiRgw3REREMrb7RDGArjOZGGC4ISIikrXfjjVcHTYyJsDBlXQehhsiIiKZyi2txnFjFRQSMILhhoiIiFxdY6/NoAhfeLupHVxN52G4ISIikqnGcHN1z0AHV9K5GG6IiIhkSAhhCzdXMdwQERGRq8suqICxshbuamWXuXlfI4YbO7JYhaNLICIiAgD8duwcAGBYtD+0qq5xf5tGDDd2kl9WgwfWFUB/30LsL1Hi5LkqCMGwQ0REjvHr0SIAXW++DeAkyy/Iwf4zpaipF3CL6I+jFcDRfWfRPUCHMb2D4ePedWaoExGR45VV19nm24zuHeTgajofe27s5Po4Pd5JCoRx/WJEeViglCScOmfCv3ecwvGiSkeXR0REXcjmwwWoswj00nsiVu/l6HI6HXtu7ESpkBDho0bVwZ8w5LFkXDugB37KKsSZ0mqsP5CHmwaEIibI09FlEhHRnwghUFZdh6IKM4oqzSgx1aGyph5VtfWorbeizmKFBAkKBaBVKeGhVcLLTY0gTy0CvTQI93V39Edo1vr9ZwEAtwwMc3AljsFw00H8PDS4PT4cmw7l40hhJTYcyMMtA8MQHejh6NKIiLouhQpFNRJOHz+H3JJqFFWYUWuxXmInAYsFqLPUo9Jcj4JyM44VNvTISxLgr1HB68rbsXN/VoeUHBgYiMjIyFa3LzPV4dejDUNSNw0I7ZCanB3DTQdSKCQk9QsB0BBwNh7Kx33DIjkHh4ioE50uNuGHwwX4bk8xImauwi+FagDFtteVCgkBHhoEeWkR4KGBl5sanloVNCoF1EoJQMPVsOZ6KyrN9Sgz1aGo0oz88hqUmupwzqyA/3UPYeFhC6q/XYfyXWtgPnPIbvW763TIysxsdcDZdDgf9VaBPiFe6BncNUcMGG46mEIhYWy/EFSYzyCvrAYbDuTh7qHdoFJwuhMRUUcQQuBwXjl+OFSAHw4XIDOv3PaaQuMGrUKge5AXIvx0CPFxg59OA6VCatWx9X95Xl5dh1+27cChM8Vw69YXutjh0MUOh7/Gin6+FgS7Xd5VswWGHHz+xjMwGo2tDjfr9+cBAG4Z2DV7bQCGm06hVEgY1z8EK3cZUFhhxtajRozuHezosoiIZEMIgQxDCdbvz8cPh/NxpqTa9ppSIWFYlD/6eNXitScnIXnB24joZZ8vfm93NUJRgp8+fxZ3zf8YNQG9cDivHMW1CvxaqED3AB2ujQ2Cv4fGLu93KfllNbarpLrqkBTAcNNpvNzUSOobgv/+fha/nylDz2BPdPPTObosIiKnYzAYYDQaW9X2dFkdfjFU41dDDQqrLLbtGiUwOESL4eFuGBrqBi+tApmZOagznoLUuk6aNtMpLBjZJxjDo/2x+2QxDuSW4dQ5Ez4vPoWh3f1xZZQfVMqO7bVfvu0E6q0Cw6L90aMLX8TCcNOJogI90D/cGwdzy5GaVYiJwyI7/BediMiVGAwG9ImLQ7XJ1GIbpWcAPPpeC4++o6HRx9i2W80mmI7ugOnINtSc2Iuj9WZ82cz+lZUde3sOD60Ko3sHY3CEL34+UoST50zYdbIYRwsrkNQvBHpvtw5530pzPVbuNAAApl/To0Pew1Uw3HSyq2MCcaKoCqWmOuw+WYKEmABHl0RE5DSMRiOqTSZMfO5N6CP/FFwEkFct4USlEgU1EoCG7hcJAiHuAhE6K0LdVVDFXg3cdHWzx87c9TO+//Qd1NTUdMZHga9Og1sHheFYUSV+zi5CiakOX+w5jRE9AjCkux8Udu5C+mL3aVTU1KNHoAeu79O1pz4w3HQyrVqJUb2DsOFAPvacKkafEC/4ddJYLBGRq9BHxqBbbD+UV9fh4NkyHD5bjqraP4adwnzc0DvEC7F6L7irW7duUoEhp6PKbZEkSYgN9kI3Px1+yizEsaJKbMs5h5PnqpDUNwTedrp6tt5ixcdbTwAAHromGopWTpCWK4YbB+gZ5ImoAB1OnjNh6zEjxg/qmjdZIiJqSbFZwv4DeThWWInG643c1Ur0DfNG/zBv+Opc6x+F7molbhoQgsy8CqQdKcTZ0hp8vtOAMX2C0CfE+7KP/5/dp5FbWg1/Dw3uvKKbHSp2bQw3DiBJEq6JDcKp4lM4bqzCmRITJxcTUYdoy+TctmjrjeVaw2IV2HGmBvqJb2BLgRpAw9yYbn7uGBjugx5Bnq2+ZNsZSZKEvmHeCPdzx6ZD+cgrq8GmQwU4dc6E0b2D2r1y99nSarzxfcMNBJ+4rifcWtmTJWcMNw7i76HBgDAf7M8tw69HjfjblRGQOmoKPxF1Sa2ZnNtebb2x3MWYauvxVfoZfLz1BE6dM8GtWz9IEOgT6o34CD8EeWntULHz8HFX464rumHXyWLsOlGMrPwKnC2txo39QxDq07blHIQQeGHtAVSa63FFpC8mJUR1TNEuhuHGgYb38EdWfgUKK8zIyq9AXOjld00SETVqaXLu5WrPjeWaU1RhxmfbT+JfO06h1FQHAPDUSMhNW4V777odsXEh9irZ6SgUEkb0CECkvw4bD+WjvKYeX6afwbAofwyN8mv1jV7/teMU0rKLoFEq8D93DXTpni17YrhxIJ1GhSuj/PBbzjlsyzmHnsGeUPPScCKys8bJuc4ip6gS//z1OL7OyEVtfcO6TpH+Oky7Jho9lUZc9eq/4P632x1cZecI83XHxOGR+CmrEEcKKrHzRDGyCyowpncwIv0vPl3hn78exz/WZwIAZibGomdw11v9uyUMNw42OMIX+3PLUFFTj72GUgyL9nd0SUREdieEwJ5TJfi/n4/jx8wC2/bBEb545NoeGNsvBEqFhIyM4oscRZ60KiVu7BeCmKBK/HykCKWmOqzdm4tQHzdEayRA0fSr+nSxCR/8nGO7p80j1/bA46Pt1zMnBww3DqZSKjAyJgCbDhVgz6li9AvzhoeW/1mI2qujJtACHTOJVu6MlWas25uLr9LPICu/wrY9MU6PR0b1wNDufpxviIbJxr30Xujur8P24+dwMLcceWU1yIMaETNXYcHP56A/uAdlpjrsOVUM6/lLyJ4e2wszxvTkOfwLfos6gd56L+w7XYqCcjN2HD+H6+P+ujQbEbVGR06gBew7iVbOKs31+PVIEb7OyEVadiHqz38Ta1QK3HlFOKZd0wMxXXhpgIvRqpUY3TsYQ6P8sddQgoNnSlCrccP+glqg4I8er2tiA/HItTG4OjbQgdU6L4YbJ9B4afhX6Wdw6Gw5Bkf4IsBTXlcHEHWGjppAC9hvEq3cCCFgrKxFZl450k+VYPfJYuw+WYw6yx+rYQ+K8MVdQ7rh1oFh8NHZ56Z1cuepVeGa2CBEiUK8t2AWXnn/34iIjICnVoW4UG/00nN+zcUw3DiJcF93xAR5IKeoCr/lnMOtvLEfUbs52wRaVyGEQE29FdW1FlTXWWCut6CuXqDOYkWtxYo6ixV19QLFxUoE3f4inv3RCON3m21XOv1Z9wAdbuwXgjuHdOMX8WWQJKDOeApjY3S44oruji7HZTDcOJGrYgJx3FiFE7yxHxF1IKtVwFhlRl5pDQorzCgx1aLUVIeaOgvEpXcHoISuVwKOFTeEGkkCuvvrMDjCF0O6++Hq2CBEB3p05EcguiiGGyfi56FB/zAfHOCN/YjIzmrrrTheVInjxiqcKjbZLsFujkalgLtaCa1KAY1KAY1SAbVSAbVSgkalQHXpOWz96kO8+fILuGZIP8QEefKuuORUGG6czPBof2Tll6OwwowjBZXoHcLuXCJqHyEEtBEDsMuoRN6Z47aJvUBDgAn1cYPe2w0BHhr46tTw0KjgplZe8kZwZ44WYuPe9RgW/gr6hfl09McgajOGGyfjoVVhSHc/7DhejG05RsQEs2uXiNqmzmLFur25eG+zESH3peC0CQAE/HRqxAZ7ITrQA8HeWijYM0wyxXDjhK6I9MOBM2Uor6nH/jNlCHZ0QUTkEuotVqzZm4v3fjoGQ3HD5fDW2hrE+KsxrE8U9N5aDnVTl8B7/TshtVKBET0CAAC7ThSjtuWhcSIi1Fus+HLPaVy/+Gc8+9V+GIpNCPDQ4IGBXsh9fzKu8LcgxMeNwYa6DPbcOKm+Yd7Yd7oU56pqcbiUE/WI6EL1FivW7TuLd386ilPnGnpqAjw0eGRUD9w/ojuyDu7HP8xVDq6SqPMx3DgphSTh2l5BWLs3FzmVCqiDox1dEhE5iXqLFf89H2pOng81/h4aPHJtDzyQ0B06Df9qp66N/wc4sUh/HWKDPXG0sBL+NzwKIVp3BwoikqfGicLvp+XghLGhR8bfQ4OHr+2BB0Z057p0ROfx/wQnd01sII4XVcCtWz+knarGkCGOrojIuZw6V4UDuWU4kl+BA8dLETThBWwtVMGnJg9atRK+7moEeGoQ7O0Gdxe9F4u53oKv03PxftoxnCmpBgD46dSYfm0PTE6IYqgh+gv+H+HkvNzUiPO24GCZCsv3leOBG2oQ7OXm6LKIHEYIgf1nyrB2b8OijI3DMo10vUeioAYoqKm8YN9gLy26B+jQS++FQBdYv63UVIvVu09jxbaTyCurAQAEemow/ZqGOTUMNUTN4/8ZLiDW24qMIzmo1Mdg/n8P4YP72X1DXY/FKrB2by7++etxZOVX2LarlRL6hfmgT4gXVDWleOethRh77yPwCAhBTZ0VxaZaGCvNKDXVobDCjMIKM3afLEGgpwZ9QrzRS+8JLzfnWswxK78cn247ibV7c1FT13C5pN5bi0eujcG9wyLhrnHNHiiizsJw4wIUEmDcsAQRU9/F9wfzseFAHm4aEOrosog6hRACmw7l460fjuBYYUNvjEalwLj+IbhpQCiu6hkIz/M9GBkZGXht3/eImj4d3SL9mhynylwPQ7EJOUWVOGGsgrGyFluPGbH1mBGR/jr0C/NGjyAPqBSdf4cMg8GAE7mF2HamGr+cqkam8Y+FKKN8Vbippweu7e4OjbIEmQdLWn3czMzMjiiXyOkx3LiIusITuL2PJ77KrMQLaw9gcIQvwnzdHV0WUYcRQmDrMSPe3JSN/WfKAAC+OjUeHRWDe6+MhI+ubb0tHloV4kK9ERfqjZo6C44WVCKroBxnS2tgKDbBUGyCm1qBuBBv9AvzRkAnDFtVmuvx9W+ZeH7Z19B0HwxJ2fCZhNUC05FtqEj/DqfOHMLPl/s+lRcO0RHJGcONC7m7ryeyy5U4kFuGJ/+zF6seHgGVkvdhJPnJMJTgfzZmYcfxYgCATqPEtKujMe3aHvC2wxCSm1qJAd18MKCbD8qq63D4bDkO5ZWhymzB3tOl2Hu6FEGeWvQK8URMkCf8dJrLfk+gIbBlF1Tg5+wipGUXYc+pYtRZBLQ9rgQA+KitiPCwIkJnhS5qODB2+GW9X+aun/H9p++gpqbGHuWTA9m7F07uvXoMNy5ErZTw3n3xuOV/t2LPqRIs3nwEz97Yx9FlEdlNVn453tp0BD9mFgAANEoF7h/RHY+PiemwCcA+7mokxARgeLQ/ThWbcOhsGU4Yq1BUaUbRMTN+O3YOvu5q+CuV0MWNQl5FPSxWccnFJYUQyCurwYHcMhzKLcOB8w9jZW2TdnoPJY78uAp33j4B/fr1tetnKzDk2PV4f9URX5By/9Jtq/LiIgDA/fff3yHHl2uvHsONi+ke4IGFdw7EjJUZeD8tB71DvHDb4HBHl0V0WU6dq8LizUfwze9nIUTDPLO7h0TgycRYhHfS8KtCISE60APRgR6orrPgWGEljhZWILekGqXVdSiFEkG3PoMZ3xdB88NGRPi7I8BTCz+dGmqlApIkoabOgvLqOhgrzThTUg1z/YVrp2hVCiTEBGB0ryCM6h2M4lNZGDrvU/jcc1unfE576OgvXEC+X7ptVV1ZDgC4+ZEX0Xug/S4mkXuvHsONC7p5YCh+P9MDH/5yHM98uR96bzfbWlTkvAwGA4xGY4ccOzAwEJGRkR1y7I500liF//slB1/uOYN6a8NNKm8eGIrZN/RCTJCnw+pyVysxINwHA8J9YK634HRxNbJPnsGh7Bx4RfRCrcWKnKIq5BRdfGkDpUJCbLBnw7G6+aB/uA/6hnrD7U/32ykxuN56Tx31hQvI/0u3vQLCuqNbbD+7Ha+je/UcjeHGRT1/Yx+cKTFhw4F8PPzZHvzn4RHoF+bj6LKoBQaDAX3i4lBtMl26cTu463TIysx0mYBzMLcMy37OwYYDeTifaTC6dxCeHtsb/cOd6/dYq1KiZ7An3MosSP3XbHy7Zw/00XEwFJtQXFWLUlMt6q0CQjTM5fF2V8Ffp0E3Px1CfNygUcl3Xpy9v3AB+X/pUudguHFRCoWExf9vMArKdyL9VAnu/XAHVkwdhiv+cvkrOQej0YhqkwkTn3sT+sgYux67wJCDz994Bkaj0anDTb3FirTsIvxrxyn8fKTItn1M7yA8PqYnrozyd2B1raeQJET46xDhr3N0KUTUAoYbF+amVmL5lCsxdflu7DlVgvv/uRPvT7wCo3sHO7o0aoE+Mqbd/9JtXFtMAJAASJLzD2dYrAK7TxZj06F8fLc/D0UVZgANc2rGDwrDI9fGoG+Yt4OrJCK5Ybhxcd5uanz20DA8/Fk6th4zYsqK3Xjyulg8eX3sJa/mIMept1hRaa5veNTUo+L8n6vM9aius8Bcb0VtvRXmeivqLFb8dc1UCYBaqYBKKUGyqhH64DtYkHYO3bP3IsBDA//zj8Y/B3hq4O+hhY+7uk2/F22dJ1RdZ8XJ0nocK6nDoSIzDhfVorL2j+K9tQqM6u6OcT116B8VgEgGGyLqAAw3MqDTqPDxg0Px8reHsXKnAe+kHsX24+fwjwn90Uvv5ejyXE57J/4KIVBTL1BmtqKsxmr7WVJjQU7eOQTdOQ8/5qlgzsux3VK/vQSAWosVtRYAkKDRx2B/YS32F5696H6S1HDps59OA1+dGv46DXx1Gvi4q+GuUcBdrYTb+YepvBTPPTMb5moTJEkBKBSApICkUEKh9YDC3RsKd28o3b2g9PSHyi8UKq/AC97TUl2B6mO7YDqyDaeOp+OAtR7vwfXmCRGR62C4kQmtSonXbx+Aod398OLag9h1ohjj3vkVD4zojoev7SHbuxkLIWCut8JcZ0VNveWCnzV1TbfVWhp6Qmrrrai3CtSd7xmpO//n4rJyfL5yFSwCgFIFSaFsuGtsS39WKCEpVZCUaijcvaBQt7SoqRa6nsNQVgcADcFGpZDg5aaCp1YFz8afWhXc1UpoVApoVUpoVQqoVYrzw1CABAmQAKtVNNRtETh7Kger330Vry96B16BYSiuqkVxVS3ONf6sNONcVS0qauohBFBqqkOpqa6FOpvyGTe7zf9N3JUCvhqBAI0VQW4CvhotFL2vAW6+xtbGVeYJNYc3UyNyfgw3MnPHFd0wLNofr353GJsOFWDFtpP4945TuGlAKG6PD8fIngHQqhyz6J4QAjV1VpTX1KG8uu78z4bhmOpaC6pq62GqtcBUW48q8/mftZaG18x/vNbw83xwaeY+IpfLfcANl7W/UhLQKgCtsuGnm1KgsuAksrduwKjb7sOgwfHw0qqgUSnsMm/GrBWoOZGBUd11uOKKHi22q623orS6FqWmOpRU1aLEVIcSUy1KTLUoq65rCIF1FlTXNZzzwuJS7NidjrCYvtC666CQJEhSw4RarUpxvoen4adOo4SvuwY+OjXc1fJc1JE3UyNyHU4RbpYuXYo333wT+fn5GDRoEN59910MGzasxfZffvkl5s6di5MnTyI2NhZvvPEGbrrppk6s2Ll189Ph/x4Yil+PFuH9LTnYfvwcvvn9LL75/Sy8tCpcGe2PKyJ9ERfqje4BHgjzdYO7WnnJL1qrVaCqth4VNfUor6lr+Fldd8HzxtDyR4j5Y3udRVz0PS6HQgI0SglqRcPPxoe68c8KQKWUoFI0tFEpJCilhp8ND6C0+BzWrfkKV9/yN/gFhUChkKCUJCgUOP9TgkKSoFRIUEg4/7PhuZtaCXe1EmqldMG5TE89gvTfNyHgjjs77E67l6JRKRDs5YZgr5Z6l5rKyMjAkFnPY+LSNegWG9HB1Tk/3kyNyHU4PNysXr0as2fPxrJlyzB8+HAsWbIESUlJyM7ORnDwhVf9bNu2Dffeey9SUlJwyy23YOXKlZgwYQIyMjLQv39/B3wC53VNbBCuiQ3CgTNl+DL9NDYezEdhhRk/ZRXip6zCJm0VEuCmkuCukuCuVkAIAYsA6q1oGL6xCFTXC9s9SS6HQgI81BI8NAro1BJ0KgXcVBK0KgluTR4N292UEtzUErTKhvoqS4vxzN9noqayHKK+DqLeDFFfB1jrL7+488Jvuwl9usfZ7XgkH7yZGpHzc3i4Wbx4MaZPn44pU6YAAJYtW4b169fjk08+wfPPP39B+3feeQc33ngjnnnmGQDAq6++is2bN+O9997DsmXLOrV2V9G4QOCC8f1wILcMGYYS7DWU4lhhJU4aK2Gqs8IqAFOdgKlOANUXH+oRljpYa6pgNZtgNVee/1kFYTbBWlMJq7nq/M+/Pq+EtaYKos4+/0K9/clXEN3bvoHWlf8VzbkgREQNHBpuamtrkZ6ejjlz5ti2KRQKJCYmYvv27c3us337dsye3XSSY1JSEtatW9eRpToFe33ZDHIHBvUG0NsDhw8bMGnqNNw1OwW+od1Rb5VQf753RtEwdxUKCVBIAhoFoJbOb5d0AHQALrw6BvgjJHTk7dm9AkJ4d1S49lwQBjIi6ggODTdGoxEWiwV6vb7Jdr1ej6ysrGb3yc/Pb7Z9fn5+s+3NZjPMZrPteVlZGQCgvLz8ckpvVuOXwJmjh2Cutt9t9k9m7gXQsYvUVZcWITgoCJo/b2xmCKq1Az91tWbbT3ueiz8fO//kEeR42PcusY3hxt7H7qjjAsDJww2/H1eOuwfdomPtdlzDkQNI//G/OJm1H2qlfZcQ6Ojf6SP799j9984VfzdYc+ccmzU3VXTmBICG70R7ftc2Hkv89cZfzREOlJubKwCIbdu2Ndn+zDPPiGHDhjW7j1qtFitXrmyybenSpSI4OLjZ9vPnzxdo+Jrmgw8++OCDDz5c/HH69OlL5guH9twEBgZCqVSioKCgyfaCggKEhIQ0u09ISEib2s+ZM6fJMJbVakVxcTECAgLsfvv68vJyRERE4PTp0/D25p1XOwrPc+fgee4cPM+dh+e6c3TUeRZCoKKiAmFhYZds69Bwo9FoMGTIEKSmpmLChAkAGsJHamoqkpOTm90nISEBqampeOqpp2zbNm/ejISEhGbba7VaaLVNL7319fW1R/kt8vb25v84nYDnuXPwPHcOnufOw3PdOTriPPv4+LSqncOvlpo9ezYmT56MoUOHYtiwYViyZAmqqqpsV09NmjQJ4eHhSElJAQDMnDkTo0aNwqJFi3DzzTdj1apV2LNnDz788ENHfgwiIiJyEg4PN/fccw+Kioowb9485OfnY/Dgwdi4caNt0rDBYIBC8cdkxpEjR2LlypV46aWX8MILLyA2Nhbr1q3jPW6IiIgIgBOEGwBITk5ucRgqLS3tgm1333037r777g6uqu20Wi3mz59/wTAY2RfPc+fgee4cPM+dh+e6czjDeZaEaM01VURERESuwb43ryAiIiJyMIYbIiIikhWGGyIiIpIVhhsiIiKSFYYbO1m6dCmioqLg5uaG4cOHY9euXY4uyaWlpKTgyiuvhJeXF4KDgzFhwgRkZ2c3aVNTU4MZM2YgICAAnp6euPPOOy+4ezW1zcKFCyFJUpObZPI8209ubi7uv/9+BAQEwN3dHQMGDMCePXtsrwshMG/ePISGhsLd3R2JiYk4evSoAyt2PRaLBXPnzkV0dDTc3d0RExODV199tcl6RDzPbffLL79g/PjxCAsLgyRJFyxW3ZpzWlxcjIkTJ8Lb2xu+vr546KGHOm5h3ksu0ECXtGrVKqHRaMQnn3wiDh06JKZPny58fX1FQUGBo0tzWUlJSWL58uXi4MGDYt++feKmm24SkZGRorKy0tbm0UcfFRERESI1NVXs2bNHjBgxQowcOdKBVbu2Xbt2iaioKDFw4EAxc+ZM23aeZ/soLi4W3bt3Fw8++KDYuXOnOH78uNi0aZM4duyYrc3ChQuFj4+PWLdunfj999/FrbfeKqKjo0V1dbUDK3ctr732mggICBDfffedOHHihPjyyy+Fp6eneOedd2xteJ7bbsOGDeLFF18Ua9asEQDE2rVrm7zemnN64403ikGDBokdO3aIX3/9VfTs2VPce++9HVIvw40dDBs2TMyYMcP23GKxiLCwMJGSkuLAquSlsLBQABA///yzEEKI0tJSoVarxZdffmlrk5mZKQCI7du3O6pMl1VRUSFiY2PF5s2bxahRo2zhhufZfp577jlx9dVXt/i61WoVISEh4s0337RtKy0tFVqtVvznP//pjBJl4eabbxZTp05tsu2OO+4QEydOFELwPNvDX8NNa87p4cOHBQCxe/duW5vvv/9eSJIkcnNz7V4jh6UuU21tLdLT05GYmGjbplAokJiYiO3btzuwMnkpKysDAPj7+wMA0tPTUVdX1+S89+nTB5GRkTzv7TBjxgzcfPPNTc4nwPNsT9988w2GDh2Ku+++G8HBwYiPj8dHH31ke/3EiRPIz89vcq59fHwwfPhwnus2GDlyJFJTU3HkyBEAwO+//46tW7di3LhxAHieO0Jrzun27dvh6+uLoUOH2tokJiZCoVBg586ddq/JKe5Q7MqMRiMsFottuYhGer0eWVlZDqpKXqxWK5566ilcddVVtmU28vPzodFoLlgEVa/XIz8/3wFVuq5Vq1YhIyMDu3fvvuA1nmf7OX78OD744APMnj0bL7zwAnbv3o0nn3wSGo0GkydPtp3P5v4u4bluveeffx7l5eXo06cPlEolLBYLXnvtNUycOBEAeJ47QGvOaX5+PoKDg5u8rlKp4O/v3yHnneGGnN6MGTNw8OBBbN261dGlyM7p06cxc+ZMbN68GW5ubo4uR9asViuGDh2K119/HQAQHx+PgwcPYtmyZZg8ebKDq5OPL774Ap9//jlWrlyJfv36Yd++fXjqqacQFhbG89yFcFjqMgUGBkKpVF5w9UhBQQFCQkIcVJV8JCcn47vvvsOWLVvQrVs32/aQkBDU1taitLS0SXue97ZJT09HYWEhrrjiCqhUKqhUKvz888/43//9X6hUKuj1ep5nOwkNDUXfvn2bbIuLi4PBYAAA2/nk3yWX55lnnsHzzz+Pv/3tbxgwYAAeeOABzJo1CykpKQB4njtCa85pSEgICgsLm7xeX1+P4uLiDjnvDDeXSaPRYMiQIUhNTbVts1qtSE1NRUJCggMrc21CCCQnJ2Pt2rX46aefEB0d3eT1IUOGQK1WNznv2dnZMBgMPO9tcP311+PAgQPYt2+f7TF06FBMnDjR9meeZ/u46qqrLridwZEjR9C9e3cAQHR0NEJCQpqc6/LycuzcuZPnug1MJhMUiqZfbUqlElarFQDPc0dozTlNSEhAaWkp0tPTbW1++uknWK1WDB8+3P5F2X2Kche0atUqodVqxYoVK8Thw4fFww8/LHx9fUV+fr6jS3NZjz32mPDx8RFpaWkiLy/P9jCZTLY2jz76qIiMjBQ//fST2LNnj0hISBAJCQkOrFoe/ny1lBA8z/aya9cuoVKpxGuvvSaOHj0qPv/8c6HT6cS///1vW5uFCxcKX19f8d///lfs379f3HbbbbxEuY0mT54swsPDbZeCr1mzRgQGBopnn33W1obnue0qKirE3r17xd69ewUAsXjxYrF3715x6tQpIUTrzumNN94o4uPjxc6dO8XWrVtFbGwsLwV3du+++66IjIwUGo1GDBs2TOzYscPRJbk0AM0+li9fbmtTXV0tHn/8ceHn5yd0Op24/fbbRV5enuOKlom/hhueZ/v59ttvRf/+/YVWqxV9+vQRH374YZPXrVarmDt3rtDr9UKr1Yrrr79eZGdnO6ha11ReXi5mzpwpIiMjhZubm+jRo4d48cUXhdlstrXheW67LVu2NPt38uTJk4UQrTun586dE/fee6/w9PQU3t7eYsqUKaKioqJD6pWE+NNtG4mIiIhcHOfcEBERkaww3BAREZGsMNwQERGRrDDcEBERkaww3BAREZGsMNwQERGRrDDcEBERkaww3BCR01uwYAH0ej0kScK6descXQ5Gjx6Np556yuHHsOdxiOSE4YaIOoS9gkhmZiZefvll/N///R/y8vIwbty4yy+uldLS0iBJ0gULhzpCS7WsWbMGr776qmOKInJSKkcXQESdx2KxQJKkCxYWdGY5OTkAgNtuuw2SJDm4Gufj7+/v6BKInI7r/A1H1MWMHj0aycnJSE5Oho+PDwIDAzF37lz8ecUUs9mMp59+GuHh4fDw8MDw4cORlpZme33FihXw9fXFN998g759+0Kr1cJgMMBsNuO5555DREQEtFotevbsiY8//ti238GDBzFu3Dh4enpCr9fjgQcegNFobFLbk08+iWeffRb+/v4ICQnBggULbK9HRUUBAG6//XZIkmR73pwDBw7guuuug7u7OwICAvDwww+jsrISQMNw1Pjx4wEACoWixXDT2KuxadMmxMfHw93dHddddx0KCwvx/fffIy4uDt7e3rjvvvtgMpls+1mtVqSkpCA6Ohru7u4YNGgQvvrqKwDAyZMnMWbMGACAn58fJEnCgw8+2GTflj7/1KlTccsttzSpsa6uDsHBwU3O85/961//wtChQ+Hl5YWQkBDcd999KCwsvGQtfx2WKikpwaRJk+Dn5wedTodx48bh6NGjttcbfyc2bdqEuLg4eHp64sYbb0ReXl6zdRG5pA5ZsYqILtuoUaOEp6enmDlzpsjKyhL//ve/hU6na7LY4rRp08TIkSPFL7/8Io4dOybefPNNodVqxZEjR4QQQixfvlyo1WoxcuRI8dtvv4msrCxRVVUl/t//+38iIiJCrFmzRuTk5Igff/xRrFq1SgghRElJiQgKChJz5swRmZmZIiMjQ9xwww1izJgxTWrz9vYWCxYsEEeOHBGffvqpkCRJ/PDDD0IIIQoLC20Lnebl5YnCwsJmP2NlZaUIDQ0Vd9xxhzhw4IBITU0V0dHRtsX4KioqxPLlywUA28rwzWlc1G/EiBFi69atIiMjQ/Ts2VOMGjVKjB07VmRkZIhffvlFBAQEiIULF9r2+8c//iH69OkjNm7cKHJycsTy5cuFVqsVaWlpor6+Xnz99dcCgMjOzhZ5eXmitLS0VZ//t99+E0qlUpw9e9b2XmvWrBEeHh62hQL/ukDpxx9/LDZs2CBycnLE9u3bRUJCghg3bpwQQlyylj8f59ZbbxVxcXHil19+Efv27RNJSUmiZ8+eora2tsnvRGJioti9e7dIT08XcXFx4r777mv23BK5IoYbIic1atQoERcXJ6xWq23bc889J+Li4oQQQpw6dUoolUqRm5vbZL/rr79ezJkzRwghbMFg3759ttezs7MFALF58+Zm3/fVV18VY8eObbLt9OnTti/WxtquvvrqJm2uvPJK8dxzz9meAxBr16696Gf88MMPhZ+fn6isrLRtW79+vVAoFCI/P18IIcTatWvFpf4d1hhufvzxR9u2lJQUAUDk5OTYtj3yyCMiKSlJCCFETU2N0Ol0Ytu2bU2O9dBDD4l77723yXFLSkqatGnN5+/bt6944403bM/Hjx8vHnzwwSbH+HMo+avdu3cLALYwdLFaGo9z5MgRAUD89ttvtteNRqNwd3cXX3zxhRDij9+JY8eO2dosXbpU6PX6FmshcjUcliJyYiNGjGgyFJOQkICjR4/CYrHgwIEDsFgs6NWrFzw9PW2Pn3/+2TZPBQA0Gg0GDhxoe75v3z4olUqMGjWq2ff8/fffsWXLlibH7NOnDwA0Oe6fjwkAoaGhtmGU1srMzMSgQYPg4eFh23bVVVfBarUiOzu7Tcf6a016vR46nQ49evRosq2xxmPHjsFkMuGGG25o8lk/++yzJp+zNe8FXPj5p02bhuXLlwMACgoK8P3332Pq1KktHi89PR3jx49HZGQkvLy8bP99DAZDKz55g8zMTKhUKgwfPty2LSAgAL1790ZmZqZtm06nQ0xMTIu1E7k6TigmclGVlZVQKpVIT0+HUqls8pqnp6ftz+7u7k0Ckru7+yWPO378eLzxxhsXvBYaGmr7s1qtbvKaJEmwWq1t+gz29ueaJEm6aI2N83rWr1+P8PDwJu20Wm2b3uuvxwaASZMm4fnnn8f27duxbds2REdH45prrmn2WFVVVUhKSkJSUhI+//xzBAUFwWAwICkpCbW1tZespa2aq138aS4XkatjuCFyYjt37mzyfMeOHYiNjYVSqUR8fDwsFgsKCwtb/NJszoABA2C1WvHzzz8jMTHxgtevuOIKfP3114iKioJK1f6/ItRqNSwWy0XbxMXFYcWKFaiqqrL13vz2229QKBTo3bt3u9+7Nf48wbqlXiyNRgMAl/wczQkICMCECROwfPlybN++HVOmTGmxbVZWFs6dO4eFCxciIiICALBnz5421xIXF4f6+nrs3LkTI0eOBACcO3cO2dnZ6Nu3b5s/A5Gr4rAUkRMzGAyYPXs2srOz8Z///AfvvvsuZs6cCQDo1asXJk6ciEmTJmHNmjU4ceIEdu3ahZSUFKxfv77FY0ZFRWHy5MmYOnUq1q1bhxMnTiAtLQ1ffPEFAGDGjBkoLi7Gvffei927dyMnJwebNm3ClClT2vQlHxUVhdTUVOTn56OkpKTZNhMnToSbmxsmT56MgwcPYsuWLXjiiSfwwAMPQK/Xt+FMtZ2XlxeefvppzJo1C59++ilycnKQkZGBd999F59++ikAoHv37pAkCd999x2KiopsvT2tNW3aNHz66afIzMzE5MmTW2wXGRkJjUaDd999F8ePH8c333xzwb1rWlNLbGwsbrvtNkyfPh1bt27F77//jvvvvx/h4eG47bbb2lQ7kStjuCFyYpMmTUJ1dTWGDRuGGTNmYObMmXj44Ydtry9fvhyTJk3C3//+d/Tu3RsTJkzA7t27ERkZedHjfvDBB7jrrrvw+OOPo0+fPpg+fTqqqqoAAGFhYfjtt99gsVgwduxYDBgwAE899RR8fX3bdH+cRYsWYfPmzYiIiEB8fHyzbXQ6HTZt2oTi4mJceeWVuOuuu3D99dfjvffea/X7XI5XX30Vc+fORUpKCuLi4nDjjTdi/fr1iI6OBgCEh4fj5ZdfxvPPPw+9Xo/k5OQ2HT8xMRGhoaFISkpCWFhYi+2CgoKwYsUKfPnll+jbty8WLlyIt956q0mb1tayfPlyDBkyBLfccgsSEhIghMCGDRsuGIoikjNJcKCVyCmNHj0agwcPxpIlSxxdCrVTZWUlwsPDsXz5ctxxxx2OLoeoy+CcGyIiO7NarTAajVi0aBF8fX1x6623Orokoi6F4YaIyM4MBgOio6PRrVs3rFix4rImZhNR23FYioiIiGSFE4qJiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhW/j/TwqqmW3QPKwAAAABJRU5ErkJggg==\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "sns.histplot(data = icm, x = 'percent of methylation', bins = 20, kde = True, kde_kws = {'bw_adjust': 2})\n",
        "plt.title('icm')\n",
        "plt.savefig('icm.png')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 472
        },
        "id": "ScTcWTzvXLI2",
        "outputId": "17376bd7-15b9-48dd-9cdd-7e15248fc516"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjcAAAHHCAYAAABDUnkqAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAUdZJREFUeJzt3XtcVHX+P/DXmSv3uwwXQfAOXkmU0PqphaKV1ba1ZZZmq+2WlunWlltqly10K3O/5ea3NrW+m2ltZhdNU4xMJS/gNVEUURAZLiJ3mIGZz+8PZJJEBZzhzBxfz8djHjpnzjnzngPCy8/ncz4fSQghQERERKQQKrkLICIiIrInhhsiIiJSFIYbIiIiUhSGGyIiIlIUhhsiIiJSFIYbIiIiUhSGGyIiIlIUhhsiIiJSFIYbIiIiUhSGGyJyWitXroQkSTh16pTcpRCRC2G4ISIiIkWRuLYUETkri8WChoYG6PV6SJIkdzlE5CIYboiIiEhR2C1FRE6rtTE33333HUaOHAlvb2/4+Phg6NChWLVqle31UaNGoX///jh48CBGjhwJDw8P9OzZE//9738BAD/++CMSEhLg7u6OPn36YMuWLZ39sYjIwRhuiMhlrFy5ErfffjvKysowd+5cLFy4EIMHD8bGjRtb7Hf+/HnccccdSEhIwD/+8Q/o9Xo88MADWLNmDR544AHcdtttWLhwIWpqanDvvfeiqqpKpk9ERI6gkbsAIqK2qKiowFNPPYVhw4YhLS0Nbm5uttd+27t+9uxZrFq1ChMnTgQAjBkzBn379sWDDz6InTt3IiEhAQAQExOD5ORkfPHFF3jkkUc67bMQkWMx3BCRS9i8eTOqqqrw/PPPtwg2AC4ZbOzl5YUHHnjA9rxPnz7w8/NDeHi4LdgAsP395MmTDqyciDrbdd0ttW3bNkyYMAFhYWGQJAnr1q1r9zmEEHjzzTfRu3dv6PV6hIeH47XXXrN/sUTXuZycHABA//79r7pv165dLwk8vr6+iIiIuGQb0NSNRUTKcV233NTU1GDQoEF49NFHcc8993ToHLNmzcL333+PN998EwMGDEBZWRnKysrsXCkRtYdarW7Xdt40SqQs13W4GT9+PMaPH3/Z100mE1544QV8+umnKC8vR//+/bFo0SKMGjUKAJCVlYX33nsPhw8fRp8+fQAA0dHRnVE60XWnR48eAIDDhw+jZ8+eMldDRM7suu6WupqZM2ciPT0dq1evxsGDB3Hfffdh3LhxOH78OADgm2++Qffu3fHtt98iOjoaUVFRmDZtGltuiBxg7Nix8Pb2RkpKCurr61u8xpYXIrrYdd1ycyV5eXlYsWIF8vLyEBYWBgB45plnsHHjRqxYsQKvv/46Tp48idOnT+Pzzz/Hxx9/DIvFgtmzZ+Pee+/F1q1bZf4ERMri4+ODt99+G9OmTcPQoUPx4IMPwt/fHwcOHEBtbS0++ugjuUskIifBcHMZhw4dgsViQe/evVtsN5lMCAwMBABYrVaYTCZ8/PHHtv0+/PBDDBkyBMeOHbN1VRGRffzxj39EcHAwFi5ciFdffRVarRZ9+/bF7Nmz5S6NiJwIw81lVFdXQ61WIyMj45JBiF5eXgCA0NBQaDSaFgEoJiYGQFPLD8MN0bV55JFHLpl/ZsKECZgwYcJlj0lLS2t1++VWFmeXFpHyMNxcRlxcHCwWC4qLi3HzzTe3us+IESPQ2NiInJwc22DH7OxsAEC3bt06rVYiIiL61XW9cGZ1dTVOnDgBoCnMLF68GKNHj0ZAQAAiIyPx0EMPYceOHXjrrbcQFxeHkpISpKamYuDAgbj99tthtVoxdOhQeHl5YcmSJbBarZgxYwZ8fHzw/fffy/zpiIiIrk/XdbhJS0vD6NGjL9k+ZcoUrFy5Eg0NDfj73/+Ojz/+GAUFBQgKCsKNN96Il19+GQMGDADQNM37k08+ie+//x6enp4YP3483nrrLQQEBHT2xyEiIiJc5+GGiIiIlIfz3BAREZGiMNwQERGRolx3d0tZrVacPXsW3t7elyysR0RERM5JCIGqqiqEhYVBpbpy28x1F27Onj17ycrARERE5Bry8/PRtWvXK+5z3YUbb29vAE0Xx8fHR+ZqiIiIqC0qKysRERFh+z1+JddduGnuivLx8WG4ISIicjFtGVLCAcVERESkKAw3REREpCgMN0RERKQoDDdERESkKAw3REREpCgMN0RERKQoDDdERESkKAw3REREpCgMN0RERKQoDDdERESkKAw3REREpCgMN0RERKQoDDdERESkKAw3REREpCgauQtQmry8PJSWltr9vEFBQYiMjLT7eYmIiJSG4caO8vLy0DcmBnW1tXY/t7uHB45mZTHgEBERXQXDjR2VlpairrYWk557A4bIHnY7b1FeDj5Z9CxKS0sZboiIiK6C4cYBDJE90LVXP7nLICIiui5xQDEREREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESmKrOFm27ZtmDBhAsLCwiBJEtatW3fF/deuXYsxY8agS5cu8PHxQWJiIjZt2tQ5xRIREZFLkDXc1NTUYNCgQVi6dGmb9t+2bRvGjBmDDRs2ICMjA6NHj8aECROwb98+B1dKRERErkIj55uPHz8e48ePb/P+S5YsafH89ddfx1dffYVvvvkGcXFxdq6OiIiIXJGs4eZaWa1WVFVVISAg4LL7mEwmmEwm2/PKysrOKI2IiIhk4tIDit98801UV1fjD3/4w2X3SUlJga+vr+0RERHRiRUSERFRZ3PZcLNq1Sq8/PLL+OyzzxAcHHzZ/ebOnYuKigrbIz8/vxOrJCIios7mkt1Sq1evxrRp0/D5558jKSnpivvq9Xro9fpOqoyIiIjk5nItN59++immTp2KTz/9FLfffrvc5RAREZGTkbXlprq6GidOnLA9z83Nxf79+xEQEIDIyEjMnTsXBQUF+PjjjwE0dUVNmTIF//znP5GQkACj0QgAcHd3h6+vryyfgYiIiJyLrC03e/fuRVxcnO027jlz5iAuLg7z588HABQWFiIvL8+2//vvv4/GxkbMmDEDoaGhtsesWbNkqZ+IiIicj6wtN6NGjYIQ4rKvr1y5ssXztLQ0xxZERERELs/lxtwQERERXQnDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESkKww0REREpCsMNERERKQrDDRERESmKrOFm27ZtmDBhAsLCwiBJEtatW3fVY9LS0nDDDTdAr9ejZ8+eWLlypcPrJCIiItcha7ipqanBoEGDsHTp0jbtn5ubi9tvvx2jR4/G/v378fTTT2PatGnYtGmTgyslIiIiV6GR883Hjx+P8ePHt3n/ZcuWITo6Gm+99RYAICYmBtu3b8fbb7+N5ORkR5VJRERELsSlxtykp6cjKSmpxbbk5GSkp6df9hiTyYTKysoWDyIiIlIulwo3RqMRBoOhxTaDwYDKykrU1dW1ekxKSgp8fX1tj4iIiM4olYiIiGTiUuGmI+bOnYuKigrbIz8/X+6SiIiIyIFkHXPTXiEhISgqKmqxraioCD4+PnB3d2/1GL1eD71e3xnlERERkRNwqZabxMREpKamtti2efNmJCYmylQRERERORtZw011dTX279+P/fv3A2i61Xv//v3Iy8sD0NSlNHnyZNv+f/7zn3Hy5En89a9/xdGjR/Gvf/0Ln332GWbPni1H+UREROSEZA03e/fuRVxcHOLi4gAAc+bMQVxcHObPnw8AKCwstAUdAIiOjsb69euxefNmDBo0CG+99Rb+/e9/8zZwIiIispF1zM2oUaMghLjs663NPjxq1Cjs27fPgVURERGRK3OpMTdEREREV8NwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIrCcENERESKwnBDREREisJwQ0RERIoie7hZunQpoqKi4ObmhoSEBOzevfuK+y9ZsgR9+vSBu7s7IiIiMHv2bNTX13dStUREROTsZA03a9aswZw5c7BgwQJkZmZi0KBBSE5ORnFxcav7r1q1Cs8//zwWLFiArKwsfPjhh1izZg3+9re/dXLlrSuuaYSkc5e7DCIiouuarOFm8eLFmD59OqZOnYrY2FgsW7YMHh4eWL58eav779y5EyNGjMCDDz6IqKgojB07FhMnTrxqa09nyMw7j79uOYegO/8KIeSuhoiI6PolW7gxm83IyMhAUlLSr8WoVEhKSkJ6enqrxwwfPhwZGRm2MHPy5Els2LABt91222Xfx2QyobKyssXDEdSShPpGKzx6DMXBcrVD3oOIiIiuTrZwU1paCovFAoPB0GK7wWCA0Whs9ZgHH3wQr7zyCm666SZotVr06NEDo0aNumK3VEpKCnx9fW2PiIgIu36OZoMi/PDkMD8AwIkqNQ6dqXDI+xAREdGVyT6guD3S0tLw+uuv41//+hcyMzOxdu1arF+/Hq+++uplj5k7dy4qKipsj/z8fIfVNyLCHeXb/q+p1uxi1JgaHfZeRERE1DqNXG8cFBQEtVqNoqKiFtuLiooQEhLS6jHz5s3Dww8/jGnTpgEABgwYgJqaGjz22GN44YUXoFJdmtX0ej30er39P8BlVKSvQVTSJJSbVcg9V4P+Yb6d9t5EREQkY8uNTqfDkCFDkJqaattmtVqRmpqKxMTEVo+pra29JMCo1U3jW4QTjeINdWuq5fS5WpkrISIiuv7I1nIDAHPmzMGUKVMQHx+PYcOGYcmSJaipqcHUqVMBAJMnT0Z4eDhSUlIAABMmTMDixYsRFxeHhIQEnDhxAvPmzcOECRNsIccZhLhbkVWpRl5ZLaxWAZVKkrskIiKi64as4eb+++9HSUkJ5s+fD6PRiMGDB2Pjxo22QcZ5eXktWmpefPFFSJKEF198EQUFBejSpQsmTJiA1157Ta6P0Cp/nYCbVoX6BisKK+oR7s+5b4iIiDqLrOEGAGbOnImZM2e2+lpaWlqL5xqNBgsWLMCCBQs6obKOkySgW4AnjhVV4dS5GoYbIiKiTuRSd0u5kqhADwAcd0NERNTZGG4cJPJCuCmpNvGWcCIiok7EcOMgHjoNDD5Nt6DnlbH1hoiIqLMw3DhQqG/TWJuSKpPMlRAREV0/GG4cKMhLBwAorWa4ISIi6iwMNw4U5NXULVVabXaqSQaJiIiUjOHGgQI8dZAA1DVYUGu2yF0OERHRdYHhxoG0ahX8PLQA2DVFRETUWRhuHOzirikiIiJyPIYbB2sON+fYckNERNQpOhRuunfvjnPnzl2yvby8HN27d7/mopTk1zum2HJDRETUGToUbk6dOgWL5dIBsiaTCQUFBddclJI0t9yU1ZhhsfKOKSIiIkdr18KZX3/9te3vmzZtgq+vr+25xWJBamoqoqKi7FacEni7aaBTq2C2WFFea0bghbBDREREjtGucHP33XcDACRJwpQpU1q8ptVqERUVhbfeestuxSmBJEkI9NKhsKIepdUMN0RERI7WrnBjtVoBANHR0dizZw+CgoIcUpTSBHnpL4QbE/rAW+5yiIiIFK1d4aZZbm6uvetQtMALg4rP1XBQMRERkaN1KNwAQGpqKlJTU1FcXGxr0Wm2fPnyay5MSfzcmybyq6hrkLkSIiIi5etQuHn55ZfxyiuvID4+HqGhoZAkyd51KYqfR1PLTUVdA4QQvF5EREQO1KFws2zZMqxcuRIPP/ywvetRJG+9BioJsFgFqk2N8HbTyl0SERGRYnVonhuz2Yzhw4fbuxbFUqkkW6Bh1xQREZFjdSjcTJs2DatWrbJ3LYrWPO6mnOGGiIjIoTrULVVfX4/3338fW7ZswcCBA6HVtuxmWbx4sV2KUxLf5kHFtQw3REREjtShcHPw4EEMHjwYAHD48OEWr3GwbOt8PdgtRURE1Bk6FG5++OEHe9eheLwdnIiIqHN0aMwNtZ/vRWNuhOACmkRERI7SoZab0aNHX7H7aevWrR0uSKmaw4250Yr6BivcdWqZKyIiIlKmDoWb5vE2zRoaGrB//34cPnz4kgU1qYlGrYKXXoNqUyMq6hoYboiIiBykQ+Hm7bffbnX7Sy+9hOrq6msqSMl83bWoNjWivM6MEF83ucshIiJSJLuOuXnooYe4rtQV8HZwIiIix7NruElPT4ebG1skLseXd0wRERE5XIe6pe65554Wz4UQKCwsxN69ezFv3jy7FKZEfh6cpZiIiMjROhRufH19WzxXqVTo06cPXnnlFYwdO9YuhSkRW26IiIgcr0PhZsWKFfau47rgc2HxzFqzBY0WKzRqTjNERERkbx0KN80yMjKQlZUFAOjXrx/i4uLsUpRSuWlV0KgkNFoFqkyN8PfQyV0SERGR4nQo3BQXF+OBBx5AWloa/Pz8AADl5eUYPXo0Vq9ejS5dutizRsWQJAnebhqcr21AVT3DDRERkSN0qF/kySefRFVVFX755ReUlZWhrKwMhw8fRmVlJZ566il716gozV1TVfUcd0NEROQIHWq52bhxI7Zs2YKYmBjbttjYWCxdupQDiq/C263pklfVN8pcCRERkTJ1qOXGarVCq9Vesl2r1cJqtV5zUUrmbWu5YbghIiJyhA6Fm1tuuQWzZs3C2bNnbdsKCgowe/Zs3HrrrXYrTomaW24q2S1FRETkEB0KN++++y4qKysRFRWFHj16oEePHoiOjkZlZSXeeecde9eoKOyWIiIicqwOjbmJiIhAZmYmtmzZgqNHjwIAYmJikJSUZNfilKi5W6ra1AghBCRJkrkiIiIiZWlXy83WrVsRGxuLyspKSJKEMWPG4Mknn8STTz6JoUOHol+/fvjpp58cVasieOmb8qTFKlBrtshcDRERkfK0K9wsWbIE06dPh4+PzyWv+fr64k9/+hMWL15st+KUSK2SbAGHXVNERET2165wc+DAAYwbN+6yr48dOxYZGRnXXJTS/TruhoOKiYiI7K1d4aaoqKjVW8CbaTQalJSUtKuApUuXIioqCm5ubkhISMDu3buvuH95eTlmzJiB0NBQ6PV69O7dGxs2bGjXe8qNg4qJiIgcp13hJjw8HIcPH77s6wcPHkRoaGibz7dmzRrMmTMHCxYsQGZmJgYNGoTk5GQUFxe3ur/ZbMaYMWNw6tQp/Pe//8WxY8fwwQcfIDw8vD0fQ3ac64aIiMhx2hVubrvtNsybNw/19fWXvFZXV4cFCxbgjjvuaPP5Fi9ejOnTp2Pq1KmIjY3FsmXL4OHhgeXLl7e6//Lly1FWVoZ169ZhxIgRiIqKwsiRIzFo0KD2fAzZ2VpuTOyWIiIisrd2hZsXX3wRZWVl6N27N/7xj3/gq6++wldffYVFixahT58+KCsrwwsvvNCmc5nNZmRkZLS4fVylUiEpKQnp6emtHvP1118jMTERM2bMgMFgQP/+/fH666/DYrn8XUcmkwmVlZUtHnL7dSI/ttwQERHZW7vmuTEYDNi5cycef/xxzJ07F0IIAE2rXScnJ2Pp0qUwGAxtOldpaSksFssl+xsMBtvcOb918uRJbN26FZMmTcKGDRtw4sQJPPHEE2hoaMCCBQtaPSYlJQUvv/xyOz6l43HxTCIiIsdp9yR+3bp1w4YNG3D+/HmcOHECQgj06tUL/v7+jqivBavViuDgYLz//vtQq9UYMmQICgoK8MYbb1w23MydOxdz5syxPa+srERERITDa72S5pab+gYrGixWaNUdmiiaiIiIWtGhGYoBwN/fH0OHDu3wGwcFBUGtVqOoqKjF9qKiIoSEhLR6TGhoKLRaLdRqtW1bTEwMjEYjzGYzdDrdJcfo9Xro9foO1+kIeo0aOrUKZosVVfWNCPC8tG4iIiLqGNmaDHQ6HYYMGYLU1FTbNqvVitTUVCQmJrZ6zIgRI3DixIkWK49nZ2cjNDS01WDjzDjXDRERkWPI2h8yZ84cfPDBB/joo4+QlZWFxx9/HDU1NZg6dSoAYPLkyZg7d65t/8cffxxlZWWYNWsWsrOzsX79erz++uuYMWOGXB+hw7wuhJtqEwcVExER2VOHu6Xs4f7770dJSQnmz58Po9GIwYMHY+PGjbZBxnl5eVCpfs1fERER2LRpE2bPno2BAwciPDwcs2bNwnPPPSfXR+iw5iUYqnnHFBERkV3JGm4AYObMmZg5c2arr6WlpV2yLTExET///LODq3I8W7hhyw0REZFd8TYdmXjZJvJjuCEiIrInhhuZeLPlhoiIyCEYbmTCMTdERESOwXAjk+ZuKVNj00R+REREZB8MNzLRqVXQqiUA7JoiIiKyJ4YbmUiSBG990xpT7JoiIiKyH4YbGXEiPyIiIvtjuJFR86Bi3g5ORERkPww3MuIdU0RERPbHcCMjdksRERHZH8ONjNhyQ0REZH8MNzLi+lJERET2x3AjI+8L3VJ1DRY0ciI/IiIiu2C4kZFeo4JGxYn8iIiI7InhRkaSJLFrioiIyM4YbmTGO6aIiIjsi+FGZt68Y4qIiMiuGG5k1txyw1mKiYiI7IPhRmac64aIiMi+GG5kxgHFRERE9sVwIzMOKCYiIrIvhhuZNbfc1JotsFiFzNUQERG5PoYbmblr1VBLTRP51bD1hoiI6Jox3MhMkiTeMUVERGRHDDdOgHdMERER2Q/DjRPgoGIiIiL7YbhxArwdnIiIyH4YbpxA8xIMVfUNMldCRETk+hhunAC7pYiIiOyH4cYJsFuKiIjIfhhunEBzuKkxcSI/IiKia8Vw4wQ8dGqomubxQ62ZrTdERETXguHGCUiSBE92TREREdkFw42T4ER+RERE9sFw4yS8uQQDERGRXTDcOAneMUVERGQfDDdOgt1SRERE9sFw4yQ4kR8REZF9MNw4CW+9FgBQxZYbIiKia8Jw4yRsE/mZG2EVnMiPiIiooxhunISHXg1JAoQAas0WucshIiJyWQw3TkIlSfDUcVAxERHRtWK4cSK8HZyIiOjaMdw4Ed4xRUREdO2cItwsXboUUVFRcHNzQ0JCAnbv3t2m41avXg1JknD33Xc7tsBO4s25boiIiK6Z7OFmzZo1mDNnDhYsWIDMzEwMGjQIycnJKC4uvuJxp06dwjPPPIObb765kyp1PC/bEgwNMldCRETkumQPN4sXL8b06dMxdepUxMbGYtmyZfDw8MDy5csve4zFYsGkSZPw8ssvo3v37p1YrWNxlmIiIqJrJ2u4MZvNyMjIQFJSkm2bSqVCUlIS0tPTL3vcK6+8guDgYPzxj3+86nuYTCZUVla2eDgrDigmIiK6drKGm9LSUlgsFhgMhhbbDQYDjEZjq8ds374dH374IT744IM2vUdKSgp8fX1tj4iIiGuu21EuHlAsOJEfERFRh8jeLdUeVVVVePjhh/HBBx8gKCioTcfMnTsXFRUVtkd+fr6Dq+y45nlurJzIj4iIqMM0cr55UFAQ1Go1ioqKWmwvKipCSEjIJfvn5OTg1KlTmDBhgm2b1WoFAGg0Ghw7dgw9evRocYxer4der3dA9fanVknw1KlRY7ag2tQIT72sXx4iIiKXJGvLjU6nw5AhQ5CammrbZrVakZqaisTExEv279u3Lw4dOoT9+/fbHnfeeSdGjx6N/fv3O3WXU1txrhsiIqJrI3vTwJw5czBlyhTEx8dj2LBhWLJkCWpqajB16lQAwOTJkxEeHo6UlBS4ubmhf//+LY738/MDgEu2uyovvQZFMPGOKSIiog6SPdzcf//9KCkpwfz582E0GjF48GBs3LjRNsg4Ly8PKpVLDQ26JrxjioiI6NrIHm4AYObMmZg5c2arr6WlpV3x2JUrV9q/IBl5u2kBMNwQERF11PXTJOIiOJEfERHRtWG4cTLN4aaKLTdEREQdwnDjZDiRHxER0bVhuHEyzS03FqtAXQMn8iMiImovhhsno1ZJ8NSrAQCVHHdDRETUbgw3Tsjnwh1TVXUNMldCRETkehhunJD3hXE3VWy5ISIiajeGGyfUPNdNZT1bboiIiNqL4cYJseWGiIio4xhunJBtzA3DDRERUbsx3Dih5pYbdksRERG1H8ONE2puuTE1WmFutMpcDRERkWthuHFCOo0Kek3Tl4atN0RERO3DcOOkOKiYiIioYxhunNSvg4rZckNERNQeDDdO6tdBxWy5ISIiag+GGyfFlhsiIqKOYbhxUhxzQ0RE1DEMN07KmxP5ERERdQjDjZNqbrmpNjXCKmQuhoiIyIUw3DgpD50aapUEAKi1yFwMERGRC2G4cVKSJNlab2obJZmrISIich0MN07M171p3E0Nww0REVGbMdw4MV83hhsiIqL2YrhxYr+23MhcCBERkQthuHFivh5suSEiImovhhsn5sNuKSIionZjuHFizd1SZqsESe8pczVERESugeHGiek0Krhr1QAArV+IzNUQERG5BoYbJ+d3YdyNhuGGiIioTRhunJyPO8MNERFRezDcODlfhhsiIqJ2Ybhxcr+Gm1CZKyEiInINDDdOji03RERE7cNw4+Rs4canCxqtQuZqiIiInB/DjZPz1KmhkgQklRqltRa5yyEiInJ6DDdOTpIkeGqaWmyM1Qw3REREV8Nw4wI8m+bxQ1E1V9AkIiK6GoYbF+ClbWq5OcuWGyIioqtiuHEB3he6pc5WseWGiIjoahhuXID3hZabAoYbIiKiq2K4cQHN4aa4xoL6BnZNERERXQnDjQvQqwBrfTWsAjh9rlbucoiIiJyaU4SbpUuXIioqCm5ubkhISMDu3bsvu+8HH3yAm2++Gf7+/vD390dSUtIV91cCSQIays4AAHJKqmWuhoiIyLnJHm7WrFmDOXPmYMGCBcjMzMSgQYOQnJyM4uLiVvdPS0vDxIkT8cMPPyA9PR0REREYO3YsCgoKOrnyztVwLh8AcKKY4YaIiOhKZA83ixcvxvTp0zF16lTExsZi2bJl8PDwwPLly1vd/5NPPsETTzyBwYMHo2/fvvj3v/8Nq9WK1NTUTq68czWcY8sNERFRW8gabsxmMzIyMpCUlGTbplKpkJSUhPT09Dado7a2Fg0NDQgICHBUmU6B4YaIiKhtNHK+eWlpKSwWCwwGQ4vtBoMBR48ebdM5nnvuOYSFhbUISBczmUwwmUy255WVlR0vWEa2MTfFNbBaBVQqSeaKiIiInJPs3VLXYuHChVi9ejW+/PJLuLm5tbpPSkoKfH19bY+IiIhOrtI+GsuNUEtAXYMFxsp6ucshIiJyWrKGm6CgIKjVahQVFbXYXlRUhJCQkCse++abb2LhwoX4/vvvMXDgwMvuN3fuXFRUVNge+fn5dqm901ktCPVuamhj1xQREdHlyRpudDodhgwZ0mIwcPPg4MTExMse949//AOvvvoqNm7ciPj4+Cu+h16vh4+PT4uHqwr3blpBM4d3TBEREV2WrGNuAGDOnDmYMmUK4uPjMWzYMCxZsgQ1NTWYOnUqAGDy5MkIDw9HSkoKAGDRokWYP38+Vq1ahaioKBiNRgCAl5cXvLy8ZPscnSHcWwPAhJySGrlLISIiclqyh5v7778fJSUlmD9/PoxGIwYPHoyNGzfaBhnn5eVBpfq1gem9996D2WzGvffe2+I8CxYswEsvvdSZpXe6cJ+mLxfnuiEiIro82cMNAMycORMzZ85s9bW0tLQWz0+dOuX4gpxUpK8WAHDUWAkhBCSJd0wRERH9llOEG2qbSB8NNCoJ52sbUFhRjzA/d7lLInI6eXl5KC0tdci5g4KCEBkZ6ZBzE5H9MNy4EK1aQs9gLxw1VuHI2UqGG6LfyMvLQ9+YGNTVOmaBWXcPDxzNymLAIXJyDDcuJjbMpyncFFYiKdZw9QOIriOlpaWoq63FpOfegCGyh13PXZSXg08WPYvS0lKGGyInx3DjYmJDfbAWBfjlbIXcpRA5LUNkD3Tt1U/uMohIJi49Q/H1qF+YLwDgSKFrLiNBRETkaAw3LiY2tGkSwvyyOlTUNchcDRERkfNhuHExvh5ahF8YSJzF1hsiIqJLMNy4oH5hTa03R84y3BAREf0Ww40Lim0ON2y5ISIiugTDjQtqHnfzC1tuiIiILsFw44L6hTfdMXW8qAqmRovM1RARETkXhhsXFObrhiAvHRqtAocLON8NERHRxTiJnwuSJAnx3QKw8Rcjdueex5BuAXKXRG3ANY+IiDoHw42LGhrdFG72nioDYN9p5sn+uOYREVHnYbhxUUOj/AEAe0+fh9UqoFJJMldEV8I1j4jI2Si5NZnhxkXFhvrAQ6dGRV0DjhdXo0+It9wlURtwzSMicgZKb01muHFRGrUKcZF+2HHiHPacKmO4ISKiNlN6azLDjQuL7xaAHSfOYe+pMjx0Yze5yyGSjRACx4qqsKugHu49h6G4XoLBYoVWzRtCia5Eqa3JDDcubFh0011Se06dl7kSInmUVJnwr7QT2HTYiLMV9QCA4N/Px0/FQHrpSXQL9MDgCD909feQuVIi6kwMNy5scIQf1CoJBeV1KCivsy2oSaR0DRYr/v1TLpb+cALVpkYAgF6jQqSPGocPH4ZfRG/UWYCckhrklNSgt8ELN/fsAi83/sgjuh6wzdaFeeo1tkU003POyVwNUecorTZh0r93YdHGo6g2NWJgV1/8e3I8DiwYi0VJQTD+318wPqwBE4dFYEC4LyQA2UXV+GTXaZw575jBk0TkXBhuXNyo3l0AAKlZRTJXQuR4hwsqMOGd7didWwYvvQZv3jcI654YgaRYA9y0att+kgQEe7vhlr7BeGBYBIK99ahvtOLLfQWc1ZvoOsA2WheXFGvA/2w9gW3ZJTA1WqDXqK9+EF2RI+Z+yMrKsuv5rkeZeecx5cPdqDI1onsXT7z/cDx6Bntd9bhgbzfcN6QrNmcVIbuoGqlHi2FutOKGbv6dUDURyYHhxsX1D/OFwUePokoTfj5ZhpEXWnKoYxw990N1dfVV97FYBYoq63Gu2oxqU6MttLppVejirUeIjxs019ldQHtOleGR5btRY7ZgWFQA/v1IPHzctG0+XqNWYVy/EPi5l2H3qTL8dKIUkgTERTLgECkRw42LU6kk3NLXgE9352HLkSKGm2vkqLkfsnb/iO8++ifq6+tbfb3RakVuSQ2yjFU4c74WDRZx2XOpVRIiAzwwINwX3QKVfxfQwTPltmAzvEcg/j0lHh669v/okiQJiT0CAQC7T5Vh2/FSaNQqDAj3tXfJRCQzhhsFGBMbjE935yE1qwiv3NUPksSlGK6Vved+KMrLaXV7XYMF+/PLcehMBeoaLLbt7lo1DD56eLtp4aZVwdRoRY2pEYUV9ag1W5BbWoPc0hr4umvR20MFQJlf85Ml1XhkxR5bsFn+yNAWY2s64sbuAbAKgb2nz+OHo8Xw1KnRvcvVu7eIyHUw3CjA8B5BcNeqcbaiHkcKK9EvjP8TdXZV9Q3Yl1eOQwUVaLQ2tdJ46tWIDfVBr2BvBHnpWg2pQgiU1ZhxpLASv5ytREVdA/bUaRAy5W1knzPjhs7+IA5krKjHwx/uRlmNGQPCffH+5PhrDjZAUwvO8B6BqDVbcKSwEt8dNuL3N3RFiK+bHaomun4JIVDfaEW5WYI2OFrWWhhuFMBNq8ZNvYKw+UgRNv1SxHDjxGqtamzJKkJWYSUuZBp08dYjvps/enbxuuoCqJIkIdBLj5t7dcGN3QOxL78ce06WAiE98bet55DbmIXZSb3tEgLkVF5rxuTlu1BQXofuQZ5YOXUovPT2+3ElSRJu6RuMGnMjTp+rxdcHzuIP8V3h56Gz23sQXQ8aLFbklFTjeFE18m1d6loE3DJN1roYbhTi9gGh2HykCF9knMGsW3tBzVXCWyitNiG7qApnyupgrKzHuWoTaswW1DdYoJIkaNUq+LhrYKqohme/W1BmkhDcaIVOc+0Dd4UQKIcHgu56Hnvqg4GzlQCAcD93DI3yR2SAR4e6ErVqFYZFBSCwvhCrN22HV/9b8L8/nsSWI0V4875BLjtYts5swaMr9yC7qBoGHz0+enQYAr30dn8ftUrCbf1D8UXmGRRXmbBuf1PA6ch4HqLrTaPVikNnKrDn1PkWXeoAoFcJ1NbXyFRZE/4rVohx/UPg940WBeV1+DG7GLf0NchdkiysVoHcczXYl1eOg2fKccxYhePF1SirMbf5HEF3zMEPRcAPRTnw0msQ6KlDoJcOXbz0CPTSI8BTd9XwKIRAUZUJuSU1yC6uQjm6wbNv0/pf0UGeGBrlj1Bf+8worVcD59Yvxt//9Ht8eLAWOSU1+P17O/HY/+uB2WN6udT0AA0WKx7/JAOZeeXwddfi40cTEBHguEHTOo0Kdw4Kw2d781FR14CvD5zF72/oyjWpiK7gzPlabD5ShMr6ptnBfdw06Bvigx7Bngjw0MF4MguLU14H5v1ethoZbhTCTavGfUO64oOfcvHJz3nXTbipb7BgX145dueWITPvPPbnl6OiruGS/SQJiAzwQHSQJ0J83BDkpYeHXg03jRpWIdBgEaisb8CxU2ex4cef4Rs9ECarhGpTI6pNjThd9uut4SoJ8PfQwdtNA283LbRqCRqVCg0WK+obLCiva0BptanFHU9qWFC+73uMShyCmwaNcMi1GBbuhgdujcfL3xzBl/sKsOzHHKQdK8bb9w9GTKiPQ97TnqxWgb/+9yDSjpXATavC8kfiO2W1e0+9BncPDsdne/NRVGnCd4eNuGNA6FW7CImuN41WK3aeOId9+eUAAC+9BgnRAYgJ9XG63gKGGwWZOCwSH/yUi63HinHmfK0iFwusqm9Axunz2J1bht25ZThwpvyS26b1mqbbewdH+CEm1Ae9Dd7oGewFd93VWzAyM2ux4vEXMGnpWgRF9UVZjRllNWaUVptQUm1CaZUZZosV52rMOHeV1iCdWoVugU2BqvLINqz5fim8Rrx/TZ//avw8dHj7/sEY1z8Ef1t7CEeNVbjr3R34y9jemHZzd6f7AdRMCIFXvm0KZRqVhPcmDcGQbgGd9v7+njpMGBSGtfsKkFtagx+yi3FLn2DeeUh0QXV9I9YfKoSxsmk6i/5hPri5Vxe7dN07AsONgnTv4oURPQOx48Q5rNmTj7+M7SN3Sdesqr4B6TnnsOtCmPnlbIVtIG6zYG89EroHIr6bP26I9EffUG+7dCu4adUI83NH2EULkgohUFXfiLJaM6rqm1p1Gi1WNFoFtGoV3DQqeLtp0cVbDz8PLVQXfjlmHLFecz3tkdwvBDdE+mPu2oPYklWMlO+OIvVoMd64dyC6BXp2ai1XI4TAy98cwcqdpwAAb9w3EKP7Bnd6HWF+7hjXLwTrDxXicEElvN20GBbVeQGLyFmdLa/D+kOFqDVboNeoMDbW4PTTJzDcKMykhG7YceIc/u/n05h2U3f4erR9Fldn0Gix4mBBBbYfL8VPx0uQmVcOy2/STGSAB4ZFB2BYdAASogM6PCC3IyRJgo+7Fj7uzn9du3jr8cHkeKzZk49Xvj2C3bllGPv2Njyd1BvTbo52inElVqvAy9/8go/ST0OSgIX3DMDv4rrKVk/PYC+M6t0FadklSM85B2+9xiW69Igc5XhRFTYdKYLFKhDopcMdA0Jd4q5ChhuFGRtrQG+DF7KLqvHO1uN48Y5YuUu6qtJqE1KzipB2rAQ7TpTaBqk1iwr0wPCeQUi4EGhCfd0vrP9UgnOnS3DutP1qUdoaUJIk4YFhkUjsEYi5aw9hZ845LNp4FGszz+CF22Mwqk/nt5A0q2+wYPaa/fjusBGSBCy6ZyD+MDRCtnqaDYrwQ1V9IzLyzmNLVhHctWpEBTlXa1d7OGKtNAAICgpCZGSk3c9LzkEIgcy8cmw/0fS90z3IE8n9Qpy2G+q3GG4URqNW4YXbYzFl+W58lH4Kk27shmgn/MGcW1qD738xYvORImTknYe4qHHGx02DET2DcHOvLri5V9Ald8s4ev0noG1rQLmSboGe+GRaAr7ILMBr64/geHHTzL839wrC00m9MaSTF5E0VtTjiQt3RenUKrxx30DcNTi8U2u4khE9A1FlakB2UTW+PViI8QNCYP+b0R3Pkf9W3D08cDQriwFHgaxWgbTsEhwqqAAADOrqi//Xu4utm90VMNwo0MjeXTCydxf8mF2ClA1ZeH9yvNwlwWoVOHCmHJuPFOH7I0U4UdwyPAwI98WtMcEY2bsLBnb1u+LAV0et/wRcfQ0oVyZJEu4d0hVjYgx4Z+txfJR+Cj8dL8VPx0txU88gPHpTFEb1Dnb4XUKbfjHiuS8Oory2AT5uGrw/OR43dg906Hu2lyRJGBsbAqsw4kRxNdYfKkR8gGv8j/Vijvq3UpSXg08WPYvS0lKGG4UxNVqw8bARp841BeL/1yvIJefMYrhRqBdvj8H2E6X4/kgR1uzJw/1DO/8HkLnRip9PnsOmCy00xVUm22saVdMihmNiDUiKMbQYtNtW9l7/Cbj8GlBK4uuhxYt3xGJyYhSW/nACX2SewfYTpdh+ohSRAR74/Q1dcdfgMLt3xeSX1WLhxqNYf7AQANA/3Af/80Cc0w5MVKskjO8Xgs2qIhw1VmHPOQ18b5oEq7j8oqbOyhH/Vkh5ymvN+OZgIcpqzFCrJIzrF4Kewc757/NqGG4UqpfBG7Nu7YXFm7Px4rrD6NHFC/GdcOdHjakRP2aXYNMvRmw9Woyqi8bPeOk1GNmnC8bGGjCqTzB8XWBQrpJFBnpg0b0DMfOWnvho5yl8tjcfeWW1eHtLNt7eko3YUB8MCdWjj68VvQKa5vPpiNzyBnyfU4vU3Fo0WpuW+LyzjyeeGhXhtMGmmUolYWysAR46NTLzyuE3YiIW7TiP/+1rRoCncw+qtFoFzlY1wqPPCBypUOHIL0ZU1jeivsECU6MVViEgoakr212rhodODV8PLfw9dAj21qOLl55z/VxH8spqseFQIUyNVnjq1bhjQJhLr7fGcKNgM0f3xFFjJTYcMuLP/8nA6sduRM9g+0+KVlZjxpasInz/ixE/HS+FqfHX256DvPQYE2tAcj8DEnsEutRsudeLiAAPvHhHLP4ytg82HCrEVwfOYseJUhwprMSRpkYWiEYzTMbjMBedRENxLhrOF6KxwghLTTlguWjSRLUGGl8DtP5h0Ef0h3vUYOgMv3aH1J3ah/Nbl+N/SnLxgYuM2ZAkCTf36gJVdSn2FAvsOQuMfftH/P3u/hjXP1Tu8mxqzY3Yn1eOvafPY+/p89h3+jyqTI3ocvdcZFUAqKi6zJGWVie+1KolGHzcEObrjjA/N4T5uTvFHXZkX0IA+/PLse14CYQADD563DEwzK5rucnBtaunK1KpJLx53yCcKq3FkcJK3PXuDrz1h6YJ3q6F1SpwpLASP2aXYFt2CfaePt/idu3IAA8k9zNgcJCEYHUt1KoGoOYMfjl45lo/EgDl3dHkLNx1avx+SFf8fkhXnKs24aPv9yDlo2/gFzMcDRod3Lr2g1vXS7s2JAioJcAigKa2gEtfD/OworuXFV0i+kH6f2+55JiNbl5WfLP4WSTMWoq8CjP+/J9MDO8RiGeS++AGGcYkFFXWY++p89h7ugwZp8/jl7OVl0yboFMDVfnZ6NOrB8JCguHjpoW7Tg2dRgX1hcGhDRYr6hosqDY1ory2AWU1ZhRV1sPUaMWZ83U4c74OQNPM3CG+bojw94BbvQSo+OvD1Uk6D+w5p0Z+bQkAICbEG7f0DYZGASGW350K56HT4OM/DsOMTzKxK7cMf/5PBu6JC8fjo3qgl6FtrTiNFiuOGquwP78ce0+VYfuJUpRWt5ydNzbUB8n9QpDc34A+Bm/k5+fzjiYXFuilx6goD/zl63/g4bFr4dW1G4wV9SipNuFctRkVdQ2orm+ERQgISGi86HeqVt00F1CIjxu6+rsjMsBDMYtRmoty8EZSELaVeeKDbbnYmXMO9/xrJ4ZFB+D++AiMHxDikM9aUdeA7KIqHC2sRGZeOfaeLkN+Wd0l+4X5umFIVACGRPohPioAdYUnMGzoHXhw6Vp0bUe3tBAC52rMKCyvx9mKOhSU16GqvhFny+txtrwegBYRsz7Fq9vKML46B8N7BCE21IfdWC7kl2ITwh59F/m1akgAbuoVhLgIP8XMyq2Mnzh0RUFeevxnWgJSNhzF8h25WLuvAGv3FWBYVACGRvvDoGuAMFXDXSPB1ChQ0yBQUmOBsaYRJ883IOd8A8wtF32Fm0bCgGAdBofocUOIHgYvDYAq1J6twr6zTa0rvKNJGaQLa2n5e+gQc9F2IQTMFivMjVZYrAJqlQTNhVmalfIDsjVatYRnk/ti4rBI/HPLcXyReca2HMjctYcwpJs/EnsEom+IN3obvBHi6wY37eW7Y61WgWpzIyrrGlBUacLZ8jrb43RZLY4Zq1BYcen3ukoC+ob4ID7KH/FRAYjv5n/JwPzMoo59HSRJQpCXHkFeegzo6gshBCrqGpB/vg75ZbU4XVoFs84d+4wm7NtwFEDTFA4xoT7oG+KNvqE+6BPijW4BHvD30DH0OJGSKhNSvsvC2swyaHyD4akRuG1QRIdu6nBmDDfXCa1ahfkTYnHX4DC8l5aDTUeM2H2qDLtPlbXpeGt9NUyF2TAVZqP+1H6YCo7imLUR/73KcZ4BBt7RpFCSJEGvUV+346i6+nvgjfsGYc7Y3vjv3jP4POMM8spqkX7yHNJPnmuxr5deAzetCmqVBLUkQa2WYLUClfUNqDY1oi03YAV5qBDpq0VPfy36BunQO1ALD60KgBmwGGE8aYTxN8fYqwtXkiT4eejg56HDgHBf5GeX4d2X/4K//XMlTte7YXduGSrrG7Ertwy7clv+TNGqJXTx0qOLjxu6eOngodPAQ6eGm1YNd50aHhf+dNM2P1SoLi9DfU0VdGoJOrUED60ET60KHlrpmtdHu14nH6yobcCHO3KxYnsuqkyNkABU7vsOd064VXHBBnCScLN06VK88cYbMBqNGDRoEN555x0MGzbssvt//vnnmDdvHk6dOoVevXph0aJFuO222zqxYtc1KMIPyx4egvyyWuzMKcXGjBP4bkcm/MO7A2od1BKgVQm4qwFPjYCPViBAb4W3Rgepd38A/QHcc9X3YesKXS9Cfd3x5K29MPOWnjhZWoPtx0txIL8cR41VOFlajfoG64XV5a98HmFpgKW6DI2VJbBUlqKxsgSNFUVoKD0Nc8lpnDbXIqODNdq7C1eSgIaSXNzZxws33HADGixWHDNW4aixCseMlRf+rEJxlQkNFoGzFfU420rrU3sJYYXVVAtrXRWs9VWw1lfDWlcJS23lhT8rYK2tgKW2Apa6ClhrK2GtrwbErzc5XG+TDx4vqsInu/LwRcYZVJma7l4dEO6Lh/pq8MCipdDedavMFTqG7OFmzZo1mDNnDpYtW4aEhAQsWbIEycnJOHbsGIKDL50afufOnZg4cSJSUlJwxx13YNWqVbj77ruRmZmJ/v37y/AJXFNEgAfuD4hEL3UpVj7+LCYtXYuuvXrb7fxsXaHrjSRJ6NHFCz0uur1dCIEqUyNKq0wwW6xotAhYhUCjtWnotY+7FqeyszBm1E2Y+JfXEDKgBwB/APb5t9hZ/8nQqlXoH+6L/uG+LbabG60orTahuMqEosp6nKs2o67BgvoGC+rMFtSaLbbnteZG1DdYUXq+ApkHDsM/LBKSRg+LABqsgEVIkCQV1G5eULt5AWjrnWoCOhWgVwFSYy0Kj+3HM6v3oltILnx0Kvi4qeCjb3r46lXw1qk6NO2Bs7QINY+R/OFoMTYdMeJwQaXttT4Gb8xK6oVx/UKwf/8+Gat0PNnDzeLFizF9+nRMnToVALBs2TKsX78ey5cvx/PPP3/J/v/85z8xbtw4PPvsswCAV199FZs3b8a7776LZcuWdWrtnc0RdwnxziPlsPfXkt8brbPXdREAKgAYT2VDNJoRosBJKXUaFcL83NvV7ZGZmYkhz8xpGgR90X+4LFZhm6OnvsGC+kYLTA1Nd3rVXQhJv/2zaVoKCWYrYLYCgCc8+4zA7jJgd9nlW7NEo7mphchUC6u5FsJU8+tzUw2Eudb2XDTUw9pogk4FfLziQ0R1DYW79tdut+a/X2t32sUaLU2tgedqzDhbXof8sjocL25qLTuQX46aiwZJqlUSbu0bjAcTIvH/enW5bsY/yRpuzGYzMjIyMHfuXNs2lUqFpKQkpKent3pMeno65syZ02JbcnIy1q1b58hSZVVZ1nSb3kMPPeSw9+CdR67L0d8fjvzecKVA5srX2VE68z9capUET70Gnu1Y5Ks5EDW3EB3Ztxvpm79Bn5tvh3dQGEwWCSYrYLIC5gt/ByRIGh3UGh3Unn7tqv2pdScBnGz1Na2qaWZ2jarps2hUgFpq+lOjkqC+8FySAKtomn/GKgSsAlCp1VBrtKgxNQ08r/ntHR6/4a3XIKF7AMbGhuCWmGAEebniymjXRtZwU1paCovFAoPB0GK7wWDA0aNHWz3GaDS2ur/R+NuhdE1MJhNMpl87uysqmhYCq6ysbHX/a9H8w+nM8V9gqrPfLdCnjjQ1Hw4dfz+6Rvey23kBIC/7EDK2fIVTRw/adYKu5v8xGk9lI8fT4yp7K//cjqzZUd8fjvreAIBTWU01OyooZB/ca9d/g4BrXmdHfd85+usHOOZraM3bh6rMbxEwqC96RVw6FYYQQKMAGgTQaJXQYG163vz3FtutQIOQYAVQU12N8+dKodLqIGn1kDQ6SBo9VNpfZ/g1XXjYk5tGQqC7CkEeaoR5a9DVW4Nofy3CvdVQq+qAulwcycxt9dhjx44BsP/vKwAoOdP0ntXV1Xb9Xdt8LtGWEfhCRgUFBQKA2LlzZ4vtzz77rBg2bFirx2i1WrFq1aoW25YuXSqCg4Nb3X/BggUCTS3AfPDBBx988MGHiz/y8/Ovmi9kbbkJCgqCWq1GUVFRi+1FRUUICWl9Ft2QkJB27T937twW3VhWqxVlZWUIDAy0+1wclZWViIiIQH5+Pnx8fOx6bvoVr3Pn4HXuHLzOnYfXunM46joLIVBVVYWwsLCr7itruNHpdBgyZAhSU1Nx9913A2gKH6mpqZg5c2arxyQmJiI1NRVPP/20bdvmzZuRmJjY6v56vR56fcv+Rj8/P3uUf1k+Pj78h9MJeJ07B69z5+B17jy81p3DEdfZ19e3TfvJfrfUnDlzMGXKFMTHx2PYsGFYsmQJampqbHdPTZ48GeHh4UhJSQEAzJo1CyNHjsRbb72F22+/HatXr8bevXvx/vvvy/kxiIiIyEnIHm7uv/9+lJSUYP78+TAajRg8eDA2btxoGzScl5cHlerXQXbDhw/HqlWr8OKLL+Jvf/sbevXqhXXr1nGOGyIiIgLgBOEGAGbOnHnZbqi0tLRLtt1333247777HFxV++n1eixYsOCSbjCyL17nzsHr3Dl4nTsPr3XncIbrLAnRlnuqiIiIiFyDfSdVICIiIpIZww0REREpCsMNERERKQrDDRERESkKw42dLF26FFFRUXBzc0NCQgJ2794td0kuLSUlBUOHDoW3tzeCg4Nx991329ZCaVZfX48ZM2YgMDAQXl5e+P3vf3/J7NXUPgsXLoQkSS0myeR1tp+CggI89NBDCAwMhLu7OwYMGIC9e/faXhdCYP78+QgNDYW7uzuSkpJw/PhxGSt2PRaLBfPmzUN0dDTc3d3Ro0cPvPrqqy3WI+J1br9t27ZhwoQJCAsLgyRJlyxW3ZZrWlZWhkmTJsHHxwd+fn744x//6LgFY6+6QANd1erVq4VOpxPLly8Xv/zyi5g+fbrw8/MTRUVFcpfmspKTk8WKFSvE4cOHxf79+8Vtt90mIiMjRXV1tW2fP//5zyIiIkKkpqaKvXv3ihtvvFEMHz5cxqpd2+7du0VUVJQYOHCgmDVrlm07r7N9lJWViW7duolHHnlE7Nq1S5w8eVJs2rRJnDhxwrbPwoULha+vr1i3bp04cOCAuPPOO0V0dLSoq6uTsXLX8tprr4nAwEDx7bffitzcXPH5558LLy8v8c9//tO2D69z+23YsEG88MILYu3atQKA+PLLL1u83pZrOm7cODFo0CDx888/i59++kn07NlTTJw40SH1MtzYwbBhw8SMGTNszy0WiwgLCxMpKSkyVqUsxcXFAoD48ccfhRBClJeXC61WKz7//HPbPllZWQKASE9Pl6tMl1VVVSV69eolNm/eLEaOHGkLN7zO9vPcc8+Jm2666bKvW61WERISIt544w3btvLycqHX68Wnn37aGSUqwu233y4effTRFtvuueceMWnSJCEEr7M9/DbctOWaHjlyRAAQe/bsse3z3XffCUmSREFBgd1rZLfUNTKbzcjIyEBSUpJtm0qlQlJSEtLT02WsTFkqKioAAAEBAQCAjIwMNDQ0tLjuffv2RWRkJK97B8yYMQO33357i+sJ8Drb09dff434+Hjcd999CA4ORlxcHD744APb67m5uTAajS2uta+vLxISEnit22H48OFITU1FdnY2AODAgQPYvn07xo8fD4DX2RHack3T09Ph5+eH+Ph42z5JSUlQqVTYtWuX3WtyihmKXVlpaSksFottuYhmBoMBR48elakqZbFarXj66acxYsQI2zIbRqMROp3ukkVQDQYDjEajDFW6rtWrVyMzMxN79uy55DVeZ/s5efIk3nvvPcyZMwd/+9vfsGfPHjz11FPQ6XSYMmWK7Xq29rOE17rtnn/+eVRWVqJv375Qq9WwWCx47bXXMGnSJADgdXaAtlxTo9GI4ODgFq9rNBoEBAQ45Loz3JDTmzFjBg4fPozt27fLXYri5OfnY9asWdi8eTPc3NzkLkfRrFYr4uPj8frrrwMA4uLicPjwYSxbtgxTpkyRuTrl+Oyzz/DJJ59g1apV6NevH/bv34+nn34aYWFhvM7XEXZLXaOgoCCo1epL7h4pKipCSEiITFUpx8yZM/Htt9/ihx9+QNeuXW3bQ0JCYDabUV5e3mJ/Xvf2ycjIQHFxMW644QZoNBpoNBr8+OOP+J//+R9oNBoYDAZeZzsJDQ1FbGxsi20xMTHIy8sDANv15M+Sa/Pss8/i+eefxwMPPIABAwbg4YcfxuzZs5GSkgKA19kR2nJNQ0JCUFxc3OL1xsZGlJWVOeS6M9xcI51OhyFDhiA1NdW2zWq1IjU1FYmJiTJW5tqEEJg5cya+/PJLbN26FdHR0S1eHzJkCLRabYvrfuzYMeTl5fG6t8Ott96KQ4cOYf/+/bZHfHw8Jk2aZPs7r7N9jBgx4pLpDLKzs9GtWzcAQHR0NEJCQlpc68rKSuzatYvXuh1qa2uhUrX81aZWq2G1WgHwOjtCW65pYmIiysvLkZGRYdtn69atsFqtSEhIsH9Rdh+ifB1avXq10Ov1YuXKleLIkSPiscceE35+fsJoNMpdmst6/PHHha+vr0hLSxOFhYW2R21trW2fP//5zyIyMlJs3bpV7N27VyQmJorExEQZq1aGi++WEoLX2V52794tNBqNeO2118Tx48fFJ598Ijw8PMR//vMf2z4LFy4Ufn5+4quvvhIHDx4Ud911F29RbqcpU6aI8PBw263ga9euFUFBQeKvf/2rbR9e5/arqqoS+/btE/v27RMAxOLFi8W+ffvE6dOnhRBtu6bjxo0TcXFxYteuXWL79u2iV69evBXc2b3zzjsiMjJS6HQ6MWzYMPHzzz/LXZJLA9DqY8WKFbZ96urqxBNPPCH8/f2Fh4eH+N3vficKCwvlK1ohfhtueJ3t55tvvhH9+/cXer1e9O3bV7z//vstXrdarWLevHnCYDAIvV4vbr31VnHs2DGZqnVNlZWVYtasWSIyMlK4ubmJ7t27ixdeeEGYTCbbPrzO7ffDDz+0+jN5ypQpQoi2XdNz586JiRMnCi8vL+Hj4yOmTp0qqqqqHFKvJMRF0zYSERERuTiOuSEiIiJFYbghIiIiRWG4ISIiIkVhuCEiIiJFYbghIiIiRWG4ISIiIkVhuCEiIiJFYbghIqf30ksvwWAwQJIkrFu3Tu5yMGrUKDz99NOyn8Oe5yFSEoYbInIIewWRrKwsvPzyy/jf//1fFBYWYvz48ddeXBulpaVBkqRLFg6Vw+VqWbt2LV599VV5iiJyUhq5CyCizmOxWCBJ0iULCzqznJwcAMBdd90FSZJkrsb5BAQEyF0CkdNxnZ9wRNeZUaNGYebMmZg5cyZ8fX0RFBSEefPm4eIVU0wmE5555hmEh4fD09MTCQkJSEtLs72+cuVK+Pn54euvv0ZsbCz0ej3y8vJgMpnw3HPPISIiAnq9Hj179sSHH35oO+7w4cMYP348vLy8YDAY8PDDD6O0tLRFbU899RT++te/IiAgACEhIXjppZdsr0dFRQEAfve730GSJNvz1hw6dAi33HIL3N3dERgYiMceewzV1dUAmrqjJkyYAABQqVSXDTfNrRqbNm1CXFwc3N3dccstt6C4uBjfffcdYmJi4OPjgwcffBC1tbW246xWK1JSUhAdHQ13d3cMGjQI//3vfwEAp06dwujRowEA/v7+kCQJjzzySItjL/f5H330Udxxxx0tamxoaEBwcHCL63yx//u//0N8fDy8vb0REhKCBx98EMXFxVet5bfdUufPn8fkyZPh7+8PDw8PjB8/HsePH7e93vw9sWnTJsTExMDLywvjxo1DYWFhq3URuSSHrFhFRNds5MiRwsvLS8yaNUscPXpU/Oc//xEeHh4tFlucNm2aGD58uNi2bZs4ceKEeOONN4RerxfZ2dlCCCFWrFghtFqtGD58uNixY4c4evSoqKmpEX/4wx9ERESEWLt2rcjJyRFbtmwRq1evFkIIcf78edGlSxcxd+5ckZWVJTIzM8WYMWPE6NGjW9Tm4+MjXnrpJZGdnS0++ugjIUmS+P7774UQQhQXF9sWOi0sLBTFxcWtfsbq6moRGhoq7rnnHnHo0CGRmpoqoqOjbYvxVVVViRUrVggAtpXhW9O8qN+NN94otm/fLjIzM0XPnj3FyJEjxdixY0VmZqbYtm2bCAwMFAsXLrQd9/e//1307dtXbNy4UeTk5IgVK1YIvV4v0tLSRGNjo/jiiy8EAHHs2DFRWFgoysvL2/T5d+zYIdRqtTh79qztvdauXSs8PT1tCwX+doHSDz/8UGzYsEHk5OSI9PR0kZiYKMaPHy+EEFet5eLz3HnnnSImJkZs27ZN7N+/XyQnJ4uePXsKs9nc4nsiKSlJ7NmzR2RkZIiYmBjx4IMPtnptiVwRww2Rkxo5cqSIiYkRVqvVtu25554TMTExQgghTp8+LdRqtSgoKGhx3K233irmzp0rhBC2YLB//37b68eOHRMAxObNm1t931dffVWMHTu2xbb8/HzbL9bm2m666aYW+wwdOlQ899xztucAxJdffnnFz/j+++8Lf39/UV1dbdu2fv16oVKphNFoFEII8eWXX4qr/T+sOdxs2bLFti0lJUUAEDk5ObZtf/rTn0RycrIQQoj6+nrh4eEhdu7c2eJcf/zjH8XEiRNbnPf8+fMt9mnL54+NjRWLFi2yPZ8wYYJ45JFHWpzj4lDyW3v27BEAbGHoSrU0nyc7O1sAEDt27LC9XlpaKtzd3cVnn30mhPj1e+LEiRO2fZYuXSoMBsNlayFyNeyWInJiN954Y4uumMTERBw/fhwWiwWHDh2CxWJB79694eXlZXv8+OOPtnEqAKDT6TBw4EDb8/3790OtVmPkyJGtvueBAwfwww8/tDhn3759AaDFeS8+JwCEhobaulHaKisrC4MGDYKnp6dt24gRI2C1WnHs2LF2neu3NRkMBnh4eKB79+4ttjXXeOLECdTW1mLMmDEtPuvHH3/c4nO25b2ASz//tGnTsGLFCgBAUVERvvvuOzz66KOXPV9GRgYmTJiAyMhIeHt7274+eXl5bfjkTbKysqDRaJCQkGDbFhgYiD59+iArK8u2zcPDAz169Lhs7USujgOKiVxUdXU11Go1MjIyoFarW7zm5eVl+7u7u3uLgOTu7n7V806YMAGLFi265LXQ0FDb37VabYvXJEmC1Wpt12ewt4trkiTpijU2j+tZv349wsPDW+yn1+vb9V6/PTcATJ48Gc8//zzS09Oxc+dOREdH4+abb271XDU1NUhOTkZycjI++eQTdOnSBXl5eUhOTobZbL5qLe3VWu3iorFcRK6O4YbIie3atavF859//hm9evWCWq1GXFwcLBYLiouLL/tLszUDBgyA1WrFjz/+iKSkpEtev+GGG/DFF18gKioKGk3Hf0RotVpYLJYr7hMTE4OVK1eipqbG1nqzY8cOqFQq9OnTp8Pv3RYXD7C+XCuWTqcDgKt+jtYEBgbi7rvvxooVK5Ceno6pU6dedt+jR4/i3LlzWLhwISIiIgAAe/fubXctMTExaGxsxK5duzB8+HAAwLlz53Ds2DHExsa2+zMQuSp2SxE5sby8PMyZMwfHjh3Dp59+infeeQezZs0CAPTu3RuTJk3C5MmTsXbtWuTm5mL37t1ISUnB+vXrL3vOqKgoTJkyBY8++ijWrVuH3NxcpKWl4bPPPgMAzJgxA2VlZZg4cSL27NmDnJwcbNq0CVOnTm3XL/moqCikpqbCaDTi/Pnzre4zadIkuLm5YcqUKTh8+DB++OEHPPnkk3j44YdhMBjacaXaz9vbG8888wxmz56Njz76CDk5OcjMzMQ777yDjz76CADQrVs3SJKEb7/9FiUlJbbWnraaNm0aPvroI2RlZWHKlCmX3S8yMhI6nQ7vvPMOTp48ia+//vqSuWvaUkuvXr1w1113Yfr06di+fTsOHDiAhx56COHh4bjrrrvaVTuRK2O4IXJikydPRl1dHYYNG4YZM2Zg1qxZeOyxx2yvr1ixApMnT8Zf/vIX9OnTB3fffTf27NmDyMjIK573vffew7333osnnngCffv2xfTp01FTUwMACAsLw44dO2CxWDB27FgMGDAATz/9NPz8/No1P85bb72FzZs3IyIiAnFxca3u4+HhgU2bNqGsrAxDhw7Fvffei1tvvRXvvvtum9/nWrz66quYN28eUlJSEBMTg3HjxmH9+vWIjo4GAISHh+Pll1/G888/D4PBgJkzZ7br/ElJSQgNDUVycjLCwsIuu1+XLl2wcuVKfP7554iNjcXChQvx5ptvttinrbWsWLECQ4YMwR133IHExEQIIbBhw4ZLuqKIlEwS7GglckqjRo3C4MGDsWTJErlLoQ6qrq5GeHg4VqxYgXvuuUfucoiuGxxzQ0RkZ1arFaWlpXjrrbfg5+eHO++8U+6SiK4rDDdERHaWl5eH6OhodO3aFStXrrymgdlE1H7sliIiIiJF4YBiIiIiUhSGGyIiIlIUhhsiIiJSFIYbIiIiUhSGGyIiIlIUhhsiIiJSFIYbIiIiUhSGGyIiIlIUhhsiIiJSlP8P6BEXiUT27GYAAAAASUVORK5CYII=\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Уровень покрытия и метелирования"
      ],
      "metadata": {
        "id": "Yh76j0rKXQzD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "! bedGraphToBigWig  s_8_cell.deduplicated.bedGraph mm10.chrom.sizes s_8_cell_methylation.bigWig.bw\n",
        "! bedGraphToBigWig  s_epiblast.deduplicated.bedGraph mm10.chrom.sizes epiblast_methylation.bigWig.bw\n",
        "! bedGraphToBigWig  s_icm.deduplicated.bedGraph mm10.chrom.sizes icm_methylation.bigWig.bw"
      ],
      "metadata": {
        "id": "3TFjav6IXUUN"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "! bedGraphToBigWig   s_8_cell.cover.bedGraph mm10.chrom.sizes s_8_cell_coverage.bigWig.bw\n",
        "! bedGraphToBigWig   s_epiblast.cover.bedGraph mm10.chrom.sizes epiblast_coverage.bigWig.bw\n",
        "! bedGraphToBigWig   s_icm.cover.bedGraph mm10.chrom.sizes icm_coverage.bigWig.bw"
      ],
      "metadata": {
        "id": "86o0n7DOYj99"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "! make_tracks_file --trackFiles s_8_cell_coverage.bigWig.bw epiblast_coverage.bigWig.bw icm_coverage.bigWig.bw -o tracks_cov.ini.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yz8mmWZ3ZMyf",
        "outputId": "231e9175-ca4b-455a-cf88-a53c01d8fe63"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Adding bigwig file: s_8_cell_coverage.bigWig.bw\n",
            "Adding bigwig file: epiblast_coverage.bigWig.bw\n",
            "Adding bigwig file: icm_coverage.bigWig.bw\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! make_tracks_file --trackFiles s_8_cell_methylation.bigWig.bw epiblast_methylation.bigWig.bw icm_methylation.bigWig.bw -o tracks_met.ini.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EJ9OimvBZSgK",
        "outputId": "77747751-2ebe-4a08-b00a-b2ae688b8bfd"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Adding bigwig file: s_8_cell_methylation.bigWig.bw\n",
            "Adding bigwig file: epiblast_methylation.bigWig.bw\n",
            "Adding bigwig file: icm_methylation.bigWig.bw\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! pyGenomeTracks --tracks tracks_cov.ini.txt --region chr11:3100030-3500030 -o image_cov.png"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "nTiF7VruZVLJ",
        "outputId": "e42b8ac1-c35c-405c-d4a1-cee53f910729"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO:pygenometracks.tracksClass:initialize 1. [x-axis]\n",
            "INFO:pygenometracks.tracksClass:initialize 2. [spacer]\n",
            "INFO:pygenometracks.tracksClass:initialize 3. [s_8_cell_coverage.bigWig]\n",
            "INFO:pygenometracks.tracksClass:initialize 4. [epiblast_coverage.bigWig]\n",
            "INFO:pygenometracks.tracksClass:initialize 5. [icm_coverage.bigWig]\n",
            "INFO:pygenometracks.tracksClass:time initializing track(s):\n",
            "INFO:pygenometracks.tracksClass:0.02244281768798828\n",
            "DEBUG:pygenometracks.tracksClass:Figure size in cm is 40 x 8.909574468085108. Dpi is set to 72\n",
            "\n",
            "INFO:pygenometracks.tracksClass:plotting 1. [x-axis]\n",
            "INFO:pygenometracks.tracksClass:plotting 2. [spacer]\n",
            "INFO:pygenometracks.tracksClass:plotting 3. [s_8_cell_coverage.bigWig]\n",
            "INFO:pygenometracks.tracksClass:plotting 4. [epiblast_coverage.bigWig]\n",
            "INFO:pygenometracks.tracksClass:plotting 5. [icm_coverage.bigWig]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! pyGenomeTracks --tracks tracks_met.ini.txt --region chr11:3100030-3500030 -o image_met.png"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "LTocMAJLZaId",
        "outputId": "44280817-72d7-459d-9494-e353d2b54e95"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "INFO:pygenometracks.tracksClass:initialize 1. [x-axis]\n",
            "INFO:pygenometracks.tracksClass:initialize 2. [spacer]\n",
            "INFO:pygenometracks.tracksClass:initialize 3. [s_8_cell_methylation.bigWig]\n",
            "INFO:pygenometracks.tracksClass:initialize 4. [epiblast_methylation.bigWig]\n",
            "INFO:pygenometracks.tracksClass:initialize 5. [icm_methylation.bigWig]\n",
            "INFO:pygenometracks.tracksClass:time initializing track(s):\n",
            "INFO:pygenometracks.tracksClass:0.01838374137878418\n",
            "DEBUG:pygenometracks.tracksClass:Figure size in cm is 40 x 8.909574468085108. Dpi is set to 72\n",
            "\n",
            "INFO:pygenometracks.tracksClass:plotting 1. [x-axis]\n",
            "INFO:pygenometracks.tracksClass:plotting 2. [spacer]\n",
            "INFO:pygenometracks.tracksClass:plotting 3. [s_8_cell_methylation.bigWig]\n",
            "INFO:pygenometracks.tracksClass:plotting 4. [epiblast_methylation.bigWig]\n",
            "INFO:pygenometracks.tracksClass:plotting 5. [icm_methylation.bigWig]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "QXAuFR3Yqv2F"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}